function global-variable {

  # SPECIES must be set before the call of this bash function. $SPECIESFILE
  # contains a list of Genbank formatted genome file names.
  SPECIESFILE=$ROOTANALYSISDIR/species/$SPECIES

  # The subdirectory output contains directories named after the species file.
  # The output species directory would contain all of the results from the
  # analysis.
  # The output species directory would contain 5 subdirectories. 
  # run-mauve contains genome alignments.
  # run-lcb contains alignment blocks that are generated by the genome
  # alignment.
  # run-clonalframe contains the reference species tree estimated by ClonalFrame.
  # run-clonalorigin contains the result from ClonalOrigin.
  # run-analysis should contain all the results. The final results must be from
  # this directory. A manuscript in the subdirectory doc/README will have
  # special words that will be replaced by values from files in run-analysis.
  # Whenever I change the analysis, values must be changed accordingly.
  OUTPUTDIR=$ROOTANALYSISDIR/output
  BASEDIR=$OUTPUTDIR/$SPECIES
  BASERUNANALYSIS=$BASEDIR/run-analysis
  NUMBERDIR=$BASEDIR/$REPETITION
  DATADIR=$NUMBERDIR/data
  BWADIR=$NUMBERDIR/bwa
  TRANSCRIPTDIR=$NUMBERDIR/transcript
  MAQDIR=$NUMBERDIR/maq
  BOWTIEDIR=$NUMBERDIR/bowtie
  RUNANALYSIS=$NUMBERDIR/run-analysis

  # The cluster has almost the same file system. I used to used Samba client to
  # use the file system of the cluster. This stopped working. I did not know the
  # reason, which I did not want to know. Since then, I use scp command.
  # Note that the cluster base directory does not contain run-analysis. The
  # basic analysis is done in the local machine.
  CAC_BASEDIR=$CAC_OUTPUTDIR/$SPECIES
  CAC_NUMBERDIR=$CAC_OUTPUTDIR/$SPECIES/$REPETITION_DIR
  CAC_DATADIR=$CAC_NUMBERDIR/data
  CAC_RUNANALYSIS=$CAC_NUMBERDIR/run-analysis

  X11_BASEDIR=$X11_OUTPUTDIR/$SPECIES
  X11_NUMBERDIR=$X11_OUTPUTDIR/$SPECIES/$REPETITION_DIR
  X11_DATADIR=$X11_NUMBERDIR/data
  X11_RUNANALYSIS=$X11_NUMBERDIR/run-analysis
}

