%\VignetteIndexEntry{Analyzing Streptococcus mutans RNA-Seq data}
%\VignettePackage{smutans}

% To compile this document
% library('weaver'); rm(list=ls()); Sweave('smutans.Rnw', driver=weaver()); system('pdflatex smutans')

\documentclass{article}

\usepackage{Sweave}
\usepackage[letterpaper]{geometry}
\usepackage{hyperref,graphicx}
\usepackage{color}

\SweaveOpts{keep.source=TRUE,eps=FALSE,include=FALSE,width=4,height=4.5}
\newcommand{\Robject}[1]{\texttt{#1}}
\newcommand{\Rpackage}[1]{\textit{#1}}
\newcommand{\Rclass}[1]{\textit{#1}}
\newcommand{\Rfunction}[1]{{\small\texttt{#1}}}
\newcommand{\fixme}[1]{{\textbf{Fixme:} \textit{\textcolor{blue}{#1}}}}

\renewcommand{\floatpagefraction}{0.7}

\author{Sang Chul Choi\\[1em]Cornell University,\\ Ithaca, NY\\[1em]
\texttt{schoi@cornell.edu}}

\title{\textsf{\textbf{Analyzing \textit{Streptococcus mutans} RNA-Seq data using R packages}}}

\begin{document}

\maketitle

\begin{abstract}
We analyze a number of RNA samples from \textit{Streptococcus mutans} using R
packages: \Rpackage{DESeq}, \Rpackage{edgeR}, and \Rpackage{goseq}.
Tasks include the analysis of count data from RNA-Seq for
the detection of differentially expressed genes, and tests of enrichment for
differentially expressed genes using \Rpackage{goseq}.  The
package \Rpackage{smutans} contains procedures that make use of functions
provided by the above R packages.  
\end{abstract}

\tableofcontents

%--------------------------------------------------
\section{Data preparations} \label{sec:prep}
%--------------------------------------------------

The \Rpackage{smutans} package expects count data in the form of
\Rclass{CountDataSet}. Using BWA and samtools we aligned short reads from RNA
sequencing. Using gene annotations in the UA159 genome we counted short reads
that overlapped the genes.  The count values are raw counts of sequencing reads.
\Rpackage{DESeq}'s statistical model, or negative binomial model, hold for raw
reads that allows to assess the measurement precision correctly.  We did not
have rounded values of normalized counts, or counts of covered base pairs.

Several exeriments of \textit{S.\ mutans} RNA sampling were performed.
\fixme{Find the papers that are related with the experiments.} One of them was
presented in the object called \Robject{smutansGenes}.  The
\Robject{smutansGenes} object is of class \Rclass{CountDataSet}, which is the
data container used by \Rpackage{DESeq}. We have a covenient class called
\Rclass{Smutans}. The class contains slots for \Rclass{CountDataSet} objects,
and results from analyses.  Let's load the needed packages and the data as
follows.

We can create \Rclass{CountDataSet} objects using functions that are not
exported. One should create an R script at the directory that contains the
source code of the \Rpackage{smutans}. Put the following scripts in the R script
and run the script at the directory. That should create RData files at
the directory smutans/data: sm835Genes.RData, smutansGenes.RData.

http://www.oralgen.lanl.gov/oralgen/downloads/smut.csv

http://www.oralgen.lanl.gov/cgi-bin/gene_id_search.cgi?dbname=smut&gene_id=SMu0001

\fixme{Where are flat files for genes of S. mutans UA159?}

<<samplePreparation,eval=FALSE>>=
#not run
library(DESeq)
source("smutans/R/core.R")
smutans.prepareData34( )
smutans.prepareData835NPP( )
smutans.prepareDataUA159TW1( ) 
smutans.prepareGoseq( )
q("no")
@

<<options,results=hide,echo=FALSE>>=
options(digits=3)
@

<<smutans2,results=hide>>=
library( "smutans" )
data( "smutansGenes" )
@

There are \Sexpr{length(colnames(counts(smutansGenes)))} RNA-seq samples in
\Robject{smutansGenes} that can be loaded from the package.
\Robject{smutansGenes} contains the counts and also metadata about the samples:

<<look_into_smutansGenes>>=
head( counts(smutansGenes) )
@

<<look_into_smutansGenes>>=
pData( smutansGenes )
@

The samples differ by experimental condition (glucose or
galactose) and by library type (ua159 or tw1). First, we only look at the ua159
data for now. We will look at the tw1 data later. 
We create an \Rclass{Smutans} object using the \Rclass{CountDataSet} object,
\Robject{smutansGenes}.

%--------------------------------------------------
\section{Differential expression analysis} \label{sec:de}
%--------------------------------------------------

%------------------------------------------------------------
\subsection{Inference: Using the UA159 samples for condition effect} 
\label{subsec:ua159only}
%------------------------------------------------------------

<<ua159onlySamples>>=
ua159only <- newSmutans( smutansGenes )
ua159only <- smutans.de2type( ua159only, type="ua159", 
                              condA="glucose", condB="galactose" )
@

\Rfunction{smutans.de2type} uses a two-factor data set to test differential
expression of genes. We compare 4 pairwise comparisons using a two-factor data
set. A pair of comparisons between glucose and galactose given either type ua159
or tw1. Another pairs of comparisons between types ua159 and tw1 given either
condition glucose or galactose. We focus on one of the four pairwise comparisons
first. Later, we will perform analyses for the rest of the three comparisons.
Note that the \Rclass{Smutans} contains \Rclass{CountDataSet} objects.
\Rfunction{smutans.de2type} estimates the effective library size, and the
dispersions using functions featured by \Rpackage{DESeq}. Then, it tests for
differential expression.

Because \Rpackage{DESeq} recommended to check
the dispersion estimates and to make sure that the data quality is as
expected, we plot the per-gene estimates against the normalized mean
expressions per gene, and then overlay the fitted curve in red
(Fig.\ \ref{figFit}).

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figFit}
\caption{Empirical (black dots) and fitted (red lines) dispersion values
plotted against mean expression strength.}
\label{figFit}
\end{figure}
<<figFit,fig=TRUE>>=
smutans.plotDispersionEstimates(ua159only)
@

As we estimated the dispersion from only two samples, we should expect the
estimates to scatter with quite some sampling variance around their true values.
Hence, \Rpackage{DESeq} should not use the per-gene estimates directly in the
test, because using too low dispersion values leads to false positives.  Many of
the values below the red line are likely to be underestimates of the true
dispersions, and hence, it is prudent to instead rather use the fitted value.
On the othe hand, not all of he values above the red line are overestimations,
and hence, the conservative choice is to keep them instead of replacing them
with their fitted values. 

Having estimated the dispersion for each gene, we look for differentially
expressed genes. To contrast two conditions, e.g., to see whether there is
differential expression between conditions ``glucose'' and ``galactose'',
\Rfunction{smutans.de2type} already performed the function
\Rfunction{nbinomTest} provided by \Rpackage{DESeq}.  Let us first plot the
$\log_2$ fold changes against the base means, colouring in red those genes that
are significant at 10\% FDR.

For each gene, we get its mean expression level (at the base scale) as a joint
estimate from both conditions, and estimated separately for each  condition, the
fold change from the first to the second condition, the logarithm (to basis 2)
of the fold change, and the $p$ value for the statistical significance of this
change. The \texttt{padj} column contains the $p$ values, adjusted for multiple
testing with the Benjamini-Hochberg procedure (see the R function
\Rfunction{p.adjust}), which controls false discovery rate (FDR).

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figDE}
\caption{Plot of normalised mean versus $log_2$ fold change
(this plot is sometimes also called the ``MA-plot'') for the contrast ``glucose''
versus ``galactose''.}
\label{figDE}
\end{figure}
<<figDE,fig=TRUE>>=
smutans.plotDiffExp(ua159only)
@ %$
See Figure~\ref{figDE} for the plot. 

We continue to analyze the data set using type tw1 to compare the two
conditions. Now, it should be straightfoward to perform the single-factor
analysis. Figures \ref{figFittw1only} and \ref{figDEtw1only} shows the
dispersion distribution and MA-plot for the contrast ``glucose'' versus
``galactose'' using type tw1 only. 
<<tw1onlySamples>>=
tw1only <- newSmutans( smutansGenes )
tw1only <- smutans.de2type( tw1only, type="tw1", 
                            condA="glucose", condB="galactose" )
@

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figFittw1only}
\caption{Empirical (black dots) and fitted (red lines) dispersion values
plotted against mean expression strength.}
\label{figFittw1only}
\end{figure}
<<figFittw1only,fig=TRUE>>=
smutans.plotDispersionEstimates(tw1only)
@

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figDEtw1only}
\caption{Plot of normalised mean versus $log_2$ fold change
(this plot is sometimes also called the ``MA-plot'') for the contrast ``glucose''
versus ``galactose''.}
\label{figDEtw1only}
\end{figure}
<<figDEtw1only,fig=TRUE>>=
smutans.plotDiffExp(tw1only)
@

Note that \Rfunction{smutans.de2} already performed the negative bionomial test
for each of the genes. We fetch adjusted p-values from the single-factor
analysis and the two-factor one to compare the number of genes that are found to
be differentially expressed using the same FDR rate of 0.1.
<<compareUA159onlyAndTW1only>>=
padj.ua159only <- smutans.padj( ua159only )
padj.tw1only <- smutans.padj( tw1only )
tabUA159onlyTW1only = table( "ua159 only" = padj.ua159only < .1, "tw1 only" = padj.tw1only < .1 )
addmargins( tabUA159onlyTW1only )
@
Of the \Sexpr{sum(tabUA159onlyTW1only)} gene, \Sexpr{tabUA159onlyTW1only[2,2]}
genes are found in both analyses. The analysis using all of tw1 only found
\Sexpr{tabUA159onlyTW1only[1,2]} genes. The analysis using only ua159 found
\Sexpr{tabUA159onlyTW1only[2,1]} genes. 

Until now we focused on the effect of type not condition. 
<<glucoseOnlySamples>>=
glucoseOnly <- newSmutans( smutansGenes, title="glucoseOnly" )
glucoseOnly <- smutans.de2( glucoseOnly, condition="glucose", 
                            condA="ua159", condB="tw1" )
galactoseOnly <- newSmutans( smutansGenes, title="galactoseOnly" )
galactoseOnly <- smutans.de2( galactoseOnly, condition="galactose", 
                              condA="ua159", condB="tw1" )
@

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figFitGlucoseOnly}
\caption{Empirical (black dots) and fitted (red lines) dispersion values
plotted against mean expression strength. \textbf{figFitGlucoseOnly}}
\label{figFitGlucoseOnly}
\end{figure}
<<figFitGlucoseOnly,fig=TRUE>>=
smutans.plotDispersionEstimates(glucoseOnly)
@

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figDEGlucoseOnly}
\caption{Plot of normalised mean versus $log_2$ fold change
(this plot is sometimes also called the ``MA-plot'') for the contrast ``ua159''
versus ``tw1''. \textbf{figDEGlucoseOnly}}
\label{figDEGlucoseOnly}
\end{figure}
<<figDEGlucoseOnly,fig=TRUE>>=
smutans.plotDiffExp(glucoseOnly)
@

Note that \Rfunction{smutans.de2} already performed the negative bionomial test
for each of the genes. We fetch adjusted p-values from the single-factor
analysis and the two-factor one to compare the number of genes that are found to
be differentially expressed using the same FDR rate of 0.1.

\fixme{Two objects and one qvalue cutoff can be handled by a function. Let's
make a function in core.R.}
<<compareGlucoseOnlyAndGalactoseOnly>>=
#padj.glucoseOnly <- smutans.padj( glucoseOnly )
#padj.galactoseOnly <- smutans.padj( galactoseOnly )
#tabGlucoseOnlyGalactoseOnly = table( "glucose only" = padj.glucoseOnly < .1, 
#                                     "galactose only" = padj.galactoseOnly < .1 )
tabGlucoseOnlyGalactoseOnly <- smutans.compareResult( glucoseOnly, 
                                                      galactoseOnly )
addmargins( tabGlucoseOnlyGalactoseOnly )
@
Of the \Sexpr{sum(tabGlucoseOnlyGalactoseOnly)} gene, \Sexpr{tabGlucoseOnlyGalactoseOnly[2,2]}
genes are found in both analyses. The analysis using all of galactose only found
\Sexpr{tabGlucoseOnlyGalactoseOnly[1,2]} genes. The analysis using glucose only found
\Sexpr{tabGlucoseOnlyGalactoseOnly[2,1]} genes. 

%------------------------------------------------------------
\subsection{Inference: Using all of the samples for condition effect} 
\label{subsec:allSample}
%------------------------------------------------------------

% We must have at least one citation \cite{Anders:2010:GB}.

We could compare the two conditions, glucose and galactose, using either type
ua159 or tw1. This may or may not be desirable depending on the experimental
design. Using more samples might be more powerful in detecting differentially
expressed genes. Because we have a two-factor data set, \Robject{smutansGenes},
we wish to determine which genes are differentially expressed in one of the two
factors. Here, we first find genes differentially expressed in different
conditions of glucose and galactose using both types of ua159 and tw1. We will
compare those genes and those found in the single type analysis that we did
previously. Let's start the analysis by creating a new \Rclass{Smutans} object,
called \Robject{allSample}. 
<<allSamples>>=
allSample <- newSmutans( smutansGenes )
allSample <- smutans.de2( allSample )
@
\Rfunction{smutans.de2} is the basic function that uses a two-factor data set to
perform two-factor analysis. The two factors are type and condition. We are
interested in condition effect. We first check 
the dispersion estimates and to make sure that the data quality is as
expected, we plot the per-gene estimates against the normalized mean
expressions per gene, and then overlay the fitted curve in red
(Fig.\ \ref{figFit2}).

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figFit2}
\caption{Empirical (black dots) and fitted (red lines) dispersion values
plotted against mean expression strength.}
\label{figFit2}
\end{figure}

<<figFit2,fig=TRUE>>=
smutans.plotDispersionEstimates(allSample)
@

\fixme{Say something about Figure \ref{figFit2}.}

Note that \Rfunction{smutans.de2} already performed the negative bionomial test
for each of the genes. We fetch adjusted p-values from the single-factor
analysis and the two-factor one to compare the number of genes that are found to
be differentially expressed using the same FDR rate of 0.1.
<<compareUA159onlyAndAllSample>>=
padj.ua159only <- smutans.padj( ua159only )
padj.allSample <- smutans.padj( allSample )
tab = table( "ua159 only" = padj.ua159only < .1, "all samples" = padj.allSample < .1 )
addmargins( tab )
@
Of the \Sexpr{sum(tab)} gene, \Sexpr{tab[2,2]} genes are found in both analyses. The analysis using all
of the sample found \Sexpr{tab[1,2]} genes. The analysis using only ua159 found \Sexpr{tab[2,1]} genes. 
Figure \ref{figCompareUA159onlyAndAllSample} shows the raw p-values from the
two-factor analysis versus those from the single-factor analysis. The p-values
from the two-factor analysis are overall smaller than those from the
single-factor analysis. We would expect that this was the case because of higher
power of the two-factor analysis.

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figCompareUA159onlyAndAllSample}
\caption{Comparison of $p$ values (unadjusted) from the test using only the
four paired-end samples and the test using all seven samples. We can see that
the latter tend to be smaller, as expected from the higher power of the test with all samples.}
\label{figCompareUA159onlyAndAllSample}
\end{figure}

<<figCompareUA159onlyAndAllSample,fig=TRUE>>=
pval.ua159only <- smutans.pval( ua159only )
pval.allSample <- smutans.pval( allSample )
bottom = function(x, theta=1e-12) pmax(x, theta)
plot( bottom(pval.ua159only), bottom(pval.allSample), log="xy", pch=20, cex=.3 )
abline(a=0, b=1, col="blue")
@

\fixme{Check the fitted regression using type and condition. The following are
from \Rpackage{DESeq} document.} 
The first three columns show the fitted coefficients, converted to a logarithm base 2 scale.
The log2 fold change due to the condition is shown in the third column. As indicated
by the column name, it is the effect of ``untreated'', i.e., the log ratio of
``untreated'' versus ``treated''. (This is unfortunately the other way round as
before, due to the peculiarities of contrast coding.) Note that the library type also
had noticeable influence on the expression, and hence would have increased the dispersion
estimates (and so reduced power) if we had not fitted an effect for it.
The column \emph{deviance}
is the deviance of the fit. (Comparing the deviances with a $\chi^2$ likelihood ratio test is
how \Rfunction{nbinomGLMTest} calculates the $p$ values.) The last column, \emph{converged},
indicates whether the calculation of coefficients and deviance has fully converged.
(If it is false too often, you can try to change the GLM control parameters, as
explained in the help page to \Rfunction{fitNbinomGLMs}.)

Finally, we show that taking the library type into account was important to have
good detection power by doing the analysis again using the standard workflow,
as outlined earlier, and without informing \Rpackage{DESeq} of the library types:

<<allSampleWithoutType>>=
allSampleWithoutType <- newSmutans( smutansGenes )
allSampleWithoutType <- smutans.de2WithoutType( allSampleWithoutType,
                                                condA="glucose", 
                                                condB="galactose" )
@

<<compareUA159onlyAndAllSample>>=
padj.allSampleWithoutType <- smutans.padj( allSampleWithoutType )
tab2 <- table(
   `all samples simple` = padj.allSampleWithoutType < 0.1,
   `all samples GLM`    = padj.allSample < 0.1 )
addmargins( tab2 )
largeNumber <- 10
@
Of the \Sexpr{sum(tab2)} gene, \Sexpr{tab2[2,2]} genes are found in both analyses. The analysis using all
of the sample found \Sexpr{tab2[1,2]} genes. The analysis using all of the same
sample but without considering type found \Sexpr{tab2[2,1]} genes. 

We now proceed to compare the two types by considering all of the samples.
<<allSampleType>>=
allSampleType <- newSmutans( smutansGenes, title="All samples for type" )
allSampleType <- smutans.de2( allSampleType, compareCondition="no" )
smutans.plotDispersionEstimates(allSampleType)
tabGlucoseOnlyAllSample <- smutans.compareResult( glucoseOnly, 
                                                  allSampleType )
addmargins( tabGlucoseOnlyAllSample )
@
Of the \Sexpr{sum(tabGlucoseOnlyAllSample)} gene,
\Sexpr{tabGlucoseOnlyAllSample[2,2]} genes are found in both analyses. The
analysis using all of the sample found \Sexpr{tabGlucoseOnlyAllSample[1,2]}
genes. The analysis using all of the same sample but without considering type
found \Sexpr{tabGlucoseOnlyAllSample[2,1]} genes.  

%------------------------------------------------------------
\subsection{Heatmap analyses} 
\label{subsec:heatmap}
%------------------------------------------------------------

Potentially useful figures can be plotted from \Rpackage{DESeq}. Instead of
browsing differentially expressed genes, they can be plotted in group. The
original fold changes could be moderated to be amenable to distance calculation. 
Figure \ref{figLfc} shows moderated log fold changes versus the original log
fold changes.

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figLfc}
\caption{Scatterplot of direct (\Robject{lfc}) versus \emph{moderated} log-ratios (\Robject{mod\_lfc}).
  The moderation criterion used is variance stabilisation.
  The red points correspond to values that were infinite in \Robject{lfc}
  and were abitrarily set to $\pm\Sexpr{largeNumber}$ for the purpose of plotting.
  These values vary in a finite range (as shown in the plot) in \Robject{mod\_lfc}.}
\label{figLfc}
\end{figure}

<<figLfc,fig=TRUE>>=
select <- 1:40
smutans.de2TypeLfc( ua159only, condA="glucose", condB="galactose" )
@

Now we plot the first \Sexpr{length(select)} genes clustered based on the
moderated log fold changes.

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figHeatmap1}
\caption{Heatmaps showing the expression data of the top \Sexpr{length(select)}
  differentially expressed genes. Left, the variance stabilisation
  transformed data,
  right, the original count data.
  The right plot is dominated by a small number of data points with large values.}
\label{figHeatmap1}
\end{figure}

<<figHeatmap1,fig=TRUE>>=
smutans.de2TypeHeatmap( ua159only )
@

We can cluster samples using variance stabilized data. Figure \ref{figClust}
shows the 13 sampled clustered. 
The clustering correctly reflects our experimental design, i.e., samples are
more similar when they have the same treatment or the same library type.
(To make this conclusion, it was important to re-estimate e dispersion with
mode ``blind'' (in the calculation for \Robject{cdsFullBlind} above, as only then,
the variance stabilizing transformation is not informed about the design and
hence not biased towards a result supporting it.) Such an analysis is useful
for quality control, and (even though we finish our vignette with it), it may
be useful to this first in any analysis.

%\begin{figure}
%\centering
%\includegraphics[width=.9\textwidth]{smutans-figClust}
%\caption{Heatmap showing the Euclidean distances between the samples
%  as calculated from the variance-stabilising transformation of the
%  count data.}
%\label{figClust}
%\end{figure}

<<figClust,fig=TRUE>>=
# smutans.de2Clust( allSample )
@

We find some functional categories enriched with differentially expressed genes.
<<goseqUA159only>>=
data( "smutans.feature.genes" )
data( "smutans.go.genes" )
data( "smutans.cat.desc" )
smutans.de2Goseq ( ua159only, qval=1e-2, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc )
@

%--------------------------------------------------
\section{Differential expression analysis using 835NP, 835P, and UA159 no CSP}
\label{sec:de835}
%--------------------------------------------------

A count data set is a table of numbers of short reads that are assigned to
predefined genes. We use one of R packages for testing for differential
expression: \Rpackage{DESeq}. 

To find functional categories enriched with differentially expressed genes we
first find a list of differentially expressed genes. A set of genes are assigned
to functional gene categories: \Robject{smutans.go.genes}. Features of the genes in
UA159 include gene lengths: \Robject{smutans.feature.genes}.
Functional categories with numbers of genes assigned are in
\Robject{smutans.cat.desc}.

We use RNA-seq data to define genes whose expression changes between
samples \cite{Oshlack2010}. Short reads are first mapped to a
reference genome using approaches such as BWA \cite{Li2009a}.
Reads fitting into one of the annotated intervals are counted.
These counts are passed to a statistical package which treats differences
in read counts for statistical significance. For this purpose, we made
use of DEseq \cite{Anders2010}. This prodecure returns a list of
differentially expressed genes, which can be passed to downstream
analysis, such as functionaly category enrichment test (e.g., Young et
al., 2010).

The abtract briefly describes main steps of analyzing RNA-seq data for
determining differentially expressed genes \cite{Oshlack2010}. We
used a short read aligner called BWA \cite{Li2009a} to map the
sequenced short reads on a reference genome, \textit{Streptococcus mutans}
UA159 (NCBI accession NC\_004350). Because the aligner allowed a few gaps
for efficient alignment of millions of sequences of size being 100 base
pairs long, short reads that consist of RNAs and sequencing adapter
sequences would be not mapped. We used cutadapt \cite{Martin2011} to remove
parts of adapter sequences from short reads before mapping on the reference
genome. Mapped short read alignments were converted into readable formats
\cite{Li2009b}. We counted short reads aligned to annotated genes in the
reference genome, forming a table of read counts of the genes. Statistical
software R \cite{R2011} with its one of R package
called DEseq \cite{Anders2010} were subsequently employed to infer
significantly differentially expressed in a pair of biological conditions.
We also generated new tracks for our recently released Streptococcus Genome
Browser (http://strep-genome.bscb.cornell.edu) that summarize gene
expression levels based the short read alignments. This allowed us to
visualize parts of the reference genome for comparing different RNA
samples. Further analyses such as functionaly category enrichment test
\cite{Young2010} were pursued.

We could find no functional categories enriched with differentially expressed
genes in the pairwise comparisons of 835NP, 835P, and UA159 no CSP.

Let's make a list of differentially expressed genes. It could be interesting to
see the cluster of the three kinds of samples. Another kind of heatmap plots for
grouping genes could be interesting. But, basically we have a list of genes with
some adjusted p-values for detecting differentially expressed genes.

<<sm835,results=hide>>=
data( "sm835Genes" )
@

<<look_into_smutansGenes>>=
head( counts(sm835Genes) )
pData( sm835Genes )
@

<<sm835Samples>>=
npAndP <- newSmutans( sm835Genes, title="835NP vs. 835P" )
npAndP <- smutans.de2( npAndP, type="ua159", 
                       condA="835np", condB="835p" )
npAndUA159 <- newSmutans( sm835Genes, title="835NP vs. UA159" )
npAndUA159 <- smutans.de2( npAndUA159, type="ua159", 
                           condA="835np", condB="ua159nocsp" )
pAndUA159 <- newSmutans( sm835Genes, title="835P vs. UA159" )
pAndUA159 <- smutans.de2( pAndUA159, type="ua159", 
                          condA="835p", condB="ua159nocsp" )
@

\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figFitNpAndP}
\includegraphics[width=.6\textwidth]{smutans-figFitNpAndUA159}
\includegraphics[width=.6\textwidth]{smutans-figFitPAndUA159}
\caption{Empirical (black dots) and fitted (red lines) dispersion values
plotted against mean expression strength. \textbf{figFitNpAndP}}
\label{figFitNpAndP}
\end{figure}
<<figFitNpAndP,fig=TRUE>>=
smutans.plotDispersionEstimates(npAndP)
@
<<figFitNpAndUA159,fig=TRUE>>=
smutans.plotDispersionEstimates(npAndUA159)
@
<<figFitPAndUA159,fig=TRUE>>=
smutans.plotDispersionEstimates(pAndUA159)
@

<<sm835Samples>>=
smutans.de2List( npAndP, "npAndP.csv" )
smutans.de2List( npAndUA159, "npAndUA159.csv" )
smutans.de2List( pAndUA159, "pAndUA159.csv" )
@

%\begin{figure}
%\centering
%\includegraphics[width=.9\textwidth]{smutans-figClustNpAndP}
%\caption{Empirical (black dots) and fitted (red lines) dispersion values
%plotted against mean expression strength. \textbf{figClustNpAndP}}
%\label{figClustNpAndP}
%\end{figure}
<<figClustNpAndP,fig=TRUE>>=
# smutans.de2Clust( npAndP )
@
%\begin{figure}
%\centering
%\includegraphics[width=.9\textwidth]{smutans-figClustNpAndUA159}
%\caption{Empirical (black dots) and fitted (red lines) dispersion values
%plotted against mean expression strength. \textbf{figClustNpAndUA159}}
%\label{figClustNpAndUA159}
%\end{figure}
<<figClustNpAndUA159,fig=TRUE>>=
# smutans.de2Clust( npAndUA159 )
@
%\begin{figure}
%\centering
%\includegraphics[width=.9\textwidth]{smutans-figClustPAndUA159}
%\caption{Empirical (black dots) and fitted (red lines) dispersion values
%plotted against mean expression strength. \textbf{figClustPAndUA159}}
%\label{figClustPAndUA159}
%\end{figure}
<<figClustPAndUA159,fig=TRUE>>=
# smutans.de2Clust( pAndUA159 )
@

\begin{figure}
\centering
\includegraphics[width=.9\textwidth]{smutans-figHeatmapNpAndP}
\caption{Empirical (black dots) and fitted (red lines) dispersion values
plotted against mean expression strength. \textbf{figHeatmapNpAndP}}
\label{figHeatmapNpAndP}
\end{figure}
<<figHeatmapNpAndP,fig=TRUE>>=
smutans.de2TypeHeatmap( npAndP )
@
\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figHeatmapNpAndUA159}
\caption{Empirical (black dots) and fitted (red lines) dispersion values
plotted against mean expression strength. \textbf{figHeatmapNpAndUA159}}
\label{figHeatmapNpAndUA159}
\end{figure}
<<figHeatmapNpAndUA159,fig=TRUE>>=
smutans.de2TypeHeatmap( npAndUA159 )
@
\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{smutans-figHeatmapPAndUA159}
\caption{Empirical (black dots) and fitted (red lines) dispersion values
plotted against mean expression strength. \textbf{figHeatmapPAndUA159}}
\label{figHeatmapPAndUA159}
\end{figure}
<<figHeatmapPAndUA159,fig=TRUE>>=
smutans.de2TypeHeatmap( pAndUA159 )
@

\fixme{OMZ175, UA159CSP, SMU86CSP pairwise comparisons should be done fairly
easily after detailed analyses.}

\section{Session Info}
<<sessi>>=
sessionInfo()
@

\bibliographystyle{unsrt}
\bibliography{siepel-mutans}

\end{document}
