%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Load smutans package and data
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<eval=FALSE,echo=FALSE,results=hide>>=
# not run when smutans R package is installed
# Use this when smutans R package is being tested.
library(DESeq)
library(goseq)
library(GenomicRanges)
library(IRanges)
library(plotrix)
library(xtable)
library(RColorBrewer)
library(lattice)
library(rtracklayer)
source("smutans/R/class_and_slots.R")
source("smutans/R/core.R")
source("smutans/R/methods.R")
load("smutans/data/smutansData.tx.RData")
load("smutans/data/smutansData.txGenes.RData")
load("smutans/data/smutansData.txGrCombined.RData")
load("smutans/data/smutansData.txAnnotation.RData")
load("smutans/data/smutansData.txRNAz.RData")
load("smutans/data/smutansData.txPileup.RData")
load("smutans/data/smutans.feature.genes.RData")
load("smutans/data/smutans.genes.criteria.RData")
load("smutans/data/smutans.go.genes.RData")
load("smutans/data/smutans.cat.desc.RData")
load("smutans/data/smutans.keg.genes.RData")
load("smutans/data/smutans.keg.desc.RData")
load("smutans/data/smutansGenes.RData")
@
<<load,echo=FALSE,results=hide>>=
# Run with smutans R package version 0.1.5 installed with R 2.14.2
# Remove eval=FALSE when the R package is installed
library(smutans)
# library(lattice)
# library(rtracklayer)
data(smutansData.tx)
data(smutansData.txGenes)
data(smutansData.txGrCombined)
data(smutansData.txAnnotation)
data(smutansData.txRNAz)
data(smutansData.txPileup)
data(smutans.feature.genes)
data(smutans.genes.criteria)
data(smutans.go.genes)
data(smutans.cat.desc)
data(smutans.keg.genes)
data(smutans.keg.desc)
data(smutansGenes)
data(smutansGenes2)
@
<<load,echo=FALSE,results=hide>>=
load(file="mss-lin.RData")
@
\documentclass{article}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{color} 
\usepackage{setspace} 
\usepackage{rotating}
\usepackage{xr}
\usepackage{url}
\usepackage{longtable}
\usepackage{lscape}
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}
\usepackage[round,comma,numbers]{natbib}

\def\@cite#1#2{(#1\if@tempswa , #2\fi)}                                         
\def\@biblabel#1{}

\externaldocument{mss-lin}

% Text layout
\oddsidemargin 0in
\evensidemargin 0in
\topmargin -.5in
\textwidth 6.5in
\textheight 9in
\parskip 0.25in

\makeatletter
\SweaveOpts{keep.source=TRUE,eps=FALSE,include=FALSE,width=6,height=6}
\newcommand{\Rpackage}[1]{R package \textit{#1}}
\newcommand{\program}[1]{{\texttt{#1}}}
\newcommand{\fixme}[1]{{\textbf{Fixme:} \textit{\textcolor{red}{#1}}}}
\newcommand{\change}[1]{\textit{\textcolor{blue}{#1}}}
\newcommand{\comment}[1]{}
\renewcommand{\@biblabel}[1]{\quad#1.}
\renewcommand{\thefootnote}{\textit{\alph{footnote}}}
\makeatother

\pagestyle{myheadings}
\markright{RNA-seq analysis in S. mutans}

% Leave date blank
\date{}

\begin{document}

\setcounter{secnumdepth}{-1} 

\noindent \begin{LARGE}{\bf CcpA and Catabolite Repression Explored\\
by RNA-Seq in {\em Streptococcus mutans}}\end{LARGE}\\[2ex]
{\bf 
Lin Zeng, Sang Chul Choi, Charles D.\ Danko, \\
Adam Siepel, Michael J.\ Stanhope, and Robert A.\ Burne}%\\[0.5ex]

\listoftables

\listoffigures

\doublespacing

\clearpage{}\setcounter{figure}{0}
\setcounter{table}{0}

\makeatletter 
\renewcommand{\thefigure}{S\@arabic\c@figure} 
\renewcommand{\thetable}{S\@arabic\c@table} 
\renewcommand{\figurename}{Supplementary Figure}
\renewcommand{\tablename}{Supplementary Table}
\makeatother

\clearpage{}
\section*{SUPPLEMENTARY TABLES}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Table S1. small RNAs neighboring context
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{landscape}
<<tabNeighborSmallRNA,echo=FALSE,results=tex>>=
print( xtable.tabNeighborSmallRNA,
       tabular.environment='longtable',
       table.placement="!h",
       caption.placement="top",
       include.rownames=FALSE,
       floating=FALSE,
       sanitize.text.function = function(x){x}) 
@
\end{landscape}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Table S2. GO UA159/TW1 grown in glucose
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf Gene Ontology (GO) enrichments for differentially expressed
genes in the comparison UA159 and TW1 grown in glucose.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq2 ( glucoseOnly, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc, 
                   goqval=1e-3)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$ We used a multiple-testing adjusted $p$-value of $10^{-3}$ to
determine differentially expressed genes.  For each category having at least ten
genes a variation of Fisher's exact test was performed, and another
multiple-testing adjusted $p$-value, or $q$-value was obtained. The listed
categories were chosen at the cutoff value of $10^{-3}$.
\\ 
\end{flushleft}
\label{tabGoseqGlucoseOnly}
\end{table}
%-----------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Table S3. KEGG UA159/TW1 grown in glucose
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf KEGG enrichments for differentially expressed
genes in the comparison UA159 and TW1 grown in glucose.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq2 ( glucoseOnly, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.keg.genes[c(4,3)],
                   cat.desc=smutans.keg.desc,
                   goqval=1e-3)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$ We used a multiple-testing adjusted $p$-value of $10^{-3}$ to
determine differentially expressed genes.  For each category having at least ten
genes a variation of Fisher's exact test was performed, and another
multiple-testing adjusted $p$-value, or $q$-value was obtained. The listed
categories were chosen at the cutoff value of $10^{-3}$.
\end{flushleft}
\label{tabKeggGlucoseOnly}
\end{table}
%-----------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Table S4. GO Glucose/Galactose with UA159
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf Gene Ontology (GO) enrichments for differentially expressed
genes in the comparison of glucose-grown and galactose-grown UA159 wild-type
cells.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq2 ( ua159only, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc, 
                   goqval=1e-3)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$ We used a multiple-testing adjusted $p$-value of $10^{-3}$ to
determine differentially expressed genes.  For each category having at least ten
genes a variation of Fisher's exact test was performed, and another
multiple-testing adjusted $p$-value, or $q$-value was obtained. The listed
categories were chosen at the cutoff value of $10^{-3}$.
\end{flushleft}
\label{tabGoseqUA159Only}
\end{table}
%-----------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Table S5. KEGG Glucose/Galactose with UA159
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf KEGG enrichments for differentially expressed
genes in the comparison of glucose-grown and galactose-grown UA159 wild-type
cells.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq2 ( ua159only, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.keg.genes[c(4,3)],
                   cat.desc=smutans.keg.desc,
                   goqval=1e-3)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$ We used a multiple-testing adjusted $p$-value of $10^{-3}$ to
determine differentially expressed genes.  For each category having at least ten
genes a variation of Fisher's exact test was performed, and another
multiple-testing adjusted $p$-value, or $q$-value was obtained. The listed
categories were chosen at the cutoff value of $10^{-3}$.
\end{flushleft}
\label{tabKeggUA159Only}
\end{table}
%-----------------------------------------------------------------------------


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Table S6. GO Glucose/Galactose with TW1
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf Gene Ontology (GO) enrichments for differentially expressed
genes in the comparison of glucose-grown and galactose-grown TW1 mutant-type
cells.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq2 ( tw1only, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc, 
                   goqval=1e-3)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$ We used a multiple-testing adjusted $p$-value of $10^{-3}$ to
determine differentially expressed genes.  For each category having at least ten
genes a variation of Fisher's exact test was performed, and another
multiple-testing adjusted $p$-value, or $q$-value was obtained. The listed
categories were chosen at the cutoff value of $10^{-3}$.
\end{flushleft}
\label{tabGoseqTW1Only}
\end{table}
%-----------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Table S7. KEGG Glucose/Galactose with TW1
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf KEGG enrichments for differentially expressed
genes in the comparison of glucose-grown and galactose-grown TW1 mutant-type
cells.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq2 ( tw1only, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.keg.genes[c(4,3)],
                   cat.desc=smutans.keg.desc,
                   goqval=1e-3)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$ We used a multiple-testing adjusted $p$-value of $10^{-3}$ to
determine differentially expressed genes.  For each category having at least ten
genes a variation of Fisher's exact test was performed, and another
multiple-testing adjusted $p$-value, or $q$-value was obtained. The listed
categories were chosen at the cutoff value of $10^{-3}$.
\end{flushleft}
\label{tabKeggTW1Only}
\end{table}
%-----------------------------------------------------------------------------

%%-----------------------------------------------------------------------------
%\begin{table}[!h]
%\caption{{\bf Gene Ontology (GO) enrichments for differentially expressed
%genes in the comparison UA159 and TW1 grown in galactose.}}
%\vspace{1ex}
%\noindent \begin{centering}
%\begin{tabular}{cl}
%\hline 
%$q^a$ & Description \\
%\hline 
%%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq2 ( galactoseOnly, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc, 
                   goqval=1e-3)
@
%%-----------------------------------------------------------------------------
%\hline 
%\end{tabular}
%\par\end{centering}
%\begin{flushleft}
%$^a$ We used a multiple-testing adjusted $p$-value of $10^{-3}$ to
%determine differentially expressed genes.  For each category having at least ten
%genes a variation of Fisher's exact test was performed, and another
%multiple-testing adjusted $p$-value, or $q$-value was obtained. The listed
%categories were chosen at the cutoff value of $10^{-3}$.
%\end{flushleft}
%\label{tab:goseqGalactoseOnly}
%\end{table}
%%-----------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Table S8. KEGG UA159/TW1 grown in galactose
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%-----------------------------------------------------------------------------
%\begin{table}[!h]
%\caption{{\bf KEGG enrichments for differentially expressed
%genes in the comparison UA159 and TW1 grown in galactose.}}
%\vspace{1ex}
%\noindent \begin{centering}
%\begin{tabular}{cl}
%\hline 
%$q^a$ & Description \\
%\hline 
%%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq2 ( galactoseOnly, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.keg.genes[c(4,3)],
                   cat.desc=smutans.keg.desc,
                   goqval=1e-3)
@
%%-----------------------------------------------------------------------------
%\hline 
%\end{tabular}
%\par\end{centering}
%\begin{flushleft}
%$^a$False discovery rate estimated by the
%Benjamini-Hochberg method.  All categories having at least ten genes and
%$q\leq 0.05$ are displayed.\\ 
%\end{flushleft}
%\label{tab:kegseqGalactoseOnly}
%\end{table}
%%-----------------------------------------------------------------------------



% END OF TABLE
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\clearpage{}
\section*{SUPPLEMENTARY FIGURES}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S1. Expressed/Not-expressed Gene length distribution
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<figLength,fig=TRUE,echo=FALSE>>=
l <- list(expressed=width(smutansData.txGenes[matchMatrix(matchGenesTx)[,1]]), 
          notexpressed=width(smutansData.txGenes[-matchMatrix(matchGenesTx)[,1]]))
bin <- c(as.integer(seq.int(0, 3000, length.out=15)), 8500)
multhist(l,breaks=bin,ylab="Frequency",xlab="Length of transcript in base pairs")
@
\begin{figure}[!h]
\centering
\includegraphics[width=.9\textwidth]{mss-lin-supp-figLength}
\caption{{\bf Length distribution for expressed and not-expressed genes.}
Darker gray
bars represent frequency of expressed genes, and lighter gray ones that of
not-expressed genes. The last bin sums from 3000 bp to 8500 bp.
}
\label{figLength}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S2. Transcript length distribution with/without annotated genes
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<figTranscriptLength,fig=TRUE,echo=FALSE>>=
l <- list(coding=width(grTxWGenes), 
          noncoding=width(grTxWOGenes))
bin <- c(as.integer(seq.int(0, 10000, length.out=15)), 30000)
multhist(l,breaks=bin,ylab="Frequency",xlab="Length of transcript in base pairs")
@
\begin{figure}
\centering
\includegraphics[width=.9\textwidth]{mss-lin-supp-figTranscriptLength}
\caption{{\bf Length distribution for expressed transcripts with annotated genes
and unannotated genes.}
Darker gray bars represent frequency of expressed transcripts with annotated
genes, and lighter gray ones that of without annotated genes that can be
potential non-coding regions. The last bin sums from 10000 bp to 30000 bp.
}
\label{figTranscriptLength}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S3. UTR length distribution of transcripts
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<figUTRLength,fig=TRUE,echo=FALSE>>=
l <- list(five=length5UTR, three=length3UTR)
bin <- c(as.integer(seq.int(0, 300, length.out=19)), 6500)
multhist(l,breaks=bin,ylab="Frequency",xlab="Length of UTRs in base pairs")
@
\begin{figure}
\centering
\includegraphics[width=.9\textwidth]{mss-lin-supp-figUTRLength}
\caption{{\bf Length distribution for 5' and 3' UTR.}
Darker gray bars represent frequency of 5' UTR lengths, and lighter gray 3' UTR
lengths. The last bin sums from 300 bp to 6500 bp.
}
\label{figUTRLength}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S4. RNA secondary structure
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\begin{figure}
\centering
\includegraphics[width=1.0\textwidth]{figures/PredSmallRNA-204}
\caption{{\bf MFE structures drawing encoding positional entropy for
differentially expressed regions.}
}
\label{figRNA204}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S5. Pie chart UA159/TW1 grown in Glucose
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\clearpage{}
<<figPieGlucoseOnly,fig=TRUE,echo=FALSE,width=6,height=8>>=
par(mai=c(0,0,0,0),fig=c(0,1,0,1),mfrow=c(2,1))
y <- tableGlucoseOnly[,"Functional Class"]
z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
z <- table(as.factor(sort(unlist(z))))

slices <- as.vector(z)
lbls <- names(z)
pct <- round(slices/sum(slices)*100)
c1 <- brewer.pal(n=9,name="Set1")
pie(z,labels=pct,main="",col=c1,clockwise=TRUE)
plot(0,0,axes=FALSE,frame.plot=FALSE,xlab="",ylab="",cex=0)
legend(-0.8,1,names(z),bty='n',cex=0.7,fill=c1)
@
\begin{figure}
\centering
\includegraphics[width=0.9\textwidth]{mss-lin-supp-figPieGlucoseOnly}
\caption{{\bf Pie chart distribution of functional classes of differntially expressed genes
in comparing UA159 wild-type and TW1 mutant-type grown in glucose.}
Labels for functional categories are colored with nine different colors. The
first label with red color, 
``Biosynthesis of cofactors, prosthetic groups, and carriers; Thiamine'', starts
at the top, and next labels read to the clockwise. The numbers circling the pie
chart is in per cent.
}
\label{figPieGlucoseOnly}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S6. Pie chart Glucose/Galactose in UA159
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\clearpage{}
<<figPieUA159Only,fig=TRUE,echo=FALSE,width=6,height=10>>=
par(mai=c(0,0,0,0),fig=c(0,1,0,1),mfrow=c(2,1))
trim <- function (x) gsub("^\\s+|\\s+$", "", x)
y <- tableUA159Only[,"Functional Class"]
z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
z <- table(as.factor(sort(unlist(z))))

slices <- as.vector(z)
lbls <- names(z)
pct <- round(slices/sum(slices)*100)
c1 <- brewer.pal(n=9,name="Set1")
pie(z,labels=pct,main="",col=c1,clockwise=TRUE)
plot(0,0,axes=FALSE,frame.plot=FALSE,xlab="",ylab="",cex=0)

legend(-0.8,1,names(z),bty='n',cex=0.7,fill=c1)
@
\begin{figure}
\centering
\includegraphics[width=0.9\textwidth]{mss-lin-supp-figPieUA159Only}
\caption{{\bf Pie chart distribution of functional classes of differntially expressed genes
in comparing UA159 wild-type grown in glucose and galactose.}
Labels for functional categories are colored with nine different colors. The
first label with red color, 
``Amino acid biosynthesis; Glutamate family'', starts
at the top, and next labels read to the clockwise. The numbers circling the pie
chart is in per cent.
}
\label{figPieUA159Only}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S7. Pie chart Glucose/Galactose in TW1
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\clearpage{}
<<figPieTW1Only,fig=TRUE,echo=FALSE,width=6,height=8>>=
par(mai=c(0,0,0,0),fig=c(0,1,0,1),mfrow=c(2,1))
trim <- function (x) gsub("^\\s+|\\s+$", "", x)
y <- tableTW1Only[,"Functional Class"]
z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
z <- table(as.factor(sort(unlist(z))))

slices <- as.vector(z)
lbls <- names(z)
pct <- round(slices/sum(slices)*100)
c1 <- brewer.pal(n=9,name="Set1")
pie(z,labels=pct,main="",col=c1,clockwise=TRUE)
plot(0,0,axes=FALSE,frame.plot=FALSE,xlab="",ylab="",cex=0)
legend(-0.8,1,names(z),bty='n',cex=0.7,fill=c1)
@
\begin{figure}
\centering
\includegraphics[width=0.9\textwidth]{mss-lin-supp-figPieTW1Only}
\caption{{\bf Pie chart distribution of functional classes of differntially expressed genes
in comparing TW1 mutant-type grown in glucose and galactose.}
Labels for functional categories are colored with nine different colors. The
first label with red color, 
``Amino acid biosynthesis; Histidine family'', starts
at the top, and next labels read to the clockwise. The numbers circling the pie
chart is in per cent.
}
\label{figPieTW1Only}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S8. Dot chart Glucose/Galactose in TW1
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<figDotplotTW1Only,fig=TRUE,echo=FALSE,width=12,height=6>>=
#par(mar=c(4.1, 3.1, 0.1, 0.1))
par(mai=c(0, 0, 0, 0))
y <- tableTW1Only[,"Functional Class"]
# z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
# z <- table(as.factor(sort(unlist(z))))
z <- sapply(y, function(x) {x1=strsplit(as.character(x),"\\|\\|"); lapply(x1, trim)})
xm <- tableTW1Only[rep(seq(length(z)),sapply(z,length)),]
xm[,"Functional Class"] <- unlist(z)
# Use xm to plot a dotplot.
dotplot(xm[,"Functional Class"]~xm[,2],
        main="",xlab="Log2 fold change",col="black",pch=88)
@
\begin{sidewaysfigure}
\centering
\includegraphics[width=0.9\textwidth]{mss-lin-supp-figDotplotTW1Only}
\caption{{\bf Dot chart distribution of functional classes of differntially expressed genes
in comparing TW1 wild-type grown in glucose and galactose.}
}
\label{figDotplotTW1Only}
\end{sidewaysfigure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S9. Pie chart UA159/TW1 grown in Galactose
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\clearpage{}
<<figPieGalactoseOnly,fig=TRUE,echo=FALSE,width=6,height=8>>=
par(mai=c(0,0,0,0),fig=c(0,1,0,1),mfrow=c(2,1))
y <- tableGalactoseOnly[,"Functional Class"]
z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
z <- table(as.factor(sort(unlist(z))))

slices <- as.vector(z)
lbls <- names(z)
pct <- round(slices/sum(slices)*100)
c1 <- brewer.pal(n=9,name="Set1")
pie(z,labels=pct,main="",col=c1,clockwise=TRUE)
plot(0,0,axes=FALSE,frame.plot=FALSE,xlab="",ylab="",cex=0)
legend(-0.8,1,names(z),bty='n',cex=0.7,fill=c1)
@
\begin{figure}
\centering
\includegraphics[width=0.9\textwidth]{mss-lin-supp-figPieGalactoseOnly}
\caption{{\bf Pie chart distribution of functional classes of differntially expressed genes
in comparing UA159 wild-type and TW1 mutant-type grown in galactose.}
Labels for functional categories are colored with nine different colors. The
first label with red color, 
``Amino acid biosynthesis; Glutamate family'', starts
at the top, and next labels read to the clockwise. The numbers circling the pie
chart is in per cent.
}
\label{figPieGalactoseOnly}
\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Figure S10. Dot chart UA159/TW1 grown in Galactose
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
<<figDotplotGalactoseOnly,fig=TRUE,echo=FALSE,width=12,height=6>>=
par(mai=c(0, 0, 0, 0))
y <- tableGalactoseOnly[,"Functional Class"]
# z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
# z <- table(as.factor(sort(unlist(z))))
z <- sapply(y, function(x) {x1=strsplit(as.character(x),"\\|\\|"); lapply(x1, trim)})
xm <- tableGalactoseOnly[rep(seq(length(z)),sapply(z,length)),]
xm[,"Functional Class"] <- unlist(z)
# Use xm to plot a dotplot.
dotplot(xm[,"Functional Class"]~xm[,2],
        main="",xlab="Log2 fold change",col="black",pch=88)
@
\begin{sidewaysfigure}
\centering
\includegraphics[width=0.9\textwidth]{mss-lin-supp-figDotplotGalactoseOnly}
\caption{{\bf Dot chart distribution of functional classes of differntially expressed genes
in comparing UA159 wild-type and TW1 mutant-type grown in galactose.}
}
\label{figDotplotGalactoseOnly}
\end{sidewaysfigure}

\end{document}

%\clearpage{}
%\begin{figure}
%\centering
%\includegraphics[width=\textwidth]{figures/ucsc}
%\caption{{\bf UCSC genome browser tracks of RNA-seq data.}
%Tracks are shown from the top: known genes, pileup value tracks for UA159 grown
%in glucose, those for TW1 in glucose, and the five segments predicted based on
%RNA-seq data and known genes. Known genes with forward strand are colored red,
%and those of reverse strand blue. The most top of the five predicted segments
%are also colored in the same way. 
%}
%\label{figUCSC}
%\end{figure}

