<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="application/xhtml+xml; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.6.7" />
<title>User Manual of RNAseq Analysis</title>
<style type="text/css">
/* Shared CSS for AsciiDoc xhtml11 and html5 backends */

/* Default font. */
body {
  font-family: Georgia,serif;
}

/* Title font. */
h1, h2, h3, h4, h5, h6,
div.title, caption.title,
thead, p.table.header,
#toctitle,
#author, #revnumber, #revdate, #revremark,
#footer {
  font-family: Arial,Helvetica,sans-serif;
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}
h5 {
  font-size: 1.0em;
}

div.sectionbody {
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}
ul > li     { color: #aaa; }
ul > li > * { color: black; }

pre {
  padding: 0;
  margin: 0;
}

#author {
  color: #527bbd;
  font-weight: bold;
  font-size: 1.1em;
}
#email {
}
#revnumber, #revdate, #revremark {
}

#footer {
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

#preamble {
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.0em;
  margin-bottom: 2.0em;
  margin-right: 10%;
  color: #606060;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid #dddddd;
  border-left: 4px solid #f0f0f0;
  padding: 0.5em;
}

div.listingblock > div.content {
  border: 1px solid #dddddd;
  border-left: 5px solid #f0f0f0;
  background: #f8f8f8;
  padding: 0.5em;
}

div.quoteblock, div.verseblock {
  padding-left: 1.0em;
  margin-left: 1.0em;
  margin-right: 10%;
  border-left: 5px solid #f0f0f0;
  color: #888;
}

div.quoteblock > div.attribution {
  padding-top: 0.5em;
  text-align: right;
}

div.verseblock > pre.content {
  font-family: inherit;
  font-size: inherit;
}
div.verseblock > div.attribution {
  padding-top: 0.75em;
  text-align: left;
}
/* DEPRECATED: Pre version 8.2.7 verse style literal block. */
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 3px solid #dddddd;
}

div.exampleblock > div.content {
  border-left: 3px solid #dddddd;
  padding-left: 0.5em;
}

div.imageblock div.content { padding-left: 0; }
span.image img { border-style: none; }
a.image:visited { color: white; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
  color: navy;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
ol.arabic {
  list-style-type: decimal;
}
ol.loweralpha {
  list-style-type: lower-alpha;
}
ol.upperalpha {
  list-style-type: upper-alpha;
}
ol.lowerroman {
  list-style-type: lower-roman;
}
ol.upperroman {
  list-style-type: upper-roman;
}

div.compact ul, div.compact ol,
div.compact p, div.compact p,
div.compact div, div.compact div {
  margin-top: 0.1em;
  margin-bottom: 0.1em;
}

tfoot {
  font-weight: bold;
}
td > div.verse {
  white-space: pre;
}

div.hdlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hdlist tr {
  padding-bottom: 15px;
}
dt.hdlist1.strong, td.hdlist1.strong {
  font-weight: bold;
}
td.hdlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
  color: navy;
}
td.hdlist2 {
  vertical-align: top;
}
div.hdlist.compact tr {
  margin: 0;
  padding-bottom: 0;
}

.comment {
  background: yellow;
}

.footnote, .footnoteref {
  font-size: 0.8em;
}

span.footnote, span.footnoteref {
  vertical-align: super;
}

#footnotes {
  margin: 20px 0 20px 0;
  padding: 7px 0 0 0;
}

#footnotes div.footnote {
  margin: 0 0 5px 0;
}

#footnotes hr {
  border: none;
  border-top: 1px solid silver;
  height: 1px;
  text-align: left;
  margin-left: 0;
  width: 20%;
  min-width: 100px;
}

div.colist td {
  padding-right: 0.5em;
  padding-bottom: 0.3em;
  vertical-align: top;
}
div.colist td img {
  margin-top: 0.3em;
}

@media print {
  #footer-badges { display: none; }
}

#toc {
  margin-bottom: 2.5em;
}

#toctitle {
  color: #527bbd;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel0, div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}

span.aqua { color: aqua; }
span.black { color: black; }
span.blue { color: blue; }
span.fuchsia { color: fuchsia; }
span.gray { color: gray; }
span.green { color: green; }
span.lime { color: lime; }
span.maroon { color: maroon; }
span.navy { color: navy; }
span.olive { color: olive; }
span.purple { color: purple; }
span.red { color: red; }
span.silver { color: silver; }
span.teal { color: teal; }
span.white { color: white; }
span.yellow { color: yellow; }

span.aqua-background { background: aqua; }
span.black-background { background: black; }
span.blue-background { background: blue; }
span.fuchsia-background { background: fuchsia; }
span.gray-background { background: gray; }
span.green-background { background: green; }
span.lime-background { background: lime; }
span.maroon-background { background: maroon; }
span.navy-background { background: navy; }
span.olive-background { background: olive; }
span.purple-background { background: purple; }
span.red-background { background: red; }
span.silver-background { background: silver; }
span.teal-background { background: teal; }
span.white-background { background: white; }
span.yellow-background { background: yellow; }

span.big { font-size: 2em; }
span.small { font-size: 0.6em; }

span.underline { text-decoration: underline; }
span.overline { text-decoration: overline; }
span.line-through { text-decoration: line-through; }

div.unbreakable { page-break-inside: avoid; }


/*
 * xhtml11 specific
 *
 * */

tt {
  font-family: "Courier New", Courier, monospace;
  font-size: inherit;
  color: navy;
}

div.tableblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.tableblock > table {
  border: 3px solid #527bbd;
}
thead, p.table.header {
  font-weight: bold;
  color: #527bbd;
}
p.table {
  margin-top: 0;
}
/* Because the table frame attribute is overriden by CSS in most browsers. */
div.tableblock > table[frame="void"] {
  border-style: none;
}
div.tableblock > table[frame="hsides"] {
  border-left-style: none;
  border-right-style: none;
}
div.tableblock > table[frame="vsides"] {
  border-top-style: none;
  border-bottom-style: none;
}


/*
 * html5 specific
 *
 * */

.monospaced {
  font-family: "Courier New", Courier, monospace;
  font-size: inherit;
  color: navy;
}

table.tableblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
thead, p.tableblock.header {
  font-weight: bold;
  color: #527bbd;
}
p.tableblock {
  margin-top: 0;
}
table.tableblock {
  border-width: 3px;
  border-spacing: 0px;
  border-style: solid;
  border-color: #527bbd;
  border-collapse: collapse;
}
th.tableblock, td.tableblock {
  border-width: 1px;
  padding: 4px;
  border-style: solid;
  border-color: #527bbd;
}

table.tableblock.frame-topbot {
  border-left-style: hidden;
  border-right-style: hidden;
}
table.tableblock.frame-sides {
  border-top-style: hidden;
  border-bottom-style: hidden;
}
table.tableblock.frame-none {
  border-style: hidden;
}

th.tableblock.halign-left, td.tableblock.halign-left {
  text-align: left;
}
th.tableblock.halign-center, td.tableblock.halign-center {
  text-align: center;
}
th.tableblock.halign-right, td.tableblock.halign-right {
  text-align: right;
}

th.tableblock.valign-top, td.tableblock.valign-top {
  vertical-align: top;
}
th.tableblock.valign-middle, td.tableblock.valign-middle {
  vertical-align: middle;
}
th.tableblock.valign-bottom, td.tableblock.valign-bottom {
  vertical-align: bottom;
}


/*
 * manpage specific
 *
 * */

body.manpage h1 {
  padding-top: 0.5em;
  padding-bottom: 0.5em;
  border-top: 2px solid silver;
  border-bottom: 2px solid silver;
}
body.manpage h2 {
  border-style: none;
}
body.manpage div.sectionbody {
  margin-left: 3em;
}

@media print {
  body.manpage div#toc { display: none; }
}
</style>
<script type="text/javascript">
/*<![CDATA[*/
var asciidoc = {  // Namespace.

/////////////////////////////////////////////////////////////////////
// Table Of Contents generator
/////////////////////////////////////////////////////////////////////

/* Author: Mihai Bazon, September 2002
 * http://students.infoiasi.ro/~mishoo
 *
 * Table Of Content generator
 * Version: 0.4
 *
 * Feel free to use this script under the terms of the GNU General Public
 * License, as long as you do not remove or alter this notice.
 */

 /* modified by Troy D. Hanson, September 2006. License: GPL */
 /* modified by Stuart Rackham, 2006, 2009. License: GPL */

// toclevels = 1..4.
toc: function (toclevels) {

  function getText(el) {
    var text = "";
    for (var i = el.firstChild; i != null; i = i.nextSibling) {
      if (i.nodeType == 3 /* Node.TEXT_NODE */) // IE doesn't speak constants.
        text += i.data;
      else if (i.firstChild != null)
        text += getText(i);
    }
    return text;
  }

  function TocEntry(el, text, toclevel) {
    this.element = el;
    this.text = text;
    this.toclevel = toclevel;
  }

  function tocEntries(el, toclevels) {
    var result = new Array;
    var re = new RegExp('[hH]([1-'+(toclevels+1)+'])');
    // Function that scans the DOM tree for header elements (the DOM2
    // nodeIterator API would be a better technique but not supported by all
    // browsers).
    var iterate = function (el) {
      for (var i = el.firstChild; i != null; i = i.nextSibling) {
        if (i.nodeType == 1 /* Node.ELEMENT_NODE */) {
          var mo = re.exec(i.tagName);
          if (mo && (i.getAttribute("class") || i.getAttribute("className")) != "float") {
            result[result.length] = new TocEntry(i, getText(i), mo[1]-1);
          }
          iterate(i);
        }
      }
    }
    iterate(el);
    return result;
  }

  var toc = document.getElementById("toc");
  if (!toc) {
    return;
  }

  // Delete existing TOC entries in case we're reloading the TOC.
  var tocEntriesToRemove = [];
  var i;
  for (i = 0; i < toc.childNodes.length; i++) {
    var entry = toc.childNodes[i];
    if (entry.nodeName.toLowerCase() == 'div'
     && entry.getAttribute("class")
     && entry.getAttribute("class").match(/^toclevel/))
      tocEntriesToRemove.push(entry);
  }
  for (i = 0; i < tocEntriesToRemove.length; i++) {
    toc.removeChild(tocEntriesToRemove[i]);
  }

  // Rebuild TOC entries.
  var entries = tocEntries(document.getElementById("content"), toclevels);
  for (var i = 0; i < entries.length; ++i) {
    var entry = entries[i];
    if (entry.element.id == "")
      entry.element.id = "_toc_" + i;
    var a = document.createElement("a");
    a.href = "#" + entry.element.id;
    a.appendChild(document.createTextNode(entry.text));
    var div = document.createElement("div");
    div.appendChild(a);
    div.className = "toclevel" + entry.toclevel;
    toc.appendChild(div);
  }
  if (entries.length == 0)
    toc.parentNode.removeChild(toc);
},


/////////////////////////////////////////////////////////////////////
// Footnotes generator
/////////////////////////////////////////////////////////////////////

/* Based on footnote generation code from:
 * http://www.brandspankingnew.net/archive/2005/07/format_footnote.html
 */

footnotes: function () {
  // Delete existing footnote entries in case we're reloading the footnodes.
  var i;
  var noteholder = document.getElementById("footnotes");
  if (!noteholder) {
    return;
  }
  var entriesToRemove = [];
  for (i = 0; i < noteholder.childNodes.length; i++) {
    var entry = noteholder.childNodes[i];
    if (entry.nodeName.toLowerCase() == 'div' && entry.getAttribute("class") == "footnote")
      entriesToRemove.push(entry);
  }
  for (i = 0; i < entriesToRemove.length; i++) {
    noteholder.removeChild(entriesToRemove[i]);
  }

  // Rebuild footnote entries.
  var cont = document.getElementById("content");
  var spans = cont.getElementsByTagName("span");
  var refs = {};
  var n = 0;
  for (i=0; i<spans.length; i++) {
    if (spans[i].className == "footnote") {
      n++;
      var note = spans[i].getAttribute("data-note");
      if (!note) {
        // Use [\s\S] in place of . so multi-line matches work.
        // Because JavaScript has no s (dotall) regex flag.
        note = spans[i].innerHTML.match(/\s*\[([\s\S]*)]\s*/)[1];
        spans[i].innerHTML =
          "[<a id='_footnoteref_" + n + "' href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
        spans[i].setAttribute("data-note", note);
      }
      noteholder.innerHTML +=
        "<div class='footnote' id='_footnote_" + n + "'>" +
        "<a href='#_footnoteref_" + n + "' title='Return to text'>" +
        n + "</a>. " + note + "</div>";
      var id =spans[i].getAttribute("id");
      if (id != null) refs["#"+id] = n;
    }
  }
  if (n == 0)
    noteholder.parentNode.removeChild(noteholder);
  else {
    // Process footnoterefs.
    for (i=0; i<spans.length; i++) {
      if (spans[i].className == "footnoteref") {
        var href = spans[i].getElementsByTagName("a")[0].getAttribute("href");
        href = href.match(/#.*/)[0];  // Because IE return full URL.
        n = refs[href];
        spans[i].innerHTML =
          "[<a href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
      }
    }
  }
},

install: function(toclevels) {
  var timerId;

  function reinstall() {
    asciidoc.footnotes();
    if (toclevels) {
      asciidoc.toc(toclevels);
    }
  }

  function reinstallAndRemoveTimer() {
    clearInterval(timerId);
    reinstall();
  }

  timerId = setInterval(reinstall, 500);
  if (document.addEventListener)
    document.addEventListener("DOMContentLoaded", reinstallAndRemoveTimer, false);
  else
    window.onload = reinstallAndRemoveTimer;
}

}
asciidoc.install(2);
/*]]>*/
</script>
</head>
<body class="article">
<div id="header">
<h1>User Manual of RNAseq Analysis</h1>
<span id="author">Sang Chul Choi</span><br />
<span id="email"><tt>&lt;<a href="mailto:goshng@yahoo.co.kr">goshng@yahoo.co.kr</a>&gt;</tt></span><br />
<span id="revnumber">version 0.1.7,</span>
<span id="revdate">August 2012</span>
<div id="toc">
  <div id="toctitle">Table of Contents</div>
  <noscript><p><b>JavaScript must be enabled in your browser to display the table of contents.</b></p></noscript>
</div>
</div>
<div id="content">
<div class="sect1">
<h2 id="_introduction">Introduction</h2>
<div class="sectionbody">
<div class="paragraph"><p>Bacteria can express genes in different intensity in two environments. The number of mRNAs is a direct measure of gene expression. Although a whole mRNA cannot be sequenced by next generation sequencing as of today as far as I know, parts of the whole mRNA can be sequenced. We shred the whole cDNA from an mRNA into smaller pieces, and sequence them to create a large number of reads of the pieces. The more expressed genes, the more reads are sequenced. Next generation sequencing technology can sequence parts of mRNAs after cDNA generation, creating a large number of reads.  We align sequenced reads from a bacterial RNA sample to a reference bacterial genome. Numbers of reads mapped on genes represent the rough expression intensity of those genes. We compare two numbers to state a statistical significance of, if any, the difference. We use four main tools to set up a pipeline that can take advantage of a UNIX cluster although you could use the pipeline on your machine that is equipped with bash shell, i.e., unix, linux, or Mac OS X. First, wetake a quality control step for filtering out raw reads data provided in FASTQ format. Second, we align reads to a bacterial genome, and count reads on annotations of the reference genome. Third, we determine whether genes are differentially expressed using the count data. Fourth, we assess functional associations of differentially expressed genes.</p></div>
<div class="paragraph"><p>Follow along to acheive the goal mentioned above. Warning! Backup your own raw data files before using commands of <strong>RNAseq Analysis</strong>. We do not guarantee that <strong>RNAseq Analysis</strong> works as you expect (see the license file, <tt>COPYING</tt> at the base directory of <strong>RNAseq Analysis</strong>, or execute menus <strong>warranty</strong> and <strong>copyright</strong>). It could delete important files of yours. Backup yours and use <strong>RNAseq Analysis</strong> at your own risk. If you have done so, you can keep reading this manual to do the analyses mentioned above. A list of RNA-seq data files are generated by sequencing bacterial transcriptome. We support RNA-seq data <strong>without</strong> strand information at the moment because I simply have only those data. If you have RNA-seq data with strand information, either you should change <strong>RNA-seq Analysis</strong> or contact me to help you (see section <a href="#help">Getting Help</a>).</p></div>
<div class="paragraph"><p>Do four steps before executing menus provided by <strong>RNAseq Analysis</strong>. First, create local and storage directories. See the section <a href="#Files">[Files]</a>. Second, create a <tt>conf</tt> directory in the local base directory, and place a proper configuration file at <tt>conf</tt> directory (We refer to a subdirectory of <strong>RNAseq Analysis</strong> without a prepending slash. The manual that you are reading, for example, is referred to as <tt>doc/Manual</tt> or <tt>doc/Manual.html</tt>. See the section <a href="#convention">[convention]</a> how we typeset words to refer to menus, files, variables in the document). See the section <a href="#Configuration">[Configuration]</a> to setup your configuration.  Third, install the software packages specified in the configuration file. See the section <a href="#Installation">[Installation]</a> for what software packages are needed. Fourth, prepare a species file at <tt><em>ROOTANALYSISDIR</em>/species</tt> directory (see also the section <a href="#convention">[convention]</a> to see what <em>ROOTANALYSISDIR</em> means), and the data files that are specified in the species file at proper places. Refer to the section <a href="#SpeciesFile">Species File</a> below.  Now, you are ready to start to analyze RNA-seq samples using <strong>RNAseq Analysis</strong>.</p></div>
<div class="listingblock">
<div class="title">Execute the main menu</div>
<div class="content">
<pre><tt>$ ./run</tt></pre>
</div></div>
<div class="paragraph"><p>First, call <strong>init-file-system</strong> to create directories for output files. Then, call <strong>choose-species</strong> menu to create directories for your project. See the subsection <a href="#txdb">[txdb]</a> for setting up gene annotations. You should configure a set of gene annotations because most of analyses use it. If you configured a gene annotation set by reading the subsection <a href="#txdb">[txdb]</a>, then select, for example, <strong>batch3</strong>, and follow the instruction in the menu <strong>batch3</strong> below. Menus might provide scripts in bash or R. Execute them at the base directory of <strong>RNAseq Analysis</strong>. For example, menu <strong>mcl</strong> allows you to execute <tt>job-mcl</tt> to find core genes. Because they may use Perl or R scripts that are available at the base directories <tt>pl</tt> or <tt>R</tt>, you must run them at the base directory of <strong>RNAseq Analysis</strong>, which can be different from <em>ROOTANALYSISDIR</em>.</p></div>
<div class="listingblock">
<div class="title">Execute scripts at the base directory</div>
<div class="content">
<pre><tt>$ cd RNAseq-Analysis
$ bash /path/to/job-mcl</tt></pre>
</div></div>
<div class="paragraph"><p><a id="convention"></a></p></div>
</div>
</div>
<div class="sect1">
<h2 id="_conventions_used_in_this_manual">Conventions used in this manual</h2>
<div class="sectionbody">
<div class="paragraph"><p>Menus are in <strong>bold</strong>, and file or software names are in <tt>typewriter</tt>. Shell variables, and variables used in configuration and species files are in <em>italic</em>. We could have two main root directories, which could refer to the same directory. We refer to the source directory of <strong>RNAseq Analysis</strong> as the root base directory of <strong>RNAseq Analysis</strong>. We often omit the directory when we refer to a subdirectory of the root base directory. This file that you are reading is called <tt>doc/Manual</tt> or <tt>doc/Manual.html</tt>. Another main directory is <em>ROOTANALYSISDIR</em>, which are intended to be used for storing data. You should set the variable in the configuration file, <tt>conf/README</tt>.</p></div>
<div class="paragraph"><p><a id="help"></a></p></div>
</div>
</div>
<div class="sect1">
<h2 id="_getting_help">Getting Help</h2>
<div class="sectionbody">
<div class="paragraph"><p>Please, post comments or help at
<a href="http://code.google.com/p/rnaseq-analysis">http://code.google.com/p/rnaseq-analysis</a>.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_kick_start">Kick Start</h2>
<div class="sectionbody">
<div class="olist arabic"><div class="title">Set up local base directoires (see <a href="#Files">[Files]</a>)</div><ol class="arabic">
<li>
<p>
Create storages and local base directories.
</p>
</li>
<li>
<p>
Use <tt>sh/make-soft-links.sh</tt> to create soft links at the local base directory
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Start a project</div><ol class="arabic">
<li>
<p>
Create a conf/README file using conf/README.template (see <a href="#Configuration">[Configuration]</a>)
</p>
</li>
<li>
<p>
Execute menu <strong>init-file-system</strong> (see <a href="#init-file-system">[init-file-system]</a>)
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Create a species file</div><ol class="arabic">
<li>
<p>
Execute menu <strong>choose-species</strong> (see <a href="#choose-species">[choose-species]</a>)
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Prepare a genomic sequence in FASTA format from NCBI</div><ol class="arabic">
<li>
<p>
Search <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria">ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria</a> for a complete bacterial genome in FASTA format
</p>
</li>
<li>
<p>
Change the sequence names of the file to a string of <em>chr</em> and an increasing number from 1, or chr1, chr2, and the like
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Prepare a genomic sequence from scratch</div><ol class="arabic">
<li>
<p>
Create a FASTA-format file with DNA sequences
</p>
</li>
<li>
<p>
Set the sequence names to a string of <em>chr</em> and an increasing number from 1
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Prepare a genome annotation file from NCBI complete genomes</div><ol class="arabic">
<li>
<p>
Download the gff file of the chosen species at <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria">ftp://ftp.ncbi.nlm.nih.gov/genomes/Bacteria</a>
</p>
</li>
<li>
<p>
Edit the gff file so that chromosome names are chr1, chr2, and the like.
</p>
</li>
<li>
<p>
Execute menu <strong>convert-gff2txdb</strong> (see <a href="#convert-gff2txdb">[convert-gff2txdb]</a>)
</p>
</li>
<li>
<p>
Specify the txdb file in your <tt>species</tt> file
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Prepare a genome annotation file from scratch</div><ol class="arabic">
<li>
<p>
Learn how to use <em>makeTranscriptDb</em> function of BioC R package <tt>GenomicFeatures</tt> available at <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html">GenomicFeatures</a>
</p>
</li>
<li>
<p>
Create a txdb file using the BioC R package
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Test the pipeline of aligning and counting reads</div><ol class="arabic">
<li>
<p>
Prepare a species file for testing purpose (a genome in FASTA format and its annotation in txdb format files are necessary)
</p>
</li>
<li>
<p>
Execute menu <strong>batch3</strong> (see <a href="#batch3">[batch3]</a>)
</p>
</li>
<li>
<p>
Run <tt>job-simulate</tt> at <em>CACWORKDIR</em> to create test FASTQ files
</p>
</li>
<li>
<p>
Run <tt>run-qcalignde.sh</tt> at <em>CACWORKDIR</em>
</p>
</li>
<li>
<p>
Run <tt>job-check</tt> at <em>CACWORKDIR</em>
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Align and count reads using FASTA-format DNA sequence and txdb files</div><ol class="arabic">
<li>
<p>
Prepare a species file for testing purpose (a genome in FASTA format and its annotation in txdb format files are necessary)
</p>
</li>
<li>
<p>
Place FASTQ files at <em>ROOTANALYSISDIR</em> directory
</p>
</li>
<li>
<p>
Execute menu <strong>batch3</strong>
</p>
</li>
<li>
<p>
Run <tt>output/your species name/push-data.sh</tt>
</p>
</li>
<li>
<p>
Go to <em>CACWORKDIR</em>, and run <tt>run-qcalignde.sh</tt>
</p>
</li>
<li>
<p>
Run <tt>job-de-sum</tt> at <em>CACWORKDIR</em>
</p>
</li>
<li>
<p>
Run <tt>job-stat-plot</tt> at <em>CACWORKDIR</em>
</p>
</li>
<li>
<p>
Run <tt>output/your species name/get-data.sh</tt>
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Compare two gene expression</div><ol class="arabic">
<li>
<p>
Align and count reads using menu <strong>batch3</strong>
</p>
</li>
<li>
<p>
Add <em>DESEQ</em> variable in the species file
</p>
</li>
<li>
<p>
Execute menu <strong>deseq</strong>
</p>
</li>
<li>
<p>
Run <tt>job-deseq</tt> at base directory of <strong>RNAseq Analysis</strong>
</p>
</li>
<li>
<p>
You might want to use a companion R package, <tt>smutans</tt>
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Find genes that unique to one of two genomes</div><ol class="arabic">
<li>
<p>
Extract protein sequences from two genomes in genbank file format
</p>
</li>
<li>
<p>
Add <em>REFCDSFASTA1</em> and <em>REFCDSFASTA2</em> to the species file
</p>
</li>
<li>
<p>
Execute menu <strong>deseq</strong>
</p>
</li>
<li>
<p>
Run <tt>job-blast</tt> at base directory of <strong>RNAseq Analysis</strong>
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Find functional associations of differentially expressed genes</div><ol class="arabic">
<li>
<p>
Use smutans R package.
</p>
</li>
</ol></div>
<div class="olist arabic"><div class="title">Upload BAM files</div><ol class="arabic">
<li>
<p>
Add KEEPBAM:YES
</p>
</li>
<li>
<p>
Execute menu <strong>batch3</strong>
</p>
</li>
</ol></div>
</div>
</div>
<div class="sect1">
<h2 id="_menus">Menus</h2>
<div class="sectionbody">
<div class="paragraph"><p><strong>RNAseq Analysis</strong> menus provide users with access to commands for analyzing bacterial RNA-seq data. We assume that we would create output directories at <em>ROOTANALYSISDIR</em> directory, which can be different from the current source root directory of <strong>RNAseq Analysis</strong>.</p></div>
<div class="paragraph"><p><a id="init-file-system"></a></p></div>
<div class="sect2">
<h3 id="_init_file_system">init-file-system</h3>
<div class="paragraph"><p>This creates <tt>ROOTANALYSISDIR/output</tt> directory. As mentioned above, we write <tt>output</tt> without <em>ROOTANALYSISDIR</em>. Because the output directory can be large, a separate hard drive can be mounted and used by symbolic links. Users might want to create <tt>data</tt>, and <tt>downloads</tt> directories. The <tt>data</tt> can be used to store raw data files including FASTQ, genome, and genome annotation files. The <tt>downloads</tt> directory can be used to store other software packages for installation.</p></div>
<div class="paragraph"><p><a id="choose-species"></a></p></div>
</div>
<div class="sect2">
<h3 id="_choose_species">choose-species</h3>
<div class="paragraph"><p>Create a file in <tt>ROOTANALYSISDIR/species</tt> directory before calling this menu. We write <tt>species</tt> without <em>ROOTANALYSISDIR</em>. Refer to the section <a href="#SpeciesFile">Species File</a>. It creates output directories at the species named subdirectory in <tt>output</tt> directory. Choose a species and a repetition number as you are asked. The repetition number can be set in a species file. Note that we use the following output directory structure. <em>DATADIR</em> is prefixed by <em>ROOTANALYSISDIR</em>. We use the same output structures in the compute node by prefixing <em>C</em>, and in the remote machine by prefixing <em>R</em>. For example, <em>CDATADIR</em> is similar to <em>DATADIR</em> except <em>OUTPUTDIR</em> being just <tt>output</tt>. <em>RDATADIR</em> is different from <em>DATADIR</em> in that <em>OUTPUTDIR</em> is replaced by <em>ROUTPUTDIR</em>, which is set in <tt>conf/README</tt>.</p></div>
<div class="paragraph"><p><a id="ShellVariable"></a></p></div>
<div class="listingblock">
<div class="title">Output directories</div>
<div class="content">
<pre><tt>SPECIES=[Your species name without spaces]
ROOTANALYSISDIR=[This is the directory where you execute menus]
OUTPUTDIR=$ROOTANALYSISDIR/output
BASEDIR=$OUTPUTDIR/$SPECIES
BASERUNANALYSIS=$BASEDIR/run-analysis
NUMBERDIR=$BASEDIR/$REPETITION
DATADIR=$NUMBERDIR/data
BWADIR=$NUMBERDIR/bwa
RUNANALYSIS=$NUMBERDIR/run-analysis</tt></pre>
</div></div>
<div class="paragraph"><p>This menu starts to create a directory named after the species file name in the main <tt>output</tt> directory: <em>BASEDIR</em>. Each species named directory can contain numbered directories. The number corresponds to <em>REPETITION</em> or repeated analyses under the species named project. We would start with a number 1 for the first repetition. You can specify <em>REPETITION</em> in the species file. Otherwise, it is set to 1 by default. In a number directory, you can have directories such as <tt>data</tt>, <tt>bwa</tt>, <tt>run-analysis</tt>, etc.  The <tt>bwa</tt> directory would contains results from analyses using <strong>BWA</strong>. In the bash scripts you can access these directories using shell variables. See <tt>sh/global-variable.sh</tt> for details.</p></div>
<div class="paragraph"><p><a id="convert-gff2txdb"></a></p></div>
</div>
<div class="sect2">
<h3 id="_convert_gff2txdb">convert-gff2txdb</h3>
<div class="paragraph"><p>We need an annotation file for differential expression study. We create a txdb file using a gff or a BED file. Because we assume chromosome names to be chr1, chr2, and the like. edit the gff file so that the chromosome name is chr1 for a single chromosome bacterial genome. Bacterial genomes with multiple chromosomes should have chr1, chr2, etc. You might want to use your own txdb file. Use an R script in <tt>R</tt> directory to learn to create a txdb file from scratch. See, for example, <tt>R/gff2txdb.R</tt>.</p></div>
<div class="listingblock">
<div class="title">To create txdb from gff</div>
<div class="content">
<pre><tt>Edit NC_004350.fna and NC_004350.gff so that the chromosome name is chr1.
Rscript R/gff2txdb.R</tt></pre>
</div></div>
<div class="paragraph"><p><a id="batch3"></a></p></div>
</div>
<div class="sect2">
<h3 id="_batch3">batch3</h3>
<div class="paragraph"><p>We use the <tt>PRINSEQ</tt> perl script for quality control. The data quality control steps includes removing low quality reads, trimming low quality 3' end parts, timming parts of adapters, and the like. We use BioConductor R packages to count reads on a reference annotation of a genome. Because the shell scripts in <strong>RNA-seq Analysis</strong> create directories and save output files in them, it is good to know what output directories are created. The structure of output directories in the local machine where you execute the main <tt>run</tt> script is similar to that in the remote machine such as a cluster of compute nodes. In the cluster, you will have a base directory where you would submit jobs. This is called a working directory at the cluster, which can be configured using <em>CACWORKDIR</em> species variable.  In compute nodes, you will also have base directories that are created in <tt>/tmp</tt> directory. Under the <tt>/tmp</tt> directory, you will have a temporary directory; e.g., <tt>1209759-3.scheduler.v4linux</tt>, which would be the base directory in a compute node. Refer to menu <strong>choose-species</strong> for the directory structure in a base direcotry.</p></div>
<div class="paragraph"><p>Menu <strong>batch3</strong> creates a number of scripts at <em>BASEDIR</em> (See <a href="#ShellVariable">Output directories</a>).  One of the shell scripts, <tt>push-data.sh</tt>, is created at <em>BASEDIR</em> by function <em>batch3-push-data</em> of shell script <tt>sh/batch3.sh</tt>. The shell script transfer input files including RNA-seq data files to the remote machine. You might want to check the script to see what files are copied to the cluster before executing it because copying large files can take minutes. I would copy large RNA-seq data files only once. You could manually copy some files, or change file paths in the remote computer because often the hard drive space at a remote computer can be limited. The script also creates data, and bwa directories in the cluster. It copies the reference genome, and the annotation file in txdb format.</p></div>
<div class="listingblock">
<div class="title">To copy data files to the cluster</div>
<div class="content">
<pre><tt>$ bash output/ua159/push-data.sh</tt></pre>
</div></div>
<div class="paragraph"><p>Wait! You might want to make sure whether the setup is complete before sending files and submiting jobs to process those files. We take two steps back before we submit jobs to get the result. First, you go through what variables need to be set in the species file (see section <a href="#SpeciesFile">Species File</a>). Second, you can use a test data set to check whether your pipeline is working as expected. To test the pipeline of <strong>RNAseq Analysis</strong>, set <em>FASTQFILES</em>, and others in the species file</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>FASTQFILES:101 102
QUALITYSCORE101:sanger
QUALITYSCORE102:sanger
ADAPTER101:GATCGGAAGAGCACACGTCTGAACTCCAGTCACCTTGTAATCTCGTATGCCGTCTTCTGCTTG
ADAPTER102:GATCGGAAGAGCACACGTCTGAACTCCAGTCACCTTGTAATCTCGTATGCCGTCTTCTGCTTG
QCALIGNDEWALLTIME:1</tt></pre>
</div></div>
<div class="paragraph"><p>Warning, again! Backup your important raw data files before using <strong>RNAseq Analysis</strong>. If you specify <em>FASTQFILES</em> to FASTQ file IDs of your real data, and execute <tt>job-simulate</tt>, then your FASTQ files will be deleted for good. Use <strong>RNAseq Analysis</strong> at your own risk.</p></div>
<div class="paragraph"><p>A simulation is not sufficient to check whether the pipeline of counting short reads mapped on a reference genome is working properly. It is a necessary condition. It also provides a way of debugging the scripts of <strong>RNAseq Analysis</strong> when it does not work as expected. The simulation and checking the pipeline takes minutes while your real data anlaysis might take hours. You might want to use a development queue of your UNIX cluster so that you do not wait for your jobs to complete when queues are full. I set both of the wall time to 1 hour to use a development queue by editting <tt>conf/README</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>QUEUENAME:v4dev</tt></pre>
</div></div>
<div class="paragraph"><p>Use the following commands at <em>CACWORKDIR</em> to simulate, and check the pipeline of aligning and counting reads on your reference genome.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>cac$ bash job-simulate
cac$ bash run-qcalignde.sh
cac$ bash job-check</tt></pre>
</div></div>
<div class="paragraph"><p>The last command, <tt>job-check</tt>, check whether two numbers are similar. The ratio of correction means the proportion of genes with equal numbers of short reads mapped on a reference genome from the simulation. The number should be near 1. Often, they were larger than 0.9 but not 0.95. Once you are finished with simulation checks, change the type of queue in the configuration file, and wall time <em>QCALIGNDEWALLTIME</em> in the species file. Go through the check list above. Then, submit the job at <em>CACWORKDIR</em>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>cac$ bash run-qcalignde.sh</tt></pre>
</div></div>
<div class="paragraph"><p>Once your jobs are finished, run the following command to create count files.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>cac$ bash job-de-sum</tt></pre>
</div></div>
<div class="paragraph"><p>You can also create base quality scores of short reads before and after (Not yet implemented) trimming 3' end of short reads.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>cac$ bash job-stat-plot</tt></pre>
</div></div>
<div class="paragraph"><p>Get result files from the cluster.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ bash output/ua159/get-data.sh</tt></pre>
</div></div>
<div class="paragraph"><p>You will have two kinds of files for determining differentially expressed genes: <tt>count</tt> and <tt>count.txt.index</tt>. You can use smutans R package to achieve this.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>output/ua159/1/bwa/count
output/ua159/1/run-analysis/count.txt.index</tt></pre>
</div></div>
<div class="paragraph"><p>We provide a few count files depending on which parts of gene annotations are used. <tt>count</tt> uses CDS genes, non-CDS genes (or ribosomal RNAs), and non-genic regions. <tt>count.tx</tt> uses genic regions including CDS and non-CDS genes. <tt>count.ng</tt> uses non-genic regions. <tt>count.cds</tt> uses only CDS genes, and <tt>count.nocds</tt> uses only non-CDS genes. We use uniquely mapped reads for counting reads for creating files just mentioned above. The cutoff value of mapping quality is <em>MINMAPQ</em>. We also use reads with mapping quality less than <em>MINMAPQ</em>, and create count files with suffix of <tt>m</tt>. For example, <tt>count.m.cds</tt> is a count file that contains number of reads multiply mapped on CDS regions.  We have all of the mapped reads in BAM files. The species variable <em>MINMAPQ</em> is used to count multiply aligned reads.</p></div>
</div>
<div class="sect2">
<h3 id="_deseq">deseq</h3>
<div class="paragraph"><p>We use an R package, <tt>smutans</tt> (ver. 0.1.9), that comes along with <strong>RNAseq Analysis</strong> for pairwise comparisons of gene expression. The menu uses BioConductor R packages including <tt>DESeq</tt> to determine differentially expressed genes. It features comparing two sets of protein sequences to determine unique genes to either of the two genomes. First, and the most useful feature is to determine differentially expressed genes. Let <strong>RNAseq Analysis</strong> know what samples you want to compare using <em>DESEQ</em>. You have nine samples numbered 1, 6, 7, 8, 11, 12, 13, 19, and 20. Each sample is labeled using <em>FASTQLABEL</em> variable. Use the following three lines in your species file to compare UA159 (1, 19, and 20) and each of other samples.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>FASTQFILES:1 6 7 8 11 12 13 19 20
FASTQLABEL:UA159 1SM1 11VS1 15JP3 N29 NLML4 NMT4863 UA159 UA159
DESEQ:UA159/1SM1,UA159/11VS1,UA159/15JP3,UA159/N29,UA159/NLML4,UA159/NMT4863</tt></pre>
</div></div>
<div class="paragraph"><p>Because each repetition is based on a particular reference genome, you would have to execute <strong>deseq</strong> for each repetition. It may be desirable to execute all of the repetitions. This would require many changes to the scripts of <strong>RNAseq Analysis</strong>. The <strong>deseq</strong> menu should provide a bash script that you can run. Run the script at the base root directory of <strong>RNAseq Analysis</strong>. The prepending slash denotes that the path is a full path.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ bash /path/to/job-blast</tt></pre>
</div></div>
<div class="paragraph"><p>The next feature is to determine genes that are unique to one of two genomes. This may belong to other menus. Because finding unique or similar genes are related with DE study, we have the bash script in menu <strong>deseq</strong>.  We extract protein sequences from a genbank file by using <tt>pl/extractcds.pl</tt> (FIXME: we have two Perl scripts for extracting protein sequences. Use only one of them. The other is called <tt>bioperl-get-protein.pl</tt>). First, extract protein sequences from the genbank file of your bacterial genome.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>perl pl/extractcds.pl -f Genbank file.gbk &gt; file.faa
perl pl/bioperl-get-protein.pl protein -in file.gbk -out file.faa</tt></pre>
</div></div>
<div class="paragraph"><p>They should produce the same sequences</p></div>
<div class="paragraph"><p>Add the following two lines to use the <tt>faa</tt> files in addition to the gff file from the first genome. Use the gff file that is downloaded from NCBI bacterial genomes. The first chromosome is contains more refined annotations that the second genome lacks.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>REF1GFF:path/to/file.gff
REFCDSFASTA1:path/to/first.faa
REFCDSFASTA2:path/to/second.faa</tt></pre>
</div></div>
<div class="listingblock">
<div class="content">
<pre><tt>bash /path/to/job-blast</tt></pre>
</div></div>
<div class="paragraph"><p>This sript creates the following files.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>_RUNANALYSIS_/genes-chr1-unique.csv
_RUNANALYSIS_/genes-chr1-similar.csv
_RUNANALYSIS_/genes-chr2-unique.csv
_RUNANALYSIS_/genes-chr2-similar.csv</tt></pre>
</div></div>
<div class="paragraph"><p>File <tt><em>RUNANALYSIS</em>/unique-genes-chr1.csv</tt> contains genes that are unique to the first set of proteins. The first column is the gene name of the first genome, and the second column is the coverage value that quantifies the proportion of genes with hits in BLAST results in per cent. File <tt><em>RUNANALYSIS</em>/genes-chr1-similar.csv</tt> contains three columns: the gene name from the first genome, that from the second genome, and the coverage. File <tt><em>RUNANALYSIS</em>/genes-chr2-unique.csv</tt> is of the same format of <tt><em>RUNANALYSIS</em>/genes-chr1-unique.csv</tt> but with chr2 genome&#8217;s unique genes. File <tt><em>RUNANALYSIS</em>/genes-chr2-similar.csv</tt> contains columns that are similar to file <tt><em>RUNANALYSIS</em>/genes-chr1-similar.csv</tt> but with one more column of description of chr1&#8217;s genes, which is the second column of the file <tt><em>RUNANALYSIS</em>/genes-chr2-similar.csv</tt>.</p></div>
<div class="paragraph"><p>More detailed accounts for the study of differential expressed genes should be found at the BioConductor R package <tt>DESeq</tt>. The companion R package, <tt>smutans</tt>, contains R scripts that use <tt>DESeq</tt>, and others in BioConductor R packages. Look at files in <tt>package/smutans</tt> for seeing how to determine differentially expressed genes. Espeically, see the tutorial of the package in <tt>package/smutans/inst/doc/smutans.pdf</tt>.</p></div>
</div>
<div class="sect2">
<h3 id="_goseq">goseq</h3>
<div class="paragraph"><p>The function <tt>smutans.de2Goseq</tt> uses <em>Smutans</em> object to find functional associations of differentially expressed genes. We prepare three required files to execute the function. First, fetures of genes in the reference genome are a table of 6 columns: chromosome, start, end, gene name, 0, and strand. Second, a file that associates genes and functional terms contains 3 columns: gene name, functional term, and a real value. Third, a file that describe functional terms contains 3 columns: functional term, number of genes that belong to the term, and description of the function.</p></div>
<div class="paragraph"><p>I download files for functional association</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>http://www.geneontology.org/gene-associations/submission/gene_association.goa_uniprot.gz
http://www.geneontology.org/ontology/obo_format_1_2/gene_ontology.1_2.obo
ftp://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref90/uniref90.fasta.gz</tt></pre>
</div></div>
<div class="paragraph"><p>Copy them to <em>ROOTANALYSISDIR</em> and specify their relative locations in your species file</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>GOAUNIPROT:data/goseq/gene_association.goa_uniprot.gz
GOBO:data/goseq/gene_ontology.1_2.obo
UNIREFFASTA:data/goseq/uniref90.fasta.gz
REF1GFF:path/to/file.gff</tt></pre>
</div></div>
<div class="paragraph"><p>We use the gff file referred by <em>REF1GFF</em> to find the list of genes. We could not use txdb because the <tt>pl/geneonlogy.pl</tt> is a Perl script which might not be able to use txdb files. We need to associate locus tags with protein IDs so that we can associate locus tags with gene ontology terms. We did not have the association of locus tags and protein IDs in txdb files, yet. Even if we did, we still need to use txdb files in a Perl file. One way to sidestep the problem is to produce a file that is easy to parse using a txdb file, and let the Perl script, <tt>pl/geneontology.pl</tt> use the file. I think that this is a good direction because I wish to have gene annotations only in a txdb file if I could. So, making a txdb file is the first step of gene annotations.</p></div>
<div class="paragraph"><p>The three file names should have the extensions of the above.</p></div>
<div class="paragraph"><p>Use txdb to create <tt><em>RUNANALYSIS</em>/feature.gene</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>REFGENOMETXDB:path/to/NC_004350.txdb</tt></pre>
</div></div>
<div class="paragraph"><p>Run the following two commands.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ bash /path/to/job-feature
$ bash /path/to/job-goseq</tt></pre>
</div></div>
<div class="paragraph"><p>There are three parameters to tune the <strong>goseq</strong> analysis. First, we could change BLAST options to associate proteins with gene ontology terms. Second, we could change the q-value of determining differentially expressed genes. Third, we could change the q-value of calling gene ontology terms significant. We use 0.05 for the q-value of the third option (This is hard-coded in <tt>smutans</tt> R package. We usually change the q-value for the second. The first option may be infrequently used because of the long (1 day) execution time of BLAST.</p></div>
<div class="paragraph"><p>EDIT THE FOLLOWING.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>GOSEQFASTA:data/analysis/NC_004350.faa
BLASTEVALUE:1e-3</tt></pre>
</div></div>
<div class="paragraph"><p><em>GOSEQFASTA</em> is for protein sequences of the reference genome.
<em>BLASTEVALUE</em> is for minimum e-value of BLAST search.</p></div>
</div>
<div class="sect2">
<h3 id="_mcl">mcl</h3>
<div class="paragraph"><p>We use <tt>BLASTP</tt> and <tt>mcl</tt> to cluster genes with their protein sequences. The main input file is a FASTA file that contains protein sequences.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>MCLFASTA=path/to/mcl.faa</tt></pre>
</div></div>
<div class="paragraph"><p>Gzip the <tt>mcl.faa</tt> although you write <tt>mcl.faa</tt> without <tt>.gz</tt> extension. The FASTA file should contain sequences with title strings that are easy to parse. The output files from menu <strong>mcl</strong> are the output files from <tt>mclblastline</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>_RUNANALYSIS_/dump.out.all.blastp.tab.*</tt></pre>
</div></div>
<div class="paragraph"><p>The file contains rows of gene names, each of which is a group of genes. We can use <tt>pl/find-core-gene.pl</tt> to find core and non-core genes in a given list of genes in a BED file. The BED file can be created using menu <strong>goseq</strong> using <tt>job-feature</tt> command (see menu <strong>goseq</strong>). You can view the cluster using <tt>cytoscape</tt>. A simple .sif file can be created from mclblastline&#8217;s dump file, e.g., dump.out.all.blastp.tab.I12. You insert a string, e.g., typeA, after the first column if there are multiple columns. You could open the new .sif file using menu File/Import/Network (Multiple File Types)&#8230; in <tt>cytoscape</tt>.</p></div>
<div class="paragraph"><p>Run <tt>job-mcl</tt> at the root directory of <strong>RNAseq Analysis</strong>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ bash /path/to/job-mcl</tt></pre>
</div></div>
<div class="paragraph"><p><a id="Configuration"></a></p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_configuration">Configuration</h2>
<div class="sectionbody">
<div class="paragraph"><p>Place a file called README in the directory of <tt>conf</tt>. You can copy
<tt>conf/README.template</tt> to create <tt>conf/README</tt>. The file contains two kinds of
lines: one for comment, and the other for variable assignment. Comments should
start with # at the first column of a line.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt># Comment</tt></pre>
</div></div>
<div class="paragraph"><p>A variable assignment should start without # at the first column. Variables tend
to be capitalized, and are followed by a colon and a variable value. No spaces
are allowed. Give, for example, a project name as follows.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>PROJECTNAME:streptoccus</tt></pre>
</div></div>
<div class="paragraph"><p>All of the output files from <strong>RNAseq Analysis</strong> created in <tt>output</tt> directory.
You might want to have a separate hard drive for saving those files because output
files can grow larger than any files in the main source files.
Set the output directory directly in the
configuration. Set <em>ROOTANALYSISDIR</em> to a full file path.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>ROOTANALYSISDIR:/path/to/RNASeq-Analysis-Storage-Directory</tt></pre>
</div></div>
<div class="paragraph"><p>We allow two modes of using <strong>RNAseq Analysis</strong>: local, and remote (Not yet checked although we tried to run the local mode). When using local mode, we simply set the <em>CAC_USERNAME</em> and <em>CAC_LOGIN</em> to localhost. Currently, <em>RUNMODE</em> is purely for the purpose of annotation of the configuration file. Copy <tt>bin/nsub</tt> to a directory that is part of PATH, so that <tt>nsub</tt> can be executed in the local machine when the run mode is <em>local</em>.</p></div>
<div class="listingblock">
<div class="title">To execute jobs in your local machine</div>
<div class="content">
<pre><tt>RUNMODE:local</tt></pre>
</div></div>
<div class="listingblock">
<div class="title">To execute jobs in a cluster machine such as CAC: <a href="http://www.cac.cornell.edu">http://www.cac.cornell.edu</a>.</div>
<div class="content">
<pre><tt>RUNMODE:remote</tt></pre>
</div></div>
<div class="paragraph"><p>Set more variables if you execute jobs in a remote machine. We let the remote machine be CAC, which is one of the cornell linux clusters. We are not sure what other linux cluster options are available. If your cluster system is different from one in CAC, you have to modify parts of the cluster job submission in the shell scripts. You should set up ssh login without password because we need log in the remote machine to create directories and copy files back and forth. See <a href="http://www.linuxproblem.org/art_9.html">http://www.linuxproblem.org/art_9.html</a> or <a href="http://smbjorklund.no/ssh-login-without-password-using-os-x">http://smbjorklund.no/ssh-login-without-password-using-os-x</a>. You should set the number of CPUs available in a computing machine.</p></div>
<div class="listingblock">
<div class="title">To check the number of CPUs</div>
<div class="content">
<pre><tt>Use "grep processor /proc/cpuinfo|wc -l" in CentOS
Use "sysctl -a|grep cpu.core_count" in MacOSX"</tt></pre>
</div></div>
<div class="listingblock">
<div class="title">To set the number of CPUs for running jobs.</div>
<div class="content">
<pre><tt>NUMBERCPU:8</tt></pre>
</div></div>
<div class="paragraph"><p>You need to specify a combination of a user name and a domain name.</p></div>
<div class="listingblock">
<div class="title">To set user ID and login address</div>
<div class="content">
<pre><tt>CAC_USERNAME:yourUserID
CAC_LOGIN:compute.domain.name</tt></pre>
</div></div>
<div class="paragraph"><p>Set the <em>CAC_LOGIN</em> variable even if you execute jobs in your local machine.</p></div>
<div class="listingblock">
<div class="title">To set user ID and login address for local mode</div>
<div class="content">
<pre><tt>CAC_USERNAME:yourUserID
CAC_LOGIN:localhost</tt></pre>
</div></div>
<div class="paragraph"><p>We could have two unix machines: one for computing and another for genome
browser. Use <em>X11_USERNAME</em> and <em>X11_LOGIN</em> to set a machine for genome
browser. <strong>RNAseq Analysis</strong> does not feature functions of creating
UCSC genome browser tracks.</p></div>
<div class="listingblock">
<div class="title">To set user ID and login address</div>
<div class="content">
<pre><tt>X11_USERNAME:yourUserIDforGenomeBrowser
X11_LOGIN:genome.browser.domain.name</tt></pre>
</div></div>
<div class="paragraph"><p>We use the CAC cluster for computing, which may need access information. A batch
script for the cluster would have following header lines.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>#PBS -A ${BATCHACCESS}
#PBS -q ${QUEUENAME}
#PBS -M ${BATCHEMAIL}</tt></pre>
</div></div>
<div class="paragraph"><p>The three variables above are replaced by values from the following lines.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>BATCHACCESS:yourAccessID
QUEUENAME:yourQueueName
BATCHEMAIL:you@email.domain.name</tt></pre>
</div></div>
<div class="paragraph"><p>Sometime you may not want to get notified from the cluster when jobs are
finished. Then, set <em>EMAILON</em> to FALSE. If you want to get notified, then set it
to TRUE.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>EMAILON:FALSE
or
EMAILON:TRUE</tt></pre>
</div></div>
<div class="paragraph"><p>In the computing machine, you need to have two kinds of directory: one is for storing input and output files, and the other for executing scripts. Please, specify a full path not a relative one. You must specify <em>ROUTPUTDIR</em> whatever <em>RUNMODE</em> is. While running jobs, input and output files are stored in <em>ROUTPUTDIR</em> directory. When jobs are finished, files can be copied to the output directory of the base directory in your local machine. This may be confusing, but convenient because we can easily delete <em>ROUTPUTDIR</em> directory when we do not need it. We would want to save files in the output directory in your local machine. Use a full path not a relative one. Create the directory at the compute machine. When the mode of <strong>RNAseq Analysis</strong> is local, create the directory in the local machine, e.g., at the local base directory.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>ROUTPUTDIR:/path/to/routput</tt></pre>
</div></div>
<div class="paragraph"><p>We use other tools or packages. You need to specify paths for those tools.
Please, find the installation section to know how to install them into the
computing machine.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>BWA:usr/bin/bwa-0.6.1/bwa
SAMTOOLS:usr/bin/samtools-0.1.18/samtools
PRINSEQ:usr/bin/prinseq-lite-0.17.4/prinseq-lite.pl
PICARD:usr/bin/picard-tools-1.67
CRAMTOOLS:usr/bin/cramtools-0.8.jar
PARSERNASEQ:usr/bin/ParseRNAseq
PYTHON:/opt/epd/bin/python2.7
CUTADAPT:/home/fs01/sc2265/usr/bin/cutadapt-1.0/cutadapt
CACRSCRIPT:/home/fs01/sc2265/Downloads/r-devel/b/bin/Rscript</tt></pre>
</div></div>
<div class="paragraph"><p><a id="txdb"></a></p></div>
<div class="sect2">
<h3 id="_txdb">txdb</h3>
<div class="paragraph"><p>The annotation of a genome afffects most of analyses such as differential expresssion of genes, functional association, and gene clusters. You had better to decide what gene annotations you use in <strong>RNAseq Analysis</strong>. A genome in genbank format from NCBI is one of the source of genome annotations. Sometimes, we use a gff file from NCBI that accompanies the genbank file. You could use <tt>pl/bioperl-get-protein.pl</tt> to create a gff file from a genbank file. Because annotation files in gff format keep changing, we could not prepare a menu in <strong>RNAseq Analysis</strong>. Many of the R scripts in <strong>RNAseq Analysis</strong> use BioConductor R packages. A BioConductor R package, called <tt>GenomicFeatures</tt>, provides a method, <em>makeTranscriptDb</em> that creates a txdb file. We use a gff file to create a txdb file. As I mentioned earlier, it is not trivial to create a txdb using a gff file although it is easy to use <em>makeTranscriptDb</em>. So, I want to write down some steps of R scripts to create a txdb file using a gff file.  The chromosome names should be named chr with a number.</p></div>
<div class="paragraph"><p>Consider that you have a file, named <tt>NC_004350.gff</tt>, in gff format. I downloaded the file from NCBI Bacterial genome FTP site, and replaced the "NC_004350.2" with "chr1" for handling the file more easily. Start an R session and load BioConductor R pacakges, <tt>rtracklayer</tt> and <tt>GenomicFeatures</tt>. We use a function <em>import.gff3</em> of <tt>rtracklayer</tt> to import the gff file.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; gffFile &lt;- "NC_004350.gff"
&gt; library(rtracklayer)
&gt; library(GenomicFeatures)
&gt; sm.gff &lt;- import.gff3(gffFile)</tt></pre>
</div></div>
<div class="paragraph"><p>The function, <em>import.gff3</em>, is helpful in reading a gff file. The output variable <em>sm.gff</em> is a <em>RangedData</em> object, contains a column called <em>type</em> that we use to subset genes and CDS regions of <em>sm.gff</em>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; sm.gene &lt;- sm.gff[sm.gff$type=="gene",]
&gt; sm.CDS &lt;- sm.gff[sm.gff$type=="CDS",]
&gt; sm.rrna &lt;- sm.gff[sm.gff$type=="rRNA",]</tt></pre>
</div></div>
<div class="paragraph"><p>The <em>sm.gene</em> object contains all of the annotations including mRNA, rRNA, and others.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; unique(sm.gff$type)
[1] region gene   CDS    rRNA   exon   tRNA
Levels: CDS exon gene region rRNA tRNA</tt></pre>
</div></div>
<div class="paragraph"><p>We call eventually <em>makeTranscriptDb</em> to create a txdb variable, which is saved using <em>saveFeatures</em> function. These two functions are part of BioConductor R package, <tt>GenomicFeatures</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; txdb &lt;- makeTranscriptDb (transcripts, splicings, chrominfo=chrominfo)
&gt; gffFileBasename &lt;- unlist(strsplit(gffFile,"\\."))[1]
&gt; saveFeatures(txdb,file=paste(gffFileBasename,"txdb",sep="."))</tt></pre>
</div></div>
<div class="paragraph"><p>You may have noticed three R variables, <em>transcripts</em>, <em>splicings</em>, and <em>chrominfo</em>. You might want to refer to the help page for <em>makeTranscriptDb</em>. The first variable is an R data frame with id, name, chrom, strand, start, and end. Prepend the 6 arguments with tx with a underscore as follows.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; transcripts &lt;-
    data.frame( tx_id=seq(length(sm.gene$locus_tag)),
                tx_name=sm.gene$locus_tag,
                tx_chrom=sm.gene$space,
                tx_strand=sm.gene$strand,
                tx_start=start(sm.gene),
                tx_end=end(sm.gene) )
&gt; head (transcripts)
  tx_id tx_name tx_chrom tx_strand tx_start tx_end
1     1  SMU_01     chr1         +      194   1552
2     2  SMU_02     chr1         +     1708   2844
3     3  SMU_05     chr1         +     3106   3297</tt></pre>
</div></div>
<div class="paragraph"><p>Let us have <em>chrominfo</em> before explaining how to create <em>splicing</em> because <em>chrominfo</em> is much easier to explain. Our example of a gff file contains a single chromosome, which we renamed it chr1. The length is 2032925. Although it is a circular genome, we set it the corresponding option to FALSE (I do not know what side effect of this is). If there are multiple chromosomes, use R vectors for chrom and length, two of which should parallel side by side.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; chrominfo &lt;-
    data.frame( chrom="chr1",
                length=2032925,
                is_circular=FALSE )
&gt; chrominfo
  chrom  length is_circular
1  chr1 2032925       FALSE</tt></pre>
</div></div>
<div class="paragraph"><p>A transcript can consist of multiple exons in eukaryotes. The <em>splicing</em> data frame contains exons and their membership in the <em>transcripts</em> data frame. For example, the following is given by the help page of <em>makeTranscriptDb</em> object. We have three transcripts. The first row of <em>splicing</em> data frame represents one and the only exon, which also encodes a protein. The next three exons of 2nd, 3rd, and 4th rows of <em>splicing</em> data frame correspond to three exons that belong to the 2nd transcript. We specify their identifiers in the 2nd transcript using exon&#8217;s rank. The three exons also encode proteins. The 3rd transcript contains two exons that does not encode a protein, which is denoted by NA in cds start and end.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; ?makeTranscriptDb
...
     transcripts &lt;- data.frame(
                        tx_id=1:3,
                        tx_chrom="chr1",
                        tx_strand=c("-", "+", "+"),
                        tx_start=c(1, 2001, 2001),
                        tx_end=c(999, 2199, 2199))
     splicings &lt;-  data.frame(
                        tx_id=c(1L, 2L, 2L, 2L, 3L, 3L),
                        exon_rank=c(1, 1, 2, 3, 1, 2),
                        exon_start=c(1, 2001, 2101, 2131, 2001, 2131),
                        exon_end=c(999, 2085, 2144, 2199, 2085, 2199),
                        cds_start=c(1, 2022, 2101, 2131, NA, NA),
                        cds_end=c(999, 2085, 2144, 2193, NA, NA))

     txdb &lt;- makeTranscriptDb(transcripts, splicings)
...</tt></pre>
</div></div>
<div class="paragraph"><p>You have to make sure that you have <em>splicings</em> object correctly. There is no unique way of creating the object. We describe two ways. First, CDS might have an attribute that tells us to which gene that the CDS belongs. The NCBI&#8217;s gff file contains Parent column, which specifies CDS&#8217;s membership. We can find the parent gene by searching <em>sm.gene</em> ID for the parent of a CDS.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; colnames(sm.CDS)
 [1] "source"        "type"          "score"         "strand"
 [5] "phase"         "Dbxref"        "gbkey"         "gene"
 [9] "gene_synonym"  "genome"        "ID"            "Is_circular"
[13] "locus_tag"     "mol_type"      "Name"          "Note"
[17] "old_locus_tag" "Parent"        "product"       "protein_id"
[21] "pseudo"        "strain"        "transl_table"</tt></pre>
</div></div>
<div class="paragraph"><p>Make sure that each CDS has a single parent, which should be true of bacterial genomes. I intentially added "&gt;" so that you cannot just copy and paste the code. You have to see what each step does. We iterate over all of the gene ID to construct <em>grCDS.start</em> and <em>grCDS.end</em>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; x &lt;- c()
&gt; y &lt;- c()
&gt; z &lt;- unlist(sm.CDS$Parent)
&gt; stopifnot(nrow(sm.CDS)==length(unlist(sm.CDS$Parent)))
for (i in sm.gene$ID) {
  if (sum(z == i) &gt; 0) { # Check if the CDS contains the gene
    if (length(start(sm.CDS)[z == i]) != 1) { # Check if there is a unique parent gene for CDS.
      print(paste("Check",gffFile))
      print(start(sm.CDS)[z == i])
      stop(paste("There are multiple CDS for gene",i))
    }
    x &lt;- c(x,start(sm.CDS)[z == i]) # Add the start position of the CDS of the gene
    y &lt;- c(y,end(sm.CDS)[z == i])   # Add the end position.
  } else {
    x &lt;- c(x,NA) # Add NA if the CDS set does not contain the gene
    y &lt;- c(y,NA) # Add NA for the end position.
  }
}
&gt; grCDS.start &lt;- x
&gt; grCDS.end &lt;- y
&gt; rm(i,x,y,z)</tt></pre>
</div></div>
<div class="paragraph"><p>Then, we can create <em>splicings</em> data frame, which is used to create a txdb file.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt; splicings &lt;-
    data.frame( tx_id=seq(length(sm.gene$locus_tag)),
                exon_rank=seq(length(sm.gene$locus_tag)),
                exon_start=start(sm.gene),
                exon_end=end(sm.gene),
                exon_name=sm.gene$locus_tag,
                cds_start=grCDS.start,
                cds_end=grCDS.end )</tt></pre>
</div></div>
<div class="paragraph"><p>FIXME: Sometimes, CDS parent information may unavailable. The second approach is to use CDS positions to find their parent genes.</p></div>
<div class="paragraph"><p>Add more attributes for genes such as protein ID, product, or some other annotations. This seems not available in <em>makeTranscriptDb</em> function at the moment. We might need to use <em>sm.gff</em> or gff file to add more annotation to CDS or transcripts.</p></div>
<div class="paragraph"><p>How do we use the txdb? We take transcripts and CDS using <em>transcripts</em> and <em>cds</em> functions of <tt>GenomicFeatures</tt> R package. Subset CDS of <em>transcripts</em> by compairing transcript IDs. Then, we add a type column. This can be a step where I can add more annodataion. This operation gives us txdb with and without CDS regions.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>library(rtracklayer)
library(GenomicFeatures)
txdb &lt;- loadFeatures("1.txdb")
feature.tx &lt;- transcripts(txdb)
feature.cds &lt;- cds(txdb,columns="tx_id")
stopifnot(length(feature.cds) == length(unlist(elementMetadata(feature.cds)$tx_id)))
x &lt;- elementMetadata(feature.tx)$tx_id %in% unlist(elementMetadata(feature.cds)$tx_id)
feature.cds &lt;- feature.tx[x]
elementMetadata(feature.cds) &lt;- data.frame(tx_id=elementMetadata(feature.cds)$tx_id, tx_name=elementMetadata(feature.cds)$tx_name,type="CDS")</tt></pre>
</div></div>
<div class="paragraph"><p>We also have transcripts without CDS, and combine <em>feature.cds</em> and <em>feature.nocds</em> to create <em>feature.txnc</em>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>feature.txnc &lt;- feature.cds
feature.nocds &lt;- feature.tx[!x]
if (length(feature.nocds) &gt; 0) {
  elementMetadata(feature.nocds) &lt;- data.frame(tx_id=elementMetadata(feature.nocds)$tx_id, tx_name=elementMetadata(feature.nocds)$tx_name,type="NOCDS")
  feature.txnc &lt;- c(feature.txnc,feature.nocds)
}</tt></pre>
</div></div>
<div class="paragraph"><p>We also find the rest of regions as non-genic loci. The combined object, <em>feature.txnc</em>, contains three types, CDS, NOCDS, and NG (non-genic).</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>feature.ng &lt;- transcripts(txdb)
strand(feature.ng) &lt;- '*'
feature.ng &lt;- gaps(reduce(feature.ng))
feature.ng &lt;- feature.ng[strand(feature.ng)=='*']
if (length(feature.ng) &gt; 0) {
  x &lt;- seq(from=length(feature.tx)+1,to=length(feature.tx)+length(feature.ng))
  y &lt;- paste("NC",x,sep="_")
  elementMetadata(feature.ng) &lt;- data.frame(tx_id=x,tx_name=y,type="NG")
  rm(x,y)
  feature.txnc &lt;- c(feature.txnc,feature.ng)
}
&gt; Rle(elementMetadata(feature.txnc)$type)
'factor' Rle of length 3743 with 3 runs
  Lengths:  1960    82  1701
  Values :   CDS NOCDS    NG
Levels(3): CDS NOCDS NG</tt></pre>
</div></div>
<div class="paragraph"><p>We can export the combined R object <em>feature.txnc</em> to save it to a BED file. The function <em>export</em> uses the file extension to recognize the export file format. Before exporting <em>feature.txnc</em>, we populate the names with transcript names, which should be used as values of the name column of the BED foramt. Otherwise, only dots will be printed out at the name column in a BED file.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>names(feature.txnc) &lt;- elementMetadata(feature.txnc)[,"tx_name"]
export(feature.txnc,"txnc.bed")</tt></pre>
</div></div>
<div class="paragraph"><p>Because the procedure is not standarized, you must make sure that your annotations are correct. You could make sure whether the extraction of CDS regions from your <em>feature.txnc</em> was correctly performed using Galaxy web interface that was available at <a href="https://main.g2.bx.psu.edu">https://main.g2.bx.psu.edu</a>. Use "Upload File" tool to send the gff file <tt>NC_004350.gff</tt>, and convert it to a BED file using "GFF-to-BED" tool. Use "Filter" tool to divide the BED file to a part of plus strand and another of minus strand. Download the two files and concatenate them to a file. Compare the concatenated file and <tt>txnc.bed</tt>. They must be the same except name columns. You may need to verify the correctness of the annotation txdb using other methods for cross check.</p></div>
<div class="paragraph"><p><a id="Installation"></a></p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_installation">Installation</h2>
<div class="sectionbody">
<div class="paragraph"><p>Software packages used in <strong>RNAseq Analysis</strong> should be installed on
the local and/or remote machine. First, create directories called <tt>downloads/build</tt>.
If you use your local machine, follow this instruction in your machine.
You have to log in your remote machine if you use other machine such as a linux
cluster.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ mkdir -p downloads/build</tt></pre>
</div></div>
<div class="paragraph"><p>Create a <tt>bin</tt> directory to save executable files.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ mkdir -p $HOME/usr/bin</tt></pre>
</div></div>
<div class="sect2">
<h3 id="_bwa">BWA</h3>
<div class="paragraph"><p>Dowanload BWA source code available at <a href="http://bio-bwa.sourceforge.net">http://bio-bwa.sourceforge.net</a>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ downloads/bwa-0.6.2.tar.bz2</tt></pre>
</div></div>
<div class="paragraph"><p>Extract the file and compile it.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ cd downloads/build
$ tar jxf ../bwa-0.6.2.tar.bz2
$ cd bwa-0.6.2; make; cd ..
$ mv bwa-0.6.2 $HOME/usr/bin</tt></pre>
</div></div>
<div class="paragraph"><p>Edit <tt>conf/README</tt> if necessary; Add the following line. The base directory is
$HOME or your home directory. Shell variable $BWA is replaced by
<tt>/home/userid/usr/bin/bwa-0.6.2/bwa</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>BWA:usr/bin/bwa-0.6.2/bwa</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_samtools">SAMTools</h3>
<div class="paragraph"><p>Download SAMtools available at <a href="http://samtools.sourceforge.net">http://samtools.sourceforge.net</a>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ downloads/samtools-0.1.18.tar.bz2</tt></pre>
</div></div>
<div class="paragraph"><p>Extract the file and compile it.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ cd downloads/build
$ tar jxf ../samtools-0.1.18.tar.bz2
$ cd samtools-0.1.18; make; cd ..
$ mv samtools-0.1.18 $HOME/usr/bin</tt></pre>
</div></div>
<div class="paragraph"><p>Edit <tt>conf/README</tt> if necessary; Add the following line. The base directory is
$HOME or your home directory. Shell variable $SAMTOOLS is replaced by
<tt>/home/userid/usr/bin/samtools-0.1.18/samtools</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>SAMTOOLS:usr/bin/samtools-0.1.18/samtools</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_prinseq">prinseq</h3>
<div class="paragraph"><p>Download a prinseq lite version available at <a href="http://prinseq.sourceforge.net">http://prinseq.sourceforge.net</a>.
Extract the file and move the directory.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ cd downloads/build
$ tar zxf ../prinseq-lite-0.17.4.tar.gz
$ mv prinseq-lite-0.17.4 $HOME/usr/bin</tt></pre>
</div></div>
<div class="paragraph"><p>Edit <tt>conf/README</tt> if necessary; Add the following line. The base directory is
$HOME or your home directory. Shell variable $PRINSEQ is replaced by
<tt>/home/userid/usr/bin/prinseq-lite-0.17.4/prinseq-lite.pl</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>PRINSEQ:usr/bin/prinseq-lite-0.17.4/prinseq-lite.pl</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_cutadapt">cutadapt</h3>
<div class="paragraph"><p>Download cutadapt available at <a href="http://code.google.com/p/cutadapt">http://code.google.com/p/cutadapt</a>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ cd downloads/build
$ tar zxf ../cutadapt-1.0.tar.gz
$ mv cutadapt-1.0 $HOME/usr/bin</tt></pre>
</div></div>
<div class="paragraph"><p>Edit <tt>conf/README</tt> if necessary; Add the following line. The base directory is
$HOME or your home directory. Note that <em>CUTADAPT</em> must be the absolute path of
the binary executable. We would call python with the executable. Do not use a
relative file path.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>CUTADAPT:/home/user/usr/bin/cutadapt-1.0/cutadapt</tt></pre>
</div></div>
<div class="paragraph"><p>NOTE:The following prorams need to be checked. I have not yet time to edit the
following description of the installation.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ python setup.py build
$ sudo python setup.py install
CUTADAPT:/usr/local/bin/cutadapt</tt></pre>
</div></div>
<div class="listingblock">
<div class="title">To install cutadapt in a linux cluster</div>
<div class="content">
<pre><tt>python2.7 setup.py build_ext -i
python2.7 path/to/cutadapt --help</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_python">python</h3>
<div class="paragraph"><p>Set <em>PYTHON</em> variable to a file.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>PYTHON:/path/to/python2.7</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_rscript">Rscript</h3>
<div class="paragraph"><p>Install R version 2.15 and Bioconductor 2.10. Set <em>CACRSCRIPT</em> variable.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>CACRSCRIPT:/path/to/Rscript</tt></pre>
</div></div>
<div class="paragraph"><p>In R, run the following line of</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>source("http://www.bioconductor.org/biocLite.R")</tt></pre>
</div></div>
<div class="paragraph"><p>to install the R packages:</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>install.packages("plotrix")
install.packages("gplots")
biocLite("DESeq")
biocLite("ShortRead")
biocLite("rtracklayer")
biocLite("GenomicFeatures")
biocLite("qrqc")
biocLite("VariantAnnotation")
biocLite("goseq")
The following are installed by the above R packages.
biocLite("GenomicRanges")
biocLite("Rsamtools")
The following are optional.
biocLite("edgeR")
biocLite("DEGseq")</tt></pre>
</div></div>
<div class="paragraph"><p>You also need to install an R package, <strong>smutans</strong>, that accompanies <strong>RNAseq Analysis</strong>. The R package uses <tt>DESeq</tt>.</p></div>
<div class="listingblock">
<div class="title">To install smutans R package.</div>
<div class="content">
<pre><tt>R CMD install smutans_0.1.9.tar.gz</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_libxml2">libxml2</h3>
<div class="paragraph"><p>Download <tt>libxml2</tt> from <a href="http://xmlsoft.org">http://xmlsoft.org</a> and install it into the machine. R package <tt>DESeq</tt> requires other R packages that need <tt>libxml2</tt>.</p></div>
</div>
<div class="sect2">
<h3 id="_zlib">zlib</h3>
<div class="paragraph"><p>Sometime this is also required by <tt>Rsamtools</tt>.</p></div>
</div>
<div class="sect2">
<h3 id="_blast">BLAST</h3>
<div class="paragraph"><p>We will use BLAST for functional association of differentially expressed genes.
We do not provide features of <strong>RNAseq Analysis</strong> that uses <tt>BLAST</tt>.
<a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/</a>
ncbi-blast-2.2.25+-x64-linux.tar.gz
is used in the cluster. You could download a dmg for mac installation, which was installed at the following directory:</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>/usr/local/ncbi/blast/bin</tt></pre>
</div></div>
<div class="paragraph"><p>Because we use a relative path for programs, copy executables in the bin directory to <tt>usr/bin</tt> of your home at the local machine or at the remote one.</p></div>
<div class="listingblock">
<div class="title">Add the following lines to <tt>conf/README</tt> for global configuration</div>
<div class="content">
<pre><tt>MAKEBLASTDB:usr/bin/makeblastdb
BLASTN:usr/bin/blastn
BLASTP:usr/bin/blastp</tt></pre>
</div></div>
<div class="paragraph"><p>Refer to documentation available in <a href="http://www.ncbi.nlm.nih.gov/books/NBK1762">http://www.ncbi.nlm.nih.gov/books/NBK1762</a>.</p></div>
<div class="paragraph"><p><tt>deseq-blast</tt></p></div>
<div class="paragraph"><p>We use <tt>uniprot90</tt> for functional association.
wget <a href="ftp://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref90/uniref90.fasta">ftp://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref90/uniref90.fasta</a></p></div>
</div>
<div class="sect2">
<h3 id="_bioperl">BioPerl</h3>
<div class="paragraph"><p>Install BioPerl. We need it when analyzing BLAST result files.</p></div>
</div>
<div class="sect2">
<h3 id="_muscle">MUSCLE</h3>
<div class="paragraph"><p>We do not provide features of <strong>RNAseq Analysis</strong> that uses <tt>MUSCLE</tt>.
<a href="http://www.drive5.com/muscle/downloads3.8.31/muscle3.8.31_i86linux64.tar.gz">http://www.drive5.com/muscle/downloads3.8.31/muscle3.8.31_i86linux64.tar.gz</a></p></div>
<div class="listingblock">
<div class="title">Create a <tt>build</tt> directory at the base directory <tt>run/rnaseq/110111</tt> of a cluster.</div>
<div class="content">
<pre><tt>cac:run/rnaseq/110111/build/muscle3.8.31_i86linux64
cac:run/rnaseq/110111/build/ncbi-blast-2.2.25+
cac:run/rnaseq/110111/build/RNAz-2.1</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_subread">subread</h3>
<div class="paragraph"><p>We have not implemented a feature of <strong>RNAseq Analysis</strong> that uses <tt>subread</tt>.
Download subread available at <a href="http://subread.sourceforge.net">http://subread.sourceforge.net</a>.
Another short read aligner. This could be used in menu <strong>batch3</strong>.
As of April 2012, it did not support Mac OS X. Authors said that they would make
a Mac OS X version available in 2012.</p></div>
</div>
<div class="sect2">
<h3 id="_bowtie">BOWTIE</h3>
<div class="paragraph"><p>We have not yet used <tt>BOWTIE</tt> in <strong>RNAseq Analysis</strong>.
Download BOWTIE and install it under <tt>src</tt> directory. The following command
should print a help message from it. Edit <tt>conf/README</tt> if necessary.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ src/bowtie-0.12.7/bowtie</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_segemehl">segemehl</h3>
<div class="paragraph"><p>We have not yet used <tt>segemehl</tt> in <strong>RNAseq Analysis</strong>.
Download segemehl at <a href="http://www.bioinf.uni-leipzig.de/Software/segemehl">http://www.bioinf.uni-leipzig.de/Software/segemehl</a>. You
could compile the source in this repository of rnaseq-analysis.
<a href="http://www.bioinf.uni-leipzig.de/Software/segemehl/">http://www.bioinf.uni-leipzig.de/Software/segemehl/</a></p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ cd src/segemehl_0_0_9_4
$ make</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_picard">Picard</h3>
<div class="paragraph"><p>We have not yet implemented a feature of <strong>RNAseq Analysis</strong> that uses <tt>Picard</tt>. We might use this when &#8220;cramming&#8221; FASTQ files.
Download Picard available at <a href="http://picard.sourceforge.net">http://picard.sourceforge.net</a>. Unzip the file and
move directories.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ cd downloads/build
$ unzip ../picard-tools-1.65.zip
$ mv picard-tools-1.65 $HOME/usr/bin
$ mv snappy-java-1.0.3-rc3.jar $HOME/usr/bin</tt></pre>
</div></div>
<div class="paragraph"><p>Edit <tt>conf/README</tt> if necessary; Add the following line. The base directory is
$HOME or your home directory. Shell variable $PICARD is replaced by
<tt>/home/userid/usr/bin/picard-tools-1.65</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>PICARD:usr/bin/picard-tools-1.65</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_cram_toolkit">CRAM Toolkit</h3>
<div class="paragraph"><p>We have not yet implemented a feature of <strong>RNAseq Analysis</strong> that uses <tt>cram</tt>.
Download CRAM Toolkit available at <a href="http://www.ebi.ac.uk/ena/about/cram_toolkit">http://www.ebi.ac.uk/ena/about/cram_toolkit</a>.
Move the file to <tt>$HOME/usr/bin</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ cd downloads/build
$ mv ../cramtools.jar $HOME/usr/bin</tt></pre>
</div></div>
<div class="paragraph"><p>Edit <tt>conf/README</tt> if necessary; Add the following line. The base directory is
$HOME or your home directory. Shell variable $CRAMTOOLS is replaced by
<tt>/home/userid/usr/bin/cramtools.jar</tt>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>CRAMTOOLS:usr/bin/cramtools.jar</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_fastax_toolkit">FASTAX-Toolkit</h3>
<div class="paragraph"><p>We have not yet used this tool kit in <strong>RNAseq Analysis</strong>.
Download FASTAX-Toolkit at
<a href="http://hannonlab.cshl.edu/fastx_toolkit/fastx_toolkit-0.0.13.tar.bz2">http://hannonlab.cshl.edu/fastx_toolkit/fastx_toolkit-0.0.13.tar.bz2</a>
and
<a href="http://hannonlab.cshl.edu/fastx_toolkit/libgtextutils-0.6.tar.bz2">http://hannonlab.cshl.edu/fastx_toolkit/libgtextutils-0.6.tar.bz2</a>.</p></div>
</div>
<div class="sect2">
<h3 id="_tophat_and_cufflinks">TopHat and Cufflinks</h3>
<div class="paragraph"><p>We have not yet used <tt>TopHat</tt> in <strong>RNAseq Analysis</strong>.
Download the Mac OS X x86_64 binary from <a href="http://tophat.cbcb.umd.edu">here</a>.
Untar it and add the directory to PATH so that I can run it without full path.
Download the Mac OS X x86_64 binary from <a href="http://cufflinks.cbcb.umd.edu">here</a>.
Untar it and add the directory to PATH so that I can run it without full path</p></div>
</div>
<div class="sect2">
<h3 id="_rnaz">RNAz</h3>
<div class="paragraph"><p>We do not provide features of <strong>RNAseq Analysis</strong> that uses <tt>RNAz</tt>.
<a href="http://www.tbi.univie.ac.at/~wash/RNAz/">http://www.tbi.univie.ac.at/~wash/RNAz/</a> or
<a href="https://github.com/wash/rnaz">https://github.com/wash/rnaz</a>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>wget http://www.tbi.univie.ac.at/~wash/RNAz/RNAz-2.1.tar.gz
wget http://www.tbi.univie.ac.at/RNA/ViennaRNA-1.8.5.tar.gz</tt></pre>
</div></div>
<div class="paragraph"><p>In the CAC:</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>mkdir b
cd b
../configure --prefix=$HOME/usr</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_rnaplex">RNAplex</h3>
<div class="paragraph"><p>We do not provide features of <strong>RNAseq Analysis</strong> that uses <tt>RNAplex</tt>.
<a href="http://www.bioinf.uni-leipzig.de/Software/RNAplex/">http://www.bioinf.uni-leipzig.de/Software/RNAplex/</a></p></div>
</div>
<div class="sect2">
<h3 id="_vienna_rna_package">Vienna RNA Package</h3>
<div class="paragraph"><p>We do not provide features of <strong>RNAseq Analysis</strong> that uses Vienna RNA package.
<a href="http://www.tbi.univie.ac.at/~ivo/RNA/RNAplfold.html">http://www.tbi.univie.ac.at/~ivo/RNA/RNAplfold.html</a></p></div>
<div class="paragraph"><p><a id="SpeciesFile"></a></p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_species_file">Species File</h2>
<div class="sectionbody">
<div class="paragraph"><p>A setup for <strong>RNAseq Analysis</strong> is described in a species file in <tt>species</tt> directory.  The <tt>species</tt> directory should be placed at <em>ROOTANALYSISDIR</em>. Choose a species name for a species file. Create a file with the name, and place it in the directory of <tt>species</tt>. You can use <tt>species/template</tt>. As the <tt>conf/README</tt> file, a species file contains two kinds of lines: one for comment, and the other for variable assignment. Comments should start with # at the first column of a line.</p></div>
<div class="listingblock">
<div class="title">Comment in a species file</div>
<div class="content">
<pre><tt># Comment</tt></pre>
</div></div>
<div class="paragraph"><p>A variable assignment should start without # at the first column. Variables tend to be capitalized, and are followed by a colon, then a variable value. No spaces are allowed.</p></div>
<div class="listingblock">
<div class="title">Variable in a species file</div>
<div class="content">
<pre><tt>VARIABLE:value</tt></pre>
</div></div>
<div class="paragraph"><p>Check a list of species file variables in the following.  Note that we must have a single variable. Multiple instances of species file variables would cause problems.</p></div>
<div class="olist arabic"><div class="title">Check the following spcies file variables</div><ol class="arabic">
<li>
<p>
fastq files
</p>
</li>
<li>
<p>
quality score scheme
</p>
</li>
<li>
<p>
adapter sequences
</p>
</li>
<li>
<p>
repetition in <em>REPETITION</em> (optional)
</p>
</li>
<li>
<p>
fastq file indices in <em>FASTQFILES</em>
</p>
</li>
<li>
<p>
fastq file labels in <em>FASTQLABEL</em>
</p>
</li>
<li>
<p>
cluster working directory in <em>CACWORKDIR</em>
</p>
</li>
<li>
<p>
reference genome fasta file in <em>REFGENOMEFASTA</em>
</p>
</li>
<li>
<p>
reference genome annotation TrancriptDb file in <em>REFGENOMETXDB</em>
</p>
</li>
<li>
<p>
reference genome annotation file in <em>REFGENOMEGFF</em> (optional)
</p>
</li>
<li>
<p>
pipeline wall time in <em>QCALIGNDEWALLTIME</em>
</p>
</li>
<li>
<p>
minimum alignment mapping quality in <em>MINMAPQ</em>
</p>
</li>
<li>
<p>
minimum base quality to trim 3' end in <em>MINTRIMQ</em>
</p>
</li>
<li>
<p>
keep BAM files using <em>KEEPBAM</em>
</p>
</li>
</ol></div>
<div class="sect2">
<h3 id="_fastq_files">fastq files</h3>
<div class="paragraph"><p>First, place fastq files in your <em>ROOTANALYSISDIR</em> in the computer. Then, specify locations of the fastq files. Variables are prefixed <em>FASTQ</em> with three-digit numbers with zeros prepended. Currently, the file extension must be fq and files must be zipped using <tt>gzip</tt>. So, files should end with <tt>fq.gz</tt>. Use a path relative to <em>ROOTANALYSISDIR</em> directory.</p></div>
<div class="listingblock">
<div class="title">fastq files</div>
<div class="content">
<pre><tt>FASTQ001:path/to/file1.fq.gz
FASTQ002:path/to/file2.fq.gz</tt></pre>
</div></div>
<div class="paragraph"><p>The path <em>ROOTANALYSISDIR</em> must be set in the configuration file, or
<tt>conf/README</tt> (see the section <a href="#Configuration">[Configuration]</a>).</p></div>
<div class="listingblock">
<div class="title">Full path of a root for storing data</div>
<div class="content">
<pre><tt>ROOTANALYSISDIR:/path/to/RNASeq-Analysis</tt></pre>
</div></div>
<div class="paragraph"><p>The full path to the FASTA file will be</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>/path/to/RNASeq-Analysis/path/to/file1.fq.gz</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_quality_score">quality score</h3>
<div class="paragraph"><p>Quality scores can be phred33 or phred64. Use a value of "illumina" for phred64, and a value of "sanger" for phred33.  Variables are prefixed <em>QUALITYSCORE</em> with three-digit numbers with zeros prepended. Because each fastq file has a distinct quality score scheme, you need to specify quality score schemes as many times as fastq files.</p></div>
<div class="listingblock">
<div class="title">Quality scores schemes</div>
<div class="content">
<pre><tt>QUALITYSCORE001:illumina
QUALITYSCORE002:sanger</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_adapter_sequences">adapter sequences</h3>
<div class="paragraph"><p>Each fastq file is associated with an Illumina adapter sequence. They must be specified, and are used to remove adapter sequences in short reads.  Variables are prefixed <em>ADAPTER</em> with three-digit numbers with zeros prepended. The adapter sequences can be found in the name of a fastq file or short reads information.</p></div>
<div class="listingblock">
<div class="title">Adapter sequences</div>
<div class="content">
<pre><tt>ADAPTER001:GATCGGAAGAGCACACGTCTGAACTCCAGTCACCGATGTATCTCGTATGCCGTCTTCTGCTTG
ADAPTER002:GATCGGAAGAGCACACGTCTGAACTCCAGTCACTGACCAATCTCGTATGCCGTCTTCTGCTTG</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_repetition">repetition</h3>
<div class="paragraph"><p>You might have want to repeat the procedure of aligning and counting reads on different reference genomes, or different annotations. Use <em>REPETITION</em> variable to store results from different reference genomes or annotations. If not specified, the default is 1.</p></div>
</div>
<div class="sect2">
<h3 id="_fastq_file_indices_and_labels">fastq file indices and labels</h3>
<div class="paragraph"><p>You may wish to process all of fastq files, or some of them. Specify which fastq files you wish to use using <em>FASTQFILES</em> variable. Use <em>FASTQLABEL</em> to group different fastq files. Labels are used for the purpose of differential expression study. They are saved in a file called <tt>count.txt.index</tt>. Use numbers beyond the largest IDs of FASTQ files when creating test FASTQ files for the purpose of testing the pipeline of <strong>RNAseq Analysis</strong>.</p></div>
<div class="listingblock">
<div class="title">File indices for fastQ files</div>
<div class="content">
<pre><tt>FASTQFILES:1 2
FASTQLABEL:UA159 TW1</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_cluster_working_directory">cluster working directory</h3>
<div class="paragraph"><p>After creating all of the scripts in your local machine, you would submit jobs at a directory in the cluster. You need to create the directory while you prepare a <strong>RNASeq Analysis</strong> by editing a species file.</p></div>
<div class="listingblock">
<div class="title">Cluster working directory</div>
<div class="content">
<pre><tt>CACWORKDIR:run/rnaseq/022712</tt></pre>
</div></div>
<div class="listingblock">
<div class="title">Create the working directory</div>
<div class="content">
<pre><tt>cac&gt; mkdir -p run/rnaseq/022712</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_reference_genome_and_annotation">reference genome and annotation</h3>
<div class="paragraph"><p>Set the path of a reference genome genbank file in <em>REFGENOMEFASTA</em>. Use a path relative to your <em>ROOTANALYSISDIR</em> directory. For example, you would have a directory called <tt>data</tt> in <strong>RNASeq-Analysis</strong> <em>ROOTANALYSISDIR</em> directory. Use a path relative to <em>ROOTANALYSISDIR</em> directory.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>REFGENOMEFASTA:data/to/NC_004350.fna</tt></pre>
</div></div>
<div class="paragraph"><p>To map short reads on a reference genome, you must specify the file paths of reference genome files and gene annotations. The reference genome sequence is in FASTA format. Each sequence represents a contiguous genome. A complete bacterial genome can be a single sequence because only one chromosome is available. Genomes can consist of multiple chromosomes when the genome sequencing is incomplete or the genome contains multiple chromosomes. The FASTA file would contain multiple FASTA sequences. Because it is convenient to assume sequence names, we rename each sequence using chromosome numbers in the FASTA file of a genome. For example, the first line of a FASTA file for a reference genome must be</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>&gt;chr1</tt></pre>
</div></div>
<div class="paragraph"><p>Change the names of other sequences accordingly: chr2 for the second chrmosome. An incomplete bacterial genome would have multiple contigs, each of which can be considered a chromosome. Use a path relative to <em>ROOTANALYSISDIR</em> directory. We also have to change chromsome name in a gff file as well. If there are other annotation files, change the names accordingly.</p></div>
<div class="listingblock">
<div class="title">Reference genomes</div>
<div class="content">
<pre><tt>REFGENOMEGFF:path/to/NC_004350.gff</tt></pre>
</div></div>
<div class="paragraph"><p>We can convert a gff file to a txdb file. See <tt>R/gff2txdb.R</tt> or <tt>R/xbasegff2txdb.R</tt> for detail. Use a path relative to <em>ROOTANALYSISDIR</em> directory.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>REFGENOMETXDB:path/to/NC_004350.txdb</tt></pre>
</div></div>
<div class="paragraph"><p>When gff files are not available, use <tt>pl/bp_genbank2gff3.pl</tt> to convert a genbank file to a gff file.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>cat file.gbk | perl pl/bp_genbank2gff3.pl -in stdin -out stdout &gt; file.gff</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_pipeline_wall_time">pipeline wall time</h3>
<div class="paragraph"><p>Set the maximum compute time for aligning and counting short reads.</p></div>
<div class="listingblock">
<div class="title">pipeline wall time</div>
<div class="content">
<pre><tt>QCALIGNDEWALLTIME:8</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_minimum_alignment_mapping_quality">minimum alignment mapping quality</h3>
<div class="paragraph"><p>We filter out short reads with bad mapping quality scores.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>MINMAPQ:30</tt></pre>
</div></div>
<div class="paragraph"><p>Use <em>DESEQ</em> to use menu <strong>deseq</strong>. A pairwise comparison is performed for each pair of <em>FASTQLABEL</em>. Pairwise comparisons are separated by a slash.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>DESEQ:UA159/1SM1,UA159/11VS1,UA159/15JP3,UA159/N29,UA159/NLML4,UA159/NMT4863</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_minimum_base_quality_to_trim_3_end">minimum base quality to trim 3' end</h3>
<div class="paragraph"><p>We trim the 3' end of short reads using a threshold of base quality.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>MINTRIMQ:20</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_keep_bam_files">keep BAM files</h3>
<div class="paragraph"><p>We might want to keep sorted BAM files by setting <em>KEEPBAM</em> variable.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>KEEPBAM:YES</tt></pre>
</div></div>
<div class="paragraph"><p><a id="Files"></a></p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_files">Files</h2>
<div class="sectionbody">
<div class="paragraph"><p>We would execute menus in a directory. The directory can be the directory you created from the git source repository, or a separate directory. We call the git source directory te Dropbox base directory because I would a directory in a Dropbox for storing source code. It could be called a git source code directory. The Dropbox base directory contains files and directories. We could have a separate directory called a local base directory. The local base directory contains soft links to some of directories in the Dropbox directory. As I mentioned earlier, a separate hard drive can be used to store an <tt>output</tt> directory. So, we could have a storage base directory. The local base directory would also contains soft links to directories in the sotrage base directory. Three base directories are used for the convenience of development.</p></div>
<div class="listingblock">
<div class="title">Dropbox base directory</div>
<div class="content">
<pre><tt>/Users/goshng/Dropbox/Documents/Projects/RNASeq-Analysis</tt></pre>
</div></div>
<div class="listingblock">
<div class="title">Local base directory</div>
<div class="content">
<pre><tt>/Users/goshng/Documents/Projects/RNASeq-Analysis</tt></pre>
</div></div>
<div class="listingblock">
<div class="title">Storage base directory</div>
<div class="content">
<pre><tt>/Volumes/Elements/Documents/Projects/RNASeq-Analysis</tt></pre>
</div></div>
<div class="paragraph"><p>The Dropbox and local base directories can be combined into one. We would clone a git repository via github at th Dropbox <tt>Projects</tt> directory, or local <tt>Projects</tt> directory. When cloning a git repository at a Dropbox base directory, you may have to still create a local base directory. In the local base directory, link some directories (<tt>pl</tt>, <tt>sh</tt>, and <tt>src</tt>) and the main script file <tt>run</tt> using soft links. The storage base directory would contain directories for input and output files such as <tt>data</tt>, <tt>downloads</tt>, <tt>email</tt>, <tt>log</tt>, and <tt>output</tt> directories. Users can create other files and soft-link them at the local base directory. This soft-links allow us to use the local base directory as the main base directory by merging source code and data. Use <tt>sh/make-soft-links.sh</tt> to create soft links in the local base directory. Edit <tt>sh/make-soft-links.sh</tt> to change the value of <em>LOCALBASEDIR</em> and <em>STORAGEBASEDIR</em>.</p></div>
<div class="paragraph"><p>In the storage base directory, create the directories for storing data.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>mkdir data downloads email log output</tt></pre>
</div></div>
<div class="paragraph"><p>Run <tt>sh/make-soft-links.sh</tt> to create soft links in the local base directory.</p></div>
<div class="listingblock">
<div class="title">Clone a git repository (read-only and edit)</div>
<div class="content">
<pre><tt>git clone git://github.com/goshng/RNASeq-Analysis.git
git clone git@github.com:goshng/RNASeq-Analysis.git</tt></pre>
</div></div>
<div class="sect2">
<h3 id="_authors">AUTHORS</h3>
<div class="paragraph"><p>Sang Chul Choi uses Perl, R, and NGS programs to analyze bacterial transcriptome. Charles Danko guides the development in the early stage of the development of <strong>RNAseq Analysis</strong>.</p></div>
</div>
<div class="sect2">
<h3 id="_copying">COPYING</h3>
<div class="paragraph"><p>The GNU General Public License.</p></div>
</div>
<div class="sect2">
<h3 id="_changelog">ChangeLog</h3>
<div class="paragraph"><p>This is a little different from a typical ChangeLog. It is more or less of history of my work. You could find some of actual codes or scripts in <tt>log</tt> directory.</p></div>
</div>
<div class="sect2">
<h3 id="_install">INSTALL</h3>
<div class="paragraph"><p>The instruction of installing RNAseq Analysis.</p></div>
</div>
<div class="sect2">
<h3 id="_news">NEWS</h3>
<div class="paragraph"><p>A list of user-visible changes to <strong>RNAseq Analysis</strong> worth mentioning.</p></div>
</div>
<div class="sect2">
<h3 id="_r">R</h3>
<div class="paragraph"><p>R scripts.</p></div>
</div>
<div class="sect2">
<h3 id="_readme">README</h3>
<div class="paragraph"><p>A README of the root directory.</p></div>
</div>
<div class="sect2">
<h3 id="_thanks">THANKS</h3>
<div class="paragraph"><p>People help the development of <strong>RNAseq Analysis</strong>.</p></div>
</div>
<div class="sect2">
<h3 id="_todo">TODO</h3>
<div class="paragraph"><p>New features of <strong>RNAseq Analysis</strong>.</p></div>
</div>
<div class="sect2">
<h3 id="_bin">bin</h3>
<div class="paragraph"><p>Binary or shell script files.</p></div>
</div>
<div class="sect2">
<h3 id="_c">c</h3>
<div class="paragraph"><p>A shell script that uses <tt>asciidoc</tt> to convert <tt>doc/Manual</tt> to an html file.</p></div>
</div>
<div class="sect2">
<h3 id="_conf">conf</h3>
<div class="paragraph"><p>The <strong>RNAseq Analysis</strong> is configured using the <tt>README</tt> in the <tt>conf</tt> directory.</p></div>
</div>
<div class="sect2">
<h3 id="_doc">doc</h3>
<div class="paragraph"><p>Users should refer to documentation in <tt>doc</tt> directory, which also contains <tt>lyx</tt> files for potential publications from the analysis of bacterial transcriptome.</p></div>
</div>
<div class="sect2">
<h3 id="_log">log</h3>
<div class="paragraph"><p>This is placed at the storage base directory, where a number of directories can be placed as you do some analyses. The directory name is MMDDYY, where MM is month, DD is a day, and YY is the year. See the ChangeLog for some working information.</p></div>
</div>
<div class="sect2">
<h3 id="_package">package</h3>
<div class="paragraph"><p>We use R-based analyses to postprocess the results from computationally intensive analyses, which are mostly done in <strong>RNAseq Analysis</strong>. Parts of the analyses can be incorporated in a manuscript for publication as a part of the result section.</p></div>
</div>
<div class="sect2">
<h3 id="_pl">pl</h3>
<div class="paragraph"><p>Perl scripts.</p></div>
</div>
<div class="sect2">
<h3 id="_run">run</h3>
<div class="paragraph"><p>It shows a menu of commands provided by <strong>RNAseq Analysis</strong>.</p></div>
</div>
<div class="sect2">
<h3 id="_sh">sh</h3>
<div class="paragraph"><p>Bash scripts.</p></div>
<div class="sect3">
<h4 id="_batch_qcalignde_sh">batch-qcalignde.sh</h4>
<div class="paragraph"><p>This explains details of <tt>run-qcalignde.sh</tt> batch script.
One job for a raw fastq file would be submitted to compute nodes. The job is
described in <tt>batch-qcalignde.sh</tt> at the CAC working directory. After a job is
in queue and changes to a run state in a compute node, the job would copy data
files to the compute node.
We copy Perl scripts of <strong>RNA-seq Analysis</strong>, and other binary executables to a
temporary directory of the compute node from the login node of the cluster: <tt>pl</tt>
directory of <strong>RNA-seq Analysis</strong>, <tt>samtools</tt>, and <tt>bwa</tt>. The compute node
would have a temporary directory, which is the actual working directory at a
compute node: <em>TMPDIR</em> is the shell variable to the temporary directory. In
addition to stand-alone executables we need several shell and R script files.
These include <tt>job-fastq-qc</tt>, <tt>job-bwa-align</tt>, <tt>job-de</tt>, and <tt>job-de.R</tt>.
Upto now we have copied stand-alone executables, and scripts.
Now we make directories at the compute node working directory, and copy small
and common data files from the head node to the the compute node.
Three directories for species name "ua159" and repetition 1 are created at the
compute node working directory: shell variables <em>CDATADIR</em>, and <em>CBWADIR</em>.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>CDATADIR=output/ua159/1/data
CBWADIR=output/ua159/1/bwa</tt></pre>
</div></div>
<div class="paragraph"><p>Most of jobs are executed in shell function <em>process-data</em> of
<tt>batch-qcalignde.sh</tt>. Quite often jobs can execute multiple instances of
executables depending on the number of CPUs available at the compute node.
Compute nodes at the CAC cluster had eight CPUs as of February 20th, 2012.
The set of multiple instances may share a common resource that needs computation
prior to the stage of <em>process-data</em>. Examples include genome indexing stages
for short read alignment procedures.</p></div>
<div class="listingblock">
<div class="title">Indexing genome files</div>
<div class="content">
<pre><tt>./bwa index -p $CBWADIR/$GENOMEFASTA-bwa -a is \\
  $CDATADIR/$GENOMEFASTA</tt></pre>
</div></div>
<div class="paragraph"><p>Let us turn to function <em>process-data</em>. We start with <tt>job-fastq-qc</tt> shell
script, which takes a three-digit number as an argument.</p></div>
<div class="listingblock">
<div class="title">Execute a quality control</div>
<div class="content">
<pre><tt>bash job-fastq-qc 056 FASTQ056.fq.gz FASTQ056.prinseq.fq.gz</tt></pre>
</div></div>
<div class="paragraph"><p>In the <tt>job-fastq-qc</tt> we take a raw fastq file to first remove adapter sequences
from the short reads in the fastq file, and to remove low quality 3' end parts.
In the <tt>job-bwa-align</tt> we take the processed short reads to align them on a
reference genome.</p></div>
<div class="listingblock">
<div class="title">Align short reads using bwa and subread</div>
<div class="content">
<pre><tt>bash job-bwa-align 056 FASTQ056.prinseq.fq.gz bwa/FASTQ056.sorted</tt></pre>
</div></div>
<div class="paragraph"><p>In the third stage of <tt>job-de</tt> we take the short read alignment files to count
short reads mapped on genes for summarizing numbers of reads see <tt>job-de</tt>,
<tt>job-de.R</tt>.</p></div>
<div class="listingblock">
<div class="title">Two pairs of input and output files of <tt>job-de</tt> stage</div>
<div class="content">
<pre><tt>output/ua159/1/bwa/FASTQ056.sorted.bam
output/ua159/1/bwa/FASTQ056.cl</tt></pre>
</div></div>
<div class="paragraph"><p>Counting mapped short reads may require memory exceeding the capacity of a
compute node. So, we divide alignments smaller pieces, for each of which we
count reads, and then sum the counts.</p></div>
</div>
</div>
<div class="sect2">
<h3 id="_species">species</h3>
<div class="paragraph"><p>Sample analysis files. This directory can be used for your analyses if <em>ROOTANALYSISDIR</em> variable in <tt>conf</tt> is set to the source code directory of <strong>RNAseq Analysis</strong>.</p></div>
</div>
<div class="sect2">
<h3 id="_src">src</h3>
<div class="paragraph"><p>C, or C++ source codes.</p></div>
</div>
<div class="sect2">
<h3 id="_ucsc">ucsc</h3>
<div class="paragraph"><p>Files for a UCSC genome browser.</p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_details_and_future_development">Details and Future Development</h2>
<div class="sectionbody">
<div class="paragraph"><p>You could read the details of implementation as a source code documentation. They may not make sense from the user perspective. Future developments include a pairwise comparison for determining differentially expressed genes, functional association of differentially expressed genes, transcript calls by applying a transcript caller such as ParseRNAseq, export of results (coverage) for showing them in UCSC genome tracks.</p></div>
<div class="sect2">
<h3 id="_useful_tools">Useful Tools</h3>
<div class="paragraph"><p>You could annotate bacterial genomes using xbase available at
<a href="http://www.xbase.ac.uk/annotation">http://www.xbase.ac.uk/annotation</a>.
You could compare two genomes to find genes unique to one of the two genome using genomeblast available at
<a href="http://bioinfo-srv1.awh.unomaha.edu/genomeblast/tutorial.php">http://bioinfo-srv1.awh.unomaha.edu/genomeblast/tutorial.php</a>. Because genbank files created by <tt>xbase</tt> were incompatible with the format required by genomeblast, we extract amino acid sequences from genbank file to find genes that are unique to one of the two genomes. A <tt>bioperl</tt> script, <tt>pl/extractcds.pl</tt> could be useful to extract protein sequences from a genbank file.</p></div>
</div>
<div class="sect2">
<h3 id="_batch3_sh">batch3.sh</h3>
<div class="paragraph"><p>Function <em>batch3-variable</em> creates often-used shell variables.
<em>ROOTANALYSISDIR</em> is set to a full file path by configuration.
<em>SPECIESFILE</em> is the species file name relative to the directory where a <tt>run</tt>
is executed. <em>OUTPUTDIR</em> is the output directory relative to <em>ROOTANALYSISDIR</em>.
<em>BASEDIR</em> is the species filenamed directory at <em>OUTPUTDIR</em>. <em>NUMBERDIR</em> is a
directory 1 (repetition number) at <em>BASEDIR</em>. <em>DATADIR</em> is placed at
<em>NUMBERDIR</em>. <em>BWADIR</em> is also at <em>NUMBERDIR</em>, and so is <em>RUNANALYSIS</em>. Consider
that <em>DATADIR</em> contains input files, <em>BWADIR</em> contains processed files, and
<em>RUNANALYSIS</em> contains output files. We have similar directory names in a remote
machine: <em>RBASEDIR</em>, <em>RNUMBERDIR</em>, <em>RDATADIR</em>, <em>RBWADIR</em>, and <em>RANALYSISDIR</em>.
We have similar directory names in a computing machine:
<em>CBASEDIR</em>, <em>CNUMBERDIR</em>, <em>CDATADIR</em>, <em>CBWADIR</em>, and <em>CANALYSISDIR</em>.</p></div>
<div class="paragraph"><p>Function <em>batch3-output</em> creates <em>DATADIR</em>, <em>BWADIR</em>, and <em>RUNANALYSIS</em>
directories.</p></div>
<div class="paragraph"><p>Function <em>batch3-speciesfile</em> creates shell variables set by <em>SPECIESFILE</em>.
<em>FASTQFILES</em> is a list of numbers separated by spaces. <em>FASTQLABEL</em> is a list of
sample names. <em>CACWORKDIR</em> is the directory at which we submit jobs.
<em>REFGENOMEFASTA</em> and <em>REFGENOMETXDB</em> are genbank and TranscriptDb files. They
should be at <em>ROOTANALYSISDIR</em>.  <em>QCALIGNDEWALLTIME</em> is the limit of execution
time in hours.  <em>MINMAPQ</em> is the minimum alignment mapping quality.
<em>MINTRIMQ</em> is the minimum base quality to trim 3' end.</p></div>
<div class="paragraph"><p>Function <em>batch3-push-data</em> help send files to the remote machine for job
submission by creating a shell script, <tt>push-data.sh</tt>. Execute the shell script
to send files for processing.</p></div>
<div class="paragraph"><p>Function <em>batch3-get-data</em> creates a shell script to get resulting files from
the remote machine. The files are <tt>count.txt</tt>, <tt>count.multiple.txt</tt>,
<tt>qualPlot.pdf</tt>, and <tt>stat1.tex</tt>.</p></div>
<div class="paragraph"><p>Function <em>batch3-run-qcalignde</em> creates most of shell and R scripts.</p></div>
<div class="paragraph"><p>File <tt>run-qcalignde.sh</tt> is the main script to be executed when submitting jobs
to the cluster. It replaces a string named PBSARRAYSIZE with the number of FASTQ
files in <tt>batch-qcalignde.sh</tt>, and submit the batch file.</p></div>
<div class="paragraph"><p>File <tt>batch-qcalignde.sh</tt> is the batch script that is submitted to the cluster.
A function called <tt>process-data</tt> in the batch script executes a series of jobs
for a FASTQ file. It first copies a FASTQ file from the remote data directory
called <em>RDATADIR</em>. We should find the FASTQ file in the remote data directory.
The file is copied to the parallel directory in the compute node, <em>CDATADIR</em>.
We use <tt>job-stat</tt> to process the FASTQ file.
We use <tt>job-fastqc</tt> to remove parts of or whole short reads with low quality.
We use <tt>job-bwa-align</tt> to align reads to a reference genome.
We use <tt>job-de</tt> to count reads aligned to the reference genome.</p></div>
<div class="paragraph"><p>Function <em>batch3-make-job</em> creates all of the job files in shell and R scripts.
Bash script <tt>job-stat</tt> takes a three-digit number, a gzipped FASTQ file name,
and an output RData file name.
Bash script <tt>job-fastqc</tt> takes a three-digit number, a gzipped FASTQ file name,
and another gzipped FASTQ file name (output).
Bash script <tt>job-bwa-align</tt> takes a three-digit number (FASTQ ID), an input
gzipped FASTQ file name, and an output BAM file base name without the extension
of bam.
Bash script <tt>job-de</tt> takes a sorted BAM file, the number of reads in the sorted
BAM file, and the number of reads in FASTQ file used as an input to bash script
<tt>job-bwa-align</tt>.
Bash scripts <tt>job-simulate</tt> and <tt>job-check</tt> create a small test data to check
the alignment and count procedure.</p></div>
<div class="paragraph"><p>File <tt>feature-txnc.txt</tt> contains R scripts to use the txdb file specified in a spceis file. This could be the key file for handling gene annotations. This file is added to R script files: <tt>job-de.R</tt>, <tt>job-de-sum.R</tt>, and <tt>job-simulate.R</tt>. We use R function <em>transcripts</em> to extract regions, on which we count reads mapped. R function <em>cds</em> extracts CDS regions. R package <em>GenomicFeatures</em> features the two functions. I need a simple example files for handling R scripts. They are easy to use, and also easy to forget. Let&#8217;s see what it does. We construct a transcript map with meta data. A transcript is of type being either "CDS" or "NOCDS." Regions not within the transcript map are of type being "NG." An R variable called <em>feature.txnc</em> contains transcripts with 6 columns: seqnames, ranges, strand, tx_id, tx_name, and type.</p></div>
<div class="paragraph"><p>R script file <tt>job-de.R</tt> takes takes a sorted BAM file, the number of reads in the sorted BAM file, and the number of reads in FASTQ file used as an input to bash script <tt>job-bwa-align</tt>. This is the same list of arguments as those of <tt>job-de</tt>. The first argument is used create the output file name by appending ".cl" to the sorted bam file. R package GenomicFeatures is used to extract annotation. The statistics of short reads are incomplete (FIXME). It bothers me that I did not have a complete information about statistics. R package GenomicRanges features summarizeOverlaps, which counts short reads mapped on amap of transcripts. <em>cl</em> contains counts. <em>cl.nc</em> is a subset of <em>cl</em> for nongenic regions. <em>cl.multiple</em> contains reads mapped on multiple places. <em>cl.multiple.nc</em> is a subset of nongenic regions of <em>cl.multiple</em>. <em>cl.multiple.nocds</em> is a subset of non-CDS regions of <em>cl.multiple</em>. The transcript map in txdb contains genes, non-genic "NG", and non-CDS "NOCDS". We have 4 values for the number of reads. <em>num.total.read</em> is the number of total reads in the FASTQ file. We align <em>num.n.read</em> reads on a reference genome. The bam file contains <em>num.mapped.read</em> reads. Among them <em>num.unique.read</em> reads are aligned with minimum quality score, <em>MINMAPQ</em>. <em>bv</em> is an R <em>GappedAlignments</em> object returned by R function <em>readBamGappedAlignments</em> of package <em>Rsamtools</em>. <em>bv</em> contains mapped reads with minimum of <em>MINMAPQ</em> mapping quality score. <em>bv.multiple</em> is an R <em>GappedAlignments</em> object that contains mapped reads with mapping quality score less than <em>MINMAPQ</em>. We report overlapping genes. &#8220;Now, I see a little bit what I did.&#8221;</p></div>
<div class="paragraph"><p>R function <em>makeTranscriptDb</em> requires a few R data frames in its arguments. An argument, transcripts, consists of transcript ID (integer), transcript name (character), chromosome name (character), strand ("+", "-"), start position (integer), end position (integer). A second argument, splicings, consists of transcript ID (integer A), exon ID (integer B), exon name (character), exon start and end (integers), cds start and end (integers). We create a txdb file based on the following rationale. An exon is a transcript. The exon or the transcript may contain a CDS. See the R script in <em>feature-txnc.txt</em> how a txdb file is used.</p></div>
<div class="paragraph"><p>Bash script <tt>job-de-sum</tt> calls its R script. It passes <em>RBWADIR</em> directory and a file that contains FASTQ IDs to the R script. R script file <tt>job-de-sum.R</tt> takes a BWADIR directory, and a file containing FASTQ IDs. Use the &#8220;cl&#8221; files for the FASTQ IDs to create a table of counts. We create several <em>count.table</em> R variables. <tt>job-de.R</tt> creates <em>cl</em>, <em>cl.cds</em>, <em>cl.nocds</em>, <em>cl.tx</em>, and <em>cl.ng</em>. For these count variables, we have <em>count.table</em>, <em>count.table.cds</em>, <em>count.table.nocds</em>, and the like. This creates a number of files: <em>count</em>, <em>count.cds</em>, <em>count.nocds</em>, <em>count.tx</em>, and <em>count.ng</em>. For multiple matched reads, we have <em>count.m</em>, <em>count.m.cds</em>, <em>count.m.nocds</em>, <em>count.m.tx</em>, and <em>count.m.ng</em>. R script <em>job-de-sum.R</em> also creates a table of statistics. Column <tt>a</tt> is the number of reads in the FASTQ file. Column <tt>b</tt> is the number of reads after quality control using PRINSEQ. Column <tt>c</tt> is the number of reads in the sorted BAM file. Column <tt>d</tt> is the number of counted reads multiply mapped on non-CDS transcripts. Column <tt>e</tt> is the number of reads uniquely mapped. Column <tt>e</tt> is the number of counted reads uniquely mapped on transcripts. Column <tt>f</tt> is the number of counted reads uniquely mapped non-transcript regions.</p></div>
<div class="paragraph"><p>Bash script <tt>job-simulate</tt> creates test FASTQ files using txdb annotation. R script <tt>job-simulate.R</tt> takes <em>FASTQNUM</em> as the only argument.  Because we execute <tt>job-simulate</tt> at the remote computer not at the compute node, we use <em>RBWADIR</em> and <em>RDATADIR</em> directories not <em>CBWADIR</em> or <em>CDATADIR</em>. We use <em>feature.txnc</em> or the complete feature to create short reads. We store sampled reads in <em>read.GA</em>. We sample reads from genes with 1,000 base pairs per chromosome. We use positions of reads to create a <em>GappedAlignments</em> object and add it to <em>read.GA</em>. We use the positions to extract DNA sequences, and write them to a FASTQ file. Ignore warnings, which arise when adding <em>GappedAlignments</em> with different chromosome configurations.</p></div>
</div>
<div class="sect2">
<h3 id="_cram">cram</h3>
<div class="paragraph"><p>Because short reads files tend to be large, it would be convenient to zip them
so that we can save storage and time of transfer. Although it is under
development, it seems to be a good idea to use a program called <tt>cram</tt>. Please,
make sure that you use the same version of the program in both zipping and
unzipping. You also need to make sure that you use the same reference genome.
Use the following command after executing <tt>batch3</tt> menu.</p></div>
<div class="listingblock">
<div class="title">Cram short reads at <em>CACWORKDIR</em></div>
<div class="content">
<pre><tt>$ bash run-cram.sh</tt></pre>
</div></div>
</div>
<div class="sect2">
<h3 id="_steps">steps</h3>
<div class="paragraph"><p>The following scripts can be run separately in the cluster. They need
documentations.
I used to have separate scripts for quality control, alignment, and counting
reads. Although this was fine, it would produce unnecessary intermediate files
that could be very large. Because accumulated large files could be too big for
the compute node to host, this might cause problems in the pipeline. The
followings are descriptions for the separate analyses. They can be useful in
understanding separate analyses that are integrated in <tt>run-qcalignde.sh</tt>.</p></div>
<div class="listingblock">
<div class="title">To align reads step-by-step</div>
<div class="content">
<pre><tt>$ bash run-fastqc.sh
$ bash run-cram.sh
$ bash run-cram2fastq.sh
$ bash run-bwa-align.sh
$ bash run-bias.sh
$ bash run-coverage.sh       &lt;- takes time
$ bash run-coverage-pull.sh
$ bash run-parsernaseq.sh
$ bash run-ucsc.sh
$ bash run-de.sh -&gt; not yet</tt></pre>
</div></div>
<div class="listingblock">
<div class="title">To create an input file for tux</div>
<div class="content">
<pre><tt>$ bash job-coverage-start-end.sh</tt></pre>
</div></div>
<div class="paragraph"><p>job files are supposed to be run in the compute node. Those with sh extension
would be run in the main node.</p></div>
<div class="listingblock">
<div class="title">To count reads</div>
<div class="content">
<pre><tt>$ bash run-qcalignde.sh</tt></pre>
</div></div>
<div class="paragraph"><p><tt>run-qcalignde.sh</tt> includes fastqc, bwa-align, and de.</p></div>
<div class="paragraph"><p><tt>run-fastqc.sh</tt> converts <tt>data/FASTQ001.fq.gz</tt> to <tt>bwa/FASTQ001.prinseq.fq.gz</tt>.</p></div>
<div class="paragraph"><p><tt>run-cram.sh</tt> converts <tt>bwa/FASTQ001.prinseq.fq.gz</tt> to <tt>bwa/FASTQ001.cram</tt>.</p></div>
<div class="paragraph"><p><tt>run-cram2fastq.sh</tt> converts <tt>bwa/FASTQ001.cram</tt> to <tt>bwa/FASTQ001.recovered.fq.gz</tt>.</p></div>
<div class="paragraph"><p><tt>job-stat</tt> converts <tt>bwa/FASTQ001.fq.gz</tt> to <tt>bwa/FASTQ001.fq.qualPlot.RData</tt>.</p></div>
<div class="paragraph"><p><tt>run-bwa-align.sh</tt> converts <tt>bwa/FASTQ001.recovered.fq.gz</tt> to <tt>bwa/FASTQ001.sorted.bam</tt>.</p></div>
<div class="paragraph"><p><tt>run-bias.sh</tt> converts <tt>bwa/FASTQ001.sorted.bam</tt> to <tt>bwa/FASTQ001.yml</tt>.</p></div>
<div class="paragraph"><p><tt>run-coverage.sh</tt> converts <tt>bwa/FASTQ001.sorted.bam</tt> and <tt>bwa/FASTQ001.yml</tt> to
<tt>bwa/FASTQ001.cvg.*</tt>.</p></div>
<div class="paragraph"><p><tt>run-coverage-pull.sh</tt> converts <tt>bwa/FASTQ001.cvg.*</tt> to <tt>bwa/FASTQ001.coverage.RData</tt>.
and <tt>bwa/FASTQ001.wig</tt>.</p></div>
<div class="paragraph"><p><tt>run-fastqc.sh</tt> can use multiple CPUs by splitting FASTQ files, each of which is
processed by cutadapt and prinseq.</p></div>
<div class="listingblock">
<div class="title">To filter out the raw fastq files for quality control</div>
<div class="content">
<pre><tt>bash job-fastqc 001 FASTQ001.fq.gz FASTQ001.prinseq.fq.gz</tt></pre>
</div></div>
<div class="listingblock">
<div class="title">To align short reads on a reference genome</div>
<div class="content">
<pre><tt>bash job-bwa-align 001 FASTQ001.prinseq.fq.gz FASTQ001.sorted</tt></pre>
</div></div>
<div class="paragraph"><p><tt>run-coverage.sh</tt> can use multiple nodes. BAM files are copied to each node
computer and copied again because readGappedAlignments(bamFile) R function
crashes when it tries to read a file from many computers.</p></div>
<div class="paragraph"><p><tt>run-parsernaseq.sh</tt> creates <tt>bwa/FASTQ001.parsernaseq</tt>, <tt>bwa/FASTQ001.bed2</tt> and
<tt>bwa/FASTQ001.operon</tt>.
File <tt>bwa/FASTQ001.bed2</tt> is the ParseRNAseq output file in BED format.
File <tt>bwa/FASTQ001.operon</tt> is another BED file of the transcript calls with
strand information inferred from genes annotations.</p></div>
<div class="paragraph"><p><tt>run-ucsc.sh</tt> creates scripts and files to upload the genome browser.</p></div>
<div class="listingblock">
<div class="content">
<pre><tt>$ bash job-ucsc-bed</tt></pre>
</div></div>
<div class="paragraph"><p><tt>run-de.sh</tt> converts <tt>bwa/FASTQ001.sorted.bam</tt> to <tt>bwa/FASTQ001.sorted.bam.cl</tt>.</p></div>
<div class="paragraph"><p>Does <tt>run-de.sh</tt> do this? We do not have that.</p></div>
<div class="paragraph"><p>Randomly sample about 5 percent of the short reads.
awk <em>BEGIN {srand()} !/^$/ {if (rand() &#8656; .05 &amp;&amp; NR%4==0) for ( o = 1; o &#8656; 4; o++ ) {getline; print}}</em></p></div>
<div class="listingblock">
<div class="title">Sample 0.1 percent of the total reads</div>
<div class="content">
<pre><tt>fastq-sample.sh FASTQ056.fq.gz FASTQ056-0.001.fq.gz 0.001</tt></pre>
</div></div>
</div>
</div>
</div>
</div>
<div id="footnotes"><hr /></div>
<div id="footer">
<div id="footer-text">
Version 0.1.7<br />
Last updated 2012-09-03 10:41:38 EDT
</div>
</div>
</body>
</html>
