\documentclass{article}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{color} 
\usepackage{setspace} 
\usepackage{rotating}
\usepackage{natbib}
\usepackage{xr}
\usepackage{../../latex/sty/genres}

% Text layout
\oddsidemargin 0in
\evensidemargin 0in
\topmargin -.5in
\textwidth 6.5in
\textheight 9in
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}


\newcommand{\Rpackage}[1]{R package \textit{#1}}
\newcommand{\program}[1]{{\texttt{#1}}}
\newcommand{\fixme}[1]{{\textbf{Fixme:} \textit{\textcolor{blue}{#1}}}}

\makeatother
\pagestyle{myheadings}
\markright{Transcriptome of Streptococcus mutans}

\usepackage{Sweave}
\begin{document}

\begin{titlepage}

\title{Defining Transcriptome using RNA Sequencing\\
of {\em Streptococcus mutans} }

\author{Sang Chul Choi$^{1}$, Charles D.\ Danko$^{1}$, 
Melissa J.\ Hubisz$^{1}$, \\
Robert A.\ Burne$^{3}$
Michael J.\ Stanhope$^{2}$, and Adam Siepel$^{1}$}

\date{ }
\maketitle

\begin{footnotesize}
\begin{center}
$^1$Department of Biological Statistics and Computational Biology,\\
Cornell University, Ithaca, NY 14853, USA
\\[1ex]
$^2$Department of Population Medicine and Diagnostic Sciences,\\
College of Veterinary Medicine, Cornell University, Ithaca, NY 14853, USA
\\[1ex]
$^3$Department of Oral Biology,\\
College of Dentistry, University of Florida, Gainesville, Florida 32610 USA
\\
\end{center}
\end{footnotesize}

\vspace{1in}

\begin{tabular}{lp{4.5in}}
{\bf Submission type:}& Research Article
\vspace{1ex}\\
{\bf Keywords:}&Carbon catabolite repression, RNA-seq, non-coding RNAs
\vspace{1ex}\\
{\bf Running Head:}&Transcriptome of {\em Streptococcus mutans}
\vspace{1ex}\\ 
{\bf Corresponding Author:}&
\begin{minipage}[t]{4in}
 Adam Siepel\\
 102E Weill Hall, Cornell University\\
 Ithaca, NY 14853\\
 Phone: +1-607-254-1157\\
 Fax: +1-607-255-4698\\
 Email: acs4@cornell.edu
\end{minipage}
\end{tabular}

\thispagestyle{empty}
\end{titlepage}

\doublespacing

\section{Abstract}
A bacterial transcriptome from the primary etiological agent of human dental
caries, \textit{Stretptococcus mutans}, is described using deep RNA sequencing
data.  We define transcripts boundaries of open reading frames based on sequence
alignments of short reads on the reference genome of UA159 from the deep
sequencing.  Based on homologous sequences and RNA secondary structures we also
report putative small RNAs, several of which are confirmed experimentally using
both RNA sequencing data, northern blot, and qRT-PCR.  CcpA is known to play a
direct role in carbon catabolite repression (CCR).  We use the developed
transcript map, open read frame, and small RNAs to determine differentially
expressed regions between the wild-type (UA159) and CcpA-deletion (TW1) strains
in order to refine the previously established importance of CcpA regulation as
central metabolism and virulence gene expression.  We find that more genes are
differentially expressed in the comparison between wild-type and CcpA-deletion
strains that are grown in galactose than in glucose.  We also find that more
genes differentially expressed in wild-type than in CcpA-deletion strain when
comparing glucose- and galactose-grown cells.  These two findings are different
from those from microarry experiment of \cite{Abranches2008}.  We also predicted
target genes of the small RNAs that were enriched in the following functional
categories: xxx.  We confirme that CcpA modulates the pathogenic potential of
\textit{S.\ mutans} through global control of gene expression.

\section{Abbreviations}

\begin{enumerate}
\item CCR Carbon catabolite repression
\item CcpA ?
\item IPS intracellular polysaccharide
\item CRE catabolite responsive elements
\end{enumerate}

\section{Introduction}
Regulation of carbon catabolite repression (CCR) in \textit{Streptococcus
mutans} can involve a phosphoenolpyruvate:sugar phosphotransferase system (PTS),
histidin protein (HPr) and a transcriptional regulator, CcpA \citep{Gorke2008}.
CcpA was shown to regulate metabolism and virulence gene expression in
\textit{S.\ mutans} by \citet{Abranches2008}. They used microarray experiments
to measure expression levels of genes. Microarray experiments tend to be
qualitative. The advent of high throughput sequencing technology motivated us to
revisit the experiment to refine the result.  The deep RNA sequencing data also
allowed us to define the bacterial transcriptome of \textit{S.\ mutans}. So, we
could compare putative operon expression levels as well as gene expression
levels between the wild-type (UA159) and its mutant-type (TW1) whose
\textit{ccpA} was deleted from the wild-type. This allowed us to profile
expression of genes and putative operons affected by the regulation of
\textit{ccpA}. 

\citet{Martin2010} developed a hidden Markov model to predict transcripts by
using site-wise expression levels. I also refined the prediction of transcripts
using multiple samples. I measured expression levels of known genes and compared
them from one experiment condition and from another. I can test whether two
genes are in the same operon or not. I assume that expression levels of two genes 
vary in similar fashion if the two genes belong to the same operon. What I mean
by similar fashion is whether they are expressed differentially with similar
fold changes. I assume that operons maps do not change between the two
experiments, or multiple experiments. This is not true. I hope that additional
samples could help to define operons. Eventually, operon maps will be determined
experimentally. Computational tools could be helpful before those ages come to
us. \fixme{This idea did not work. We should only use the prediction method.}

Although our RNA-seq data did not sampled specifically non-coding small RNAs, we
retained small RNAs in the sample by attempting to deplete only ribosomal RNAs
from the sample. Because non-coding small RNAs and coding messenger RNAs were
mixed in the sample, we sought a way of predicting small RNAs using
\program{RNAz} developed by \citet{Gruber2010}.  \program{RNAz} used homologous
sequences and RNA secondary structures to predict putative non-coding RNAs. We
confirmed the several predicted small non-coding RNAs experimentally
\citep{Altuvia2007} including the RNA-seq data set. Because small RNAs could
be associated with CCR regulation of genes (CITE), we determined differentially expressed small
RNAs between RNA samples. We found a few small RNAs to be differentially
expressed.

Bacteria can selectively use substrates from different carbon sources. In the
presence of a preferred carbon source, catabolic system of secondary substrates
are often regulated or shutdown: called carbon catabolite repression (CCR)
\cite{Gorke2008}. CCR also regulates the expression of virulence factors in many
pathogenic bacteria. A classical example is the glucose-lactose diauxie in
\textit{E.\ coli}. A significant number of genes can be subject to CCR.  CCR in
bacteria involves regulatory networks that activate or silence genes in response
to carbon source and availability.  \textit{S.\ mutans} possesses redundant
systems for CCR.  We could measure directly the expression levels of
\textit{fruA} in \textit{S.\ mutans}.  We assess the contribution of CcpA to CCR
of predicted operons.  Clearly, deletion of \textit{ccpA} resulted in
substantial relief of the CCR of \textit{fruA} expression exerted by both
glucose and fructose.

We aims at refining the results of \cite{Abranches2008} in more quantative ways.
Find CCR-sensitive transcripts. We describe predictions made for transcripts and
small RNAs based on RNA-seq data. Combining the predictions and known genes we
determine differentially expressed regions in UA159 genome. 

We need to quantify some uncertainty of transcript predictions.

\section{Methods and Materials}

\subsection{Bacterial strains and growth conditions}
\fixme{Burne group should be able to describe this part. The following is just
from somewhere for the purpose of writing template.}
\textit{S.\ mutans} UA159 is a human cariogenic isolate. The whole genome is
availabe at NCBI \cite{Ajdic2002}. Strain TW1 was derived from UA159 by
replacing a part of \textit{ccpA} gene \cite{Wen2002}.  \textit{S.\ mutans}
strains were inoculated into TV medium supplemented with 0.5\% galactose or
glucose and grown to early exponential phase (OD600, 0.2).  Then fructose was
added to the medium at concentrations (0,0.05,3,5,10 mM) for 3.5h to induce
\textit{fruA} expression.

\subsection{RNA isolation}
\fixme{Burne group should be able to describe this part. The following is just
from somewhere for the purpose of writing template.}
RNA was isolated from S. mutans (1) by using 50 ml cultures that were grown
under the desired conditions and harvested by centrifugation at $4^{\circ}$C. Pelleted
cells were resuspended in 400 ul of diethyl pyrocarbonate-treated water, 800 ul
of RNAprotect reagent (Qiagen, Inc., Chatsworth, CA) was added, and the samples
were incubated at room temperature for 5 min with vortexing for 10 s at 1-min
intervals. Cells were then pelleted, resuspended in 250 ul Tris-EDTA (50:10)
buffer, and transferred to 1.5 ml screw-cap tubes containing sterile glass beads
(average diameter, 0.1 mm; Biospec, Bartlesville, OK), 10 ul of 1\% sodium
dodecyl sulfate, and 300 ul of acid phenol-chloroform (5:1). After
centrifugation for 15 min at maximum speed at $4^{\circ}$C, RNA was further purified
using an RNeasy mini kit (Qiagen, Inc., Chatsworth, CA), including on-column
DNase digestion with RNase-free DNase (Qiagen), as recommended by the supplier.

\subsection{RNA sequencing}
\fixme{Stanhope group should be able to describe this part. The following is just
from somewhere for the purpose of writing template.}
Sequencing libraries for the Illumina GA platform were constructed by shearing
the enriched cDNA by nebulisation (35psi, 6 min) followed by end-repair with
Klenow polymerase, T4 DNA polymerase and T4 polynucleotide kinase (to blunt-end
the DNA fragments). A single 3' adenosine moiety was added to the cDNA using
Klenow exo- and dATP. The Illumina adapters (containing primer sites for
sequencing and flowcell surface annealing) were ligated onto the repaired ends
on the cDNA and gel-electrophoresis was used to separate library DNA fragments
from unligated adapters by selecting cDNA fragments between 200-250 bps in size.
Fragmentation followed by gel-electrophoresis were used to separate library DNA
fragments and size fragments were recovered following gel extraction at room
temperature to ensure representation of AT rich sequences. Ligated cDNA
fragments were recovered following gel extraction at room temperature to ensure
representation of AT rich sequences. Libraries were amplified by 18 cycles of
PCR with Phusion polymerase. Sequencing libraries were denatured with sodium
hydroxide and diluted to 3.5 pM in hybridisation buffer for loading onto a
single lane of an Illumina GA flowcell. Cluster formation, primer hybridisation
and single-end, 36 cycle sequencing were performed using proprietary reagents
according to manufacturers' recommended protocol (https://icom.illumina.com/).
The efficacy of each stage of library construction was ascertained in a quality
control step that involved measuring the adapter-cDNA on a Agilent DNA 1000
chip. A final dilution of 2 nM of the library was loaded onto the sequencing
machine.

SOLiD sequencing was performed at Agencourt Bioscience (Beverly, MA) and
SeqWright (Houston, TX). Library preparations, fragment library protocols, and
all SOLiD-run parameters followed standard Applied Biosystems protocols.
Illumina/Solexa sequencing followed standard Illumina/Solexa methods and
protocols. Both systems were used because in this study our aim was simply to
collect as much sequence coverage as possible, and in pursuing this goal, we
utilized all available sequencing systems. We note, however, that the relative
performances (overall coverage, data output, etc.) noted in Table 1 for the
Illumina and Applied Biosystems sequencing platforms cannot be compared
directly. This study was not intended or designed to be a direct comparison of
the two systems, and interpreting the differences shown in Table 1 is difficult
given the many differences in cDNA/library preparation and the rapid evolution
of each system over the past year.

\subsection{Statistical analysis for differential expressions}
We used an R package, DESeq, \cite{Anders2010} to determine differentially
expressed genes based on negative binomial model originally developed by \cite{Robinson2007}.
Steps of analyzing RNA-seq data for determining differentially expressed genes
are described in \cite{Oshlack2010}. We used a short read aligner called
\program{BWA} \cite{Li2009a} to map the sequenced short reads on a reference genome,
\textit{Streptococcus mutans} UA159 (NCBI accession NC\_004350). Because the
aligner allowed a few gaps for efficient alignment of millions of sequences of
size being 100 base pairs long, short reads that consist of RNAs and sequencing
adapter sequences would be not mapped. We used \program{cutadapt} \cite{Martin2011} to
remove parts of adapter sequences from short reads before mapping on the
reference genome. Mapped short read alignments were converted into readable
formats \cite{Li2009b}. We counted short reads aligned to annotated genes in the
reference genome, forming a table of read counts of the genes. Statistical
software R \cite{R2011} with \Rpackage{DEseq} \cite{Anders2010}
was subsequently employed to infer differentially expressed in different
biological conditions.  We also generated tracks for our recently released
Streptococcus Genome Browser (http://strep-genome.bscb.cornell.edu) that
summarize gene expression levels based the short read alignments. This allowed
us to visualize parts of the reference genome for comparing different RNA
samples. Further analyses such as functional category enrichment test
\cite{Young2010} were also employed.

\subsection{Transcript predictions}
\citet{Martin2010} developed a hidden Markov model to predict transcripts by
using site-wise expression levels. We used the following options for the
program: ``-c 10 -b 25 -force\_gp''. We binned expression levels into 25 parts.
We allowed 10 emission states. We also used the genes as the predicted parts of
transcripts. 
Because we did not have strand-specific RNA-seq of the 13 samples due to
limited resources, we could not tell in which directions transcripts based on
RNA-seq data were oriented.

\subsection{small RNAs}
RNAz is used to find small RNA candidates. Sequence alignments were a critical
step for finding small RNAs in the approach. Phylogenies of streptococcus
species could be considered to choose species related with Streptococcus mutans.
Alternatively, I find sequences similar to intergenic regions, and refine the
alignments using a multiple sequence alignment tool such as MUSCLE
\cite{Edgar2004a}. Steps for achieving this goal include: 1. Exract intergenic
regions including up- and down-stream, 2. BLAST the intergenic regions against
non-redundant DB to find similar sequences \cite{Altschul1990}, 3. Refine each
alignment using MUSCLE \cite{Edgar2004a}, 4. Apply RNAz to the alignments for
scoring regions for small RNAs \cite{Washietl2005,Gruber2010}, and 5. Find
transcripts with high score of small RNAs. The transcripts must be defined in
advance.  We used RNAplex \citep{Tafer2008} to predict targets of putative small
RNAs. The method used RNAplfold \citep{Bernhart2006}.  Please, find a review
paper about small RNA predictions and targets by \citet{Tafer2008a}.

\subsection{Gene category associations and virulence genes}
We used two approaches for finding functional categories
\citep{Young2010,Suzuki2011}. We briefly descirbe We briefly descirbe the
prodecure of \citet{Suzuki2011}.  We assigned genes to Gene Ontology (GO)
categories by comparing the {\em Streptococcus} genes to bacterial proteins from
the Uniref90 database using BLASTP, and then assigning the same GO
classification as the target gene of the uniProt GOA database if the match had
an $E$-value of $<1.0\times10^{-5}.$  A gene family was assigned a given
classification if any of its genes was assigned that classification.  To test
for associations with replacing gene transfers we used the recombination
intensity estimated for each gene, and for additive gene transfers we used the
number of additive transfers inferred for each family.  To test for
significance, we performed a Mann-Whitney $U$ test of the values (recombination
intensities or numbers of transfers) associated with a given category vs.\ the
values for the all other genes/families.  We used the \cite{Benjamini1995}
method to correct for multiple comparisons.


\section{Results}

Findings include:
\begin{enumerate}
\item New genes, open reading frames that are not annotated in NCBI's genome database.
\end{enumerate}

We sequenced a whole prokaryotic transcriptome of \textit{S\. mutans}, and
mapped short reads to the genomic sequence of the reference genome UA159
\citep{Ajdic2002}. 
\citet{Martin2010} applied a hidden Markov model to use short reads mapped to a
genomic sequence in order to parase \textit{B.\ anthracis} genome into a
sequence of transcripts. Most of annotated genes belong to the predicted
transcripts. We found discrepancies of annotated genes and their transcriptions:
1. A gene contains two transcripts, 
2. Non-coding regions contain transcripts.
\fixme{We could find more discrepancies. Try to find them}. 
Expression levels of adjacent genes within a transcript should be more
correlated to each other than those of adjacent genes belonging to separate transcripts
\citep{Martin2010}. 

We identified a set of transcripts, albeit not a full, including small RNAs, and novel
transcripts. 
We hope that our study could serve as a foundation for a comprehensive study of
the \textit{S.\ mutans} transcriptome.
Our current knowledge of the transcriptome is limited to annotated or predicted genes.

Whole transcriptome sequencing using high throughput sequencing (HTS)
technologies, or deep RNA sequencing (RNA-seq) allowed to reveal the
transcriptome topography of \textit{S.\ mutans}. 

\subsection{From Reviews}
We reconstructed the full-length transcripts by transcriptome assembly using
xxx.

We tried in vain to use short-read assemblers such as Velvet \citep{Zerbino2008} that had been developed to
assembly transcriptome.
Trinity

gene-dense transcriptomes

We ensured a high-quality transcriptome assembly. 

In the data generation phase (FIG. 1a), total RNAs or mRNAs are fragmented and
converted into a library of cDNAs containing sequencing adaptors.  
In the data analysis phase (FIG. 1b), these short reads are pre-processed to
remove sequencing errors and other artefacts. The reads are subsequently
assembled to reconstruct the original RNAs and to assess their abun- dance
(‘expression counting’).

To increase the number of assem- bled transcripts, especially the less abundant
ones, ribosomal RNA (rRNA) and abundant transcripts are removed during the
first steps of library construction.
In order to retain RNAs without a poly(A) tail in the assem- bled transcriptome,
rRNA contamination can instead be removed by hybridization-based depletion
methods. \fixme{What depletion method did we use?} 
Last, the use of strand-specific RNA-seq protocols aids in the assembly and
quantification of overlapping transcripts that are derived from opposite strands
of the genome. This consideration is especially important for gene-dense
genomes, such as those of bacteria, archaea and lower eukaryotes.

Sequencing. The major factors to consider before sequencing a sample are: the
choice of sequencing plat- form, the sequencing read length and whether to use a
paired-end protocol.

Removing artefacts from RNA- seq data sets before assembly improves the read
quality, which, in turn, improves the accuracy and computa- tional efficiency of
the assembly. This step is straight- forward and can be executed using several
tools.  In general, three types of artefacts should be removed from raw RNA-seq
data: sequencing adaptors, which origi- nate from failed or short DNA
insertions during library preparation; low-complexity reads; and
near-identical reads that are derived from PCR amplification. 
Sequencing errors in NGS reads can be removed or corrected by analysing the
quality score and/or the k-mer frequency.

Reference-based transcriptome assembly is easier to perform for the simple
transcriptomes of bac- terial, archaeal and lower eukaryotic organisms, as these
organisms have few introns and little alternative splicing. Transcription
boundaries can be inferred from regions of contiguous read coverage in the
genome even with- out graph construction and traversal37,51,52. Alternative
transcription start and stop sites can also be inferred based on the 5′ cap or
poly(A) signals (if cap- or end- specific experimental protocols are used)51,53.
However, complications arise owing to the gene-dense nature of
these genomes. Many genes overlap, resulting in adja- cent genes being assembled
into one transcript, even though they are not from a polycistronic RNA. Strand-
specific RNA-seq has successfully been used to separate adjacent overlapping
genes from opposite strands in the genome51,52. Overlapping genes that are
transcribed from the same strand and that also have comparable expres- sion
levels cannot easily be separated without using cap- or end-specific RNA-seq.

In summary, reference-based assembly is generally preferable for cases in which
a high-quality reference genome already exists.

Meanwhile, experimental RNA-seq and sequencing protocols are continually
improving and should greatly reduce the informatics challenges.  RNA-seq reads
from third-generation sequencers, such as PacBio70, are longer (up to several
kilobases). PacBio sequencers are capable of sequencing a single transcript to
its full length in a single read. If this technology reaches a throughput that
is comparable to the second-generation technologies, then the need for
transcriptome assembly will probably be eliminated. Hopefully, the future of
transcriptome assembly will be ``no assembly required''.

What is the sequencing depth?

\subsection{My Results}

CcpA affects CCR of \textit{fruA}. 
\textit{fruA} fructan hydrolase
When Wen2002 deleted CRE sequencs in the \textit{fruA} promoter region, CCR was
almost completely alleviated. 
What are CCR-sensitive genes? 
What are the mechanism of CcpA on CCR of the \textit{fruA} operon? 
Expression levels of \textit{fruA} in \textit{S.\ mutans} strains that were
grown to early exponential phase (OD600, 0.2) in TV medium supplemented with
0.5\% glucose or galactose. \fixme{What does this mean?} 
Deletion of \textit{ccpA} resulted in substantial relief of the CCR of
\textit{fruA} expression exerted by both glucose and fructose. 
CcpA could directly interact with the \textit{fruA} promoter region. 

The role of CcpA in gene expression during growth on glucose. 
\textit{S.\ mutans} UA159 and TW1 grown under catabolite-repressing conditions
(glucose), and catabolite-derepressing conditions (galactose). Under
glucose-only condition I found xx number of differentially expressed genes. What
are the functional categories of those differentially expressed genes? What are
the genes that are most profoundly up- or down-regulated (Smu.1421 to Smu.1424,
PDH enzyme complex), ... and more? ... This regulatory control was
short-circuited in the strain growing on glucose and lacking CcpA. I showed that
some of KEGG pathways were affected by TW1 grown under glucose. We found
down-regulated genes, which would have something to do with pathogenicity. 
Upto here, I compared UA159 and TW1 under glucose.

1. UA159 (A) vs. TW1 (B) : fold change is B/A.
When glucose was the sole carbohydrate for growth, 170 genes were differentially
expressed at least twofold as a result of loss of CcpA (Fig. 2; see Table S1 in
the supplemental material). Of the gene products, 22 were predicted to partici
pate in energy metabolism, 9 were PTS components, and 43 were hypothetical
proteins (Fig. 2). Among the most profoundly up-regulated genes in the ccpA
mutant strain were the genes that encoded the components of the PDH enzyme com-
plex (Smu.1421 to Smu.1424), pyruvate-formate lyase, enzymes in the
tricarboxylic acid (TCA) cycle, selected sugar-specific PTS permeases, and
components of the glycogen biosynthetic machinery (see Table S1 in the
supplemental material). 
Consistent with the profile data, TW1 cells grown in glucose had 78\% higher
PDH activity than the wild-type strain (data not shown). Given the impact on
expression of PDH, TCA enzymes, and selected glycolytic enzymes, CcpA clearly
legislates gene expression to discriminate between carbohydrate-replete and
carbohydrate-limiting conditions. In particular, when there is excess
carbohydrate, S. mutans moves carbohydrates primar- ily through lactate
dehydrogenase, sacrificing an ATP and generating NAD When carbohydrate is
limiting, S. mutans pro- duces mainly formic and acetic acids, shunts
carbohydrate through pyruvate formate lyase, and uses its partial TCA cycle to
regenerate reducing equivalents to maintain an appropriate NAD/NADH balance.
This regulatory control was clearly short-circuited in the strain growing on
glucose and lacking CcpA.
Among the more interesting genes that were down-regu- lated, significant
decreases in expression were noted for the genes encoding components of the
proton-translocating F-ATPase, phosphoglycerate kinase (Pgk), and
fructosyltrans- ferase (ftf), which catalyzes the conversion of sucrose to high-
molecular-weight polymers of fructose, as well as gtfD, which encodes the
glucosyltransferase D enzyme that produces predominantly -1,6-linked polymers
of glucose from sucrose. Since acid tolerance (ATPase), acid production from
carbohy- drate (Pgk), and glucan and fructan production (Gtf and Ftf), as well
as the ability to transport carbohydrates and produce intracellular storage
compounds (glycogen), as noted above (PTS and Glg), are central to the
pathogenic potential of S. mutans, CcpA represents a major control point for
expression of critical virulence attributes in cells presented with glucose.

Growth in derepressing conditions induces major changes in the UA159
transcriptome. Here, I compared glucose and galactose conditions for growing
UA159. Energy metabolism was enriched with differentially expressed genes. There
were more up-regulated genes in galactose condition compared with glucose
condition. Some genes are described in detail. 

2. I compared glucose (A) and galactose (B) for growing UA159
It is noteworthy that only three genes in the energy metabolism category were
down-regulated, whereas 23 genes in this category were markedly up-regulated. In
addition, the expression of 28 genes encoding hypothetical proteins was affected,
and 19 of them were up-regulated. The genes encod- ing proteins belonging to the
pathways for lactose and galac- tose utilization (2) were the most strongly
induced genes in cells growing on galactose. The galKTE genes, encoding the
enzymes for galac- tose utilization via the Leloir pathway (2), were
up-regulated between 10- and 36-fold, but the genes of the lactose (lac) operon,
encoding the lactose PTS permease and the tagatose- 6-phosphate pathway, were
up-regulated 70- to 300-fold. Since the products of the galKTE operon contribute
to the genera- tion of UDP-galactose, which is required for cell wall biogen-
esis, the lower level of induction of galKTE than of the lac operon may reflect
a need for higher constitutive levels of expression of the Leloir pathway in the
absence of galactose. Components of the multiple-sugar metabolism operon (msm)
were also induced in cells growing on galactose. The Msm pathway is induced by
raffinose and melibiose, both of which are ?-galactosides (5). The enhanced
expression of gene products involved in en- ergy metabolism, including pyruvate
formate lyase and com- ponents of the TCA cycle, in cells growing on galactose
is consistent with the idea that the cells perceive carbohydrate limitation. 

Now, I compared glucose (A) and galactose (B) for growing TW1. 

Now, I compared UA159 and TW1 under galactose. 

What genes are differentially expressed? 

Name the transcripts.
I found xx transcripts. Among them xx \% were mono-cistronic operons. xx \% were
poly-cistronic operons. xx transcripts could not defined because RNA-seq short
reads were not precise. Transripts were assigned with functional categories, and
differentially expressed transcripts were found. Transcripts and functional
categories: how can I assess that transcripts consist of genes with similar
functional cateogires. 

Name the small RNAs.
RNAz predicted a few small RNAs. Some of them were differentially expressed.
Some functional categories were enriched with target genes of the predicted
genes. 

\section{Discussion}
We used RNA-seq data to delineate transcript boundaries. Although this map of
transcripts does not precisely tell us where transcripts start and end, it
informs us of which open reading frames can be transcribed together. Potentially
polycistronic operons are defined. We focused on interpretations of operon in
light of CcpA regulations. 


\renewcommand*{\refname}{Literature Cited}
\bibliographystyle{../../latex/bst/mbe}
\bibliography{siepel-mutans}



\begin{figure}
\centering
\includegraphics[width=.9\textwidth]{manuscript-figClust}
\caption{Heatmap showing the Euclidean distances between the samples
  as calculated from the variance-stabilizing transformation of the
  count data.}
\label{figClust}
\end{figure}

\end{document}
