\documentclass{article}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{color} 
\usepackage{setspace} 
\usepackage{rotating}
\usepackage{natbib}
\usepackage{xr}
\usepackage{url}
\usepackage{../../latex/sty/genres}
\externaldocument{supplementary}

% Text layout
\oddsidemargin 0in
\evensidemargin 0in
\topmargin -.5in
\textwidth 6.5in
\textheight 9in
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

\SweaveOpts{keep.source=TRUE,eps=FALSE,include=FALSE,width=6,height=6}
\newcommand{\Rpackage}[1]{R package \textit{#1}}
\newcommand{\program}[1]{{\texttt{#1}}}
\newcommand{\fixme}[1]{{\textbf{Fixme:} \textit{\textcolor{blue}{#1}}}}
\newcommand{\comment}[1]{}

\makeatother
\pagestyle{myheadings}
\markright{Transcriptome of Streptococcus mutans}

\begin{document}

\begin{titlepage}

\title{Defining Transcriptome using RNA Sequencing\\
of {\em Streptococcus mutans} }

\author{Sang Chul Choi$^{1}$, Charles D.\ Danko$^{1}$, 
Melissa J.\ Hubisz$^{1}$, \\
Paulina Dmitrievna Pavinski Bitar$^{2}$, \\
\fixme{Add more coauthors, or change the order of authors if needed} \\
Robert A.\ Burne$^{3}$
Michael J.\ Stanhope$^{2}$, and Adam Siepel$^{1}$}


\date{ }
\maketitle


\begin{footnotesize}
\begin{center}
$^1$Department of Biological Statistics and Computational Biology,\\
Cornell University, Ithaca, NY 14853, USA
\\[1ex]
$^2$Department of Population Medicine and Diagnostic Sciences,\\
College of Veterinary Medicine, Cornell University, Ithaca, NY 14853, USA
\\[1ex]
$^3$Department of Oral Biology,\\
College of Dentistry, University of Florida, Gainesville, Florida 32610 USA
\\
\end{center}
\end{footnotesize}

\vspace{1in}

\fixme{Add more corresponding authors.}

\begin{tabular}{lp{4.5in}}
{\bf Submission type:}& Research Article
\vspace{1ex}\\
{\bf Keywords:}&Carbon catabolite repression, RNA-seq, non-coding RNAs
\vspace{1ex}\\
{\bf Running Head:}&Transcriptome of {\em Streptococcus mutans}
\vspace{1ex}\\ 
{\bf Corresponding Author:}&
\begin{minipage}[t]{4in}
 Adam Siepel\\
 102E Weill Hall, Cornell University\\
 Ithaca, NY 14853\\
 Phone: +1-607-254-1157\\
 Fax: +1-607-255-4698\\
 Email: acs4@cornell.edu
\end{minipage}
\end{tabular}

\thispagestyle{empty}
\end{titlepage}

\doublespacing

\section*{Abstract}
A bacterial transcriptome of the primary etiological agent of human dental
caries, \textit{Streptococcus mutans}, is described using deep RNA sequencing.
RNA transcripts inferred from the RNA-seq data revealed novel messenger RNAs,
and small RNAs. Differential expression studies on the role of CcpA in 
carbon catabolite repression (CCR) suggested that some of the novel mRNAs and
small RNAs found in our study could be involved in the regulation of CcpA. Our
study provides more precise landscape of transcriptome changes in response to
different carbon sources.

\section*{Introduction}
% Biological motivations: S. mutans, CCR and CcpA
About one fifth of children of ages ranging from 2 to 19 were reported to have
untreated cavities in the United States \citep{health2010}. Oral bacteria digest
foods containing carbohydrates left on the teeth to convert them to acids, which
dissolve the surface of the teeth and create cavities.  The primary etiological
agent of human dental caries, \textit{Streptococcus mutans}, is one of hundreds
bacterial species that constitute the oral microbial community
\citep{Jenkinson2005}.  Bacteria can selectively use substrates from different
carbon sources via a regulation called carbon catabolite repression (CCR)
\citep{Gorke2008}.  \textit{S.\ mutans} also possesses biological systems for
CCR.  CCR were known to regulates the expression of virulence factors in many
pathogenic bacteria, thereby subjecting a number of genes to CCR.  CcpA was
shown to regulate metabolism and virulence gene expression in 
\textit{S.\ mutans} through CCR in transcriptom study using 
the microarray technique based on DNA hybridization \citep{Abranches2008}. 

% Technological motivations: RNA-seq, and microarray, bacterial transcriptomes
It has been a decade since the first genome of the oral bacteria was finished
\citep{Ajdic2002}.  Studies of the transcriptome in the bacteria were performed
based on DNA hybridization
\citep[e.g.,][]{Klein2010,Lemme2011,Redanz2011,Wen2011} whereas those in other
bacterial species took advantage of high-throughput sequencing technologies:
e.g.,
\textit{Mycoplasma pneumoniae} \citep{Guell2009},
\textit{Methanosarcina mazei} \citep{Jaeger2009},
\textit{Bacillus subtilis}  \citep{Irnov2010},
\textit{Helicobacter pylori} \citep{Sharma2010}, and
\textit{Sulfolobus solfataricus} \citep{Wurtzel2010}, to name a few.
The advent of the high-throughput RNA deep sequencing technologies allowed
annotation-free detection of transcripts, sensitive and dynamic ranges of
measurements in RNA expression levels, and studies of non-coding RNAs. 

% Aims of the study
We used the standard RNA deep sequencing technique to sequence RNAs in
\textit{S.\ mutans}, focusing on messenger RNAs expression levels in response to
different nutrients with UA159 wild-type and \textit{ccpA}-deleted TW1
mutant-type.  We describe transcripts of the mRNAs using the RNA-seq
data, and present several small non-coding RNAs in \textit{S.\ mutans}. 
Using the map of transcripts
and expression levels of the transcripts from different biological conditions
allowed us to determine which transcripts and non-coding small RNAs were
differentially expressed. This provided more refined views of dynamic changes in
the transcriptome landscape of the etiological agent of human dental caries,
\textit{S.\ mutans}, than previously reported by \citet{Abranches2008}.

\section*{Materials and Methods}

\subsection*{Bacterial strains and growth conditions}
\fixme{Burne group describes bacterial strains and growth conditions and other
necessary works. Please, add texts and SCC will copy-edit them.}

\subsection*{RNA isolation}
\fixme{Burne group describes RNA isolations and other works. Please, add texts
and SCC will copy-edit them.}

\subsection*{RNA sequencing}
Deep sequencing of the RNA samples were performed at the Cornell University Life
Sciences Core Laboratories Center (Ithaca, NY).  RNA extraction was performed
using a modified Qiagen RNAease protocol (Qiagen, Inc., Chatsworth, CA): cells
were pelleted, resuspended in RNA protect reagent, frozen, and then resuspended
in TE, mechanically disrupted in a bead beater in the presence of glass beads
and acidic phenol. We used the Qiagen kit to obtain total RNAs.  The RNA sample
was passed through the Ambion MicrobExpress kit twice, which removed ribosomal
RNAs only using magnetic beads.  This procedure left small RNAs and transfer
RNAs intact.  Then, we followed the TruSeq Illumina kit to prepare the cDNA
libraries.  

\subsection*{Short reads alignments}
About 20 millions of short reads were sequenced for each sample.  We used a
short read aligner, \program{BWA} \citep{Li2009a}, to map the sequenced short
reads on the reference genome, \textit{S.\ mutans} UA159 (NCBI accession
NC\_004350). Because the aligner allowed a few gaps for efficient alignments of
millions of sequences of size being 100 base pairs long, short reads that
consisted of RNAs with sequencing adapter sequences would not be mapped.  We
used \program{cutadapt} \citep{Martin2011} to remove parts of adapter sequences
from short reads before mapping on the reference genome.  Mapped short read
alignments were converted into readable formats using \program{samtools}
\citep{Li2009b}. 

\subsection*{Transcript predictions}
RNA transcripts were inferred by applying a hidden Markov model to site-wise
expression levels \citep{Martin2010}.  We used the pileup command of
\program{BWA} to convert the short reads alignments to pilup values, which were
taken as the site-wise expression levels along the genome.  Site-wise expression
levels are a list of site-wise non-negative integers that represents numbers of
short reads mapped to a particular genomic position.  We used the following
options for the transcript inference program, \program{ParseRNAseq}: ``-c 10 -b
25 -force\_gp'' to bin expression levels into 25 parts, and to allow 10 emission
states for relative expression intensity. We also used the genes as the
predicted parts of transcripts.    Because we enriched mRNAs in the cDNA
preparation, we intented to group annotated genes to segments, which could be
transcripts. We did not aim at finding precise transcription start or end sites.
Because we did not have strand-specific RNA-seq of the samples due to limited
resources, we could not tell in which directions transcripts were oriented
solely based on RNA-seq data. Instead, we used strands of annotated genes to
determine strands of predicted transcripts.

\subsection*{Prediction of small RNAs and targets}
Although our RNA-seq data did not enriched non-coding small RNAs in the cDNA
preparation, we retained small RNAs in the sample by attempting to deplete only
ribosomal RNAs from the sample. Because non-coding small RNAs and coding
mRNAs were mixed in the sample, it was difficult to discriminate small
non-coding RNAs from mRNAs unless non-coding RNAs expressions were masked by
neighboring background RNA expressions. 
Therefore, we sought a way of predicting small RNAs
using \program{RNAz} \citep{Gruber2010}, which  used homologous sequences and
RNA secondary structures to predict putative non-coding RNAs. Because sequence
alignments were a critical step for finding small RNAs in the approach, we
collated sequences similar to intergenic regions, and refined the alignments using
a multiple sequence alignment tool. Exracting
intergenic regions including up- and down-stream, we searched a database of
bacterial genomes using BLAST \citep{Altschul1990} for similar intergenic
sequences, subsequently, refining the sequence alignments using MUSCLE
\citep{Edgar2004a}.  Then, we applied \program{RNAz} to the alignments for
scoring regions for small RNAs \citep{Washietl2005,Gruber2010}.  Targets of
transcripts with high scores for small RNAs were also predicted using
\program{RNAplex} \citep{Tafer2008}, and \program{RNAplfold}
\citep{Bernhart2006}.  Target genes were scored for each small RNA candidate,
which was used to perform functional category enrichment tests of targets for
each small RNA candidate.

\subsection*{Experimental validation of small RNAs}
\fixme{Burne group might have done or be able to do some experimental
validations of putative non-coding RNAs shown at table \ref{tabUnannotatedRNA}
\citep[e.g.,][]{Altuvia2007}.}

\subsection*{Statistical analysis for differential expressions}
We used \Rpackage{DESeq} \citep{Anders2010} to determine differentially
expressed genes based on negative binomial model originally developed by
\citet{Robinson2007}.  Steps of analyzing RNA-seq data for determining
differentially expressed genes were described in \cite{Oshlack2010}.  We counted
short reads aligned to annotated genes in the reference genome, forming a table
of read counts of the genes. Statistical software R \citep{R2011} with
\Rpackage{DEseq} was subsequently employed to infer differentially expressed
genes in the different biological conditions.  

\subsection*{Gene category associations}
We compiled three sets of gene categories for testing functional associations.
Firstly, we assigned genes to Gene Ontology (GO) categories by comparing the
\textit{S.\ mutans} genes to bacterial proteins from the Uniref90 database using
\program{BLASTP}, and then assigning the same GO classification as the target
gene of the uniProt GOA database if the match had an $E$-value of
$<1.0\times10^{-5}.$  A gene family was assigned a given classification if any
of its genes was assigned that classification.  Secondly, we collated functional
classes of genes in the Oral Pathogen Sequence Databases available at
\url{http://oralgen.lanl.gov}. Thirdly, we used a set of metabolite pathways for
\textit{S.\ mutans} available at Kyoto Encyclopedia of Genes and Genomes (KEGG)
\citep{Kanehisa2012}.  Using the sets of gene categories for \textit{S.\ mutans}
we performed a Mann-Whitney $U$ test of the values of target genes of small RNAs
associated with a given category versus those for the other categories. We
used a variation of Fisher's exact test accounting for gene lengths to test
enrichments of differentially expressed genes \citep{Young2010}.  Both of the
tests were corrected for multiple testing hypotheses \citep{Benjamini1995}.

\subsection*{UCSC genome browser tracks and data availability}
We created tracks for our recently released Streptococcus Genome Browser
(\url{http://strep-genome.bscb.cornell.edu}) that summarized the results of our
\textit{S.\ mutans} transcriptome analysis with known genes, gene expression
levels based on the short read alignments, predicted putative transcript
segments, and predictd small non-coding RNAs.  They should be accessed by clade
\textit{Streptococcus}, and genome \textit{S. mutans UA159}.  These tracks can
be used to inspect loci of interest and to compare the results of different
RNA-seq data sets.  They can also be queried and intersected with other tracks
using the UCSC Table Browser.

\section*{Results}

<<eval=FALSE,echo=FALSE,results=hide>>=
# not run when smutans R package is installed
# Use this when smutans R package is being tested.
library(DESeq)
library(GenomicRanges)
library(plotrix)
source("smutans/R/class_and_slots.R")
source("smutans/R/core.R")
source("smutans/R/methods.R")
load("smutans/data/smutansData.tx.RData")
load("smutans/data/smutansData.txGenes.RData")
load("smutans/data/smutansData.txGrCombined.RData")
load("smutans/data/smutansData.txAnnotation.RData")
load("smutans/data/smutansData.txRNAz.RData")
load("smutans/data/smutansData.txPileup.RData")
load("smutans/data/smutans.feature.genes.RData")
load("smutans/data/smutans.genes.criteria.RData")
load("smutans/data/smutans.go.genes.RData")
load("smutans/data/smutans.cat.desc.RData")
load("smutans/data/smutans.keg.genes.RData")
load("smutans/data/smutans.keg.desc.RData")
load("smutans/data/smutansGenes.RData")
load("smutans/data/smutansGenes2.RData")
@
<<echo=FALSE,results=hide>>=
# not run without smutans R package installed
# Remove eval=FALSE when the R package is installed
library(smutans)
inDir = system.file("extdata", package="smutans", mustWork=TRUE)
smutansGenes <- 
  readSmutans (countsFile=file.path(inDir, "count-ua159-bwa.txt"),
               indexFile =file.path(inDir, "count-ua159.txt.index"),
               condition = c("UA159GLU", "UA159GAL", "TW1GLU", "TW1GAL"),
               firstFactor = c("UA159GLU", "UA159GAL"),
               firstFactorLabel = c("ua159", "tw1"),
               secondFactor = c("UA159GLU", "TW1GLU"),
               secondFactorLabel = c("glucose","galactose"),
               name="S. mutans UA159, TW1, Glucose, and Galactose")
data(smutansData.tx)
data(smutansData.txGenes)
data(smutansData.txGrCombined)
data(smutansData.txAnnotation)
data(smutansData.txRNAz)
data(smutansData.txPileup)
data(smutans.feature.genes)
data(smutans.genes.criteria)
data(smutans.go.genes)
data(smutans.cat.desc)
data(smutans.keg.genes)
data(smutans.keg.desc)
data(smutansGenes2)
@

<<echo=FALSE,result=hide>>=
notExpressedTx <- smutansData.txGrCombined[elementMetadata(smutansData.txGrCombined)$zero > 0.1 & elementMetadata(smutansData.txGrCombined)$pileup < 5]
grTx <- smutansData.txGrCombined[!(elementMetadata(smutansData.txGrCombined)$zero > 0.1 & elementMetadata(smutansData.txGrCombined)$pileup < 5)]
# We query genes and annotations (genes and RNAs) againt tx's.
matchGenesTx <- findOverlaps(smutansData.txGenes,grTx)
matchAnnotationTx <- findOverlaps(smutansData.txAnnotation,grTx)
# I use gaps with IRanges because gaps with GenomicRanges does not work as expected.
grNoAnnotation <- gaps(ranges(smutansData.txAnnotation))
grNoncoding <- gaps(ranges(smutansData.txGenes))
grRNAz <- ranges(smutansData.txRNAz)
# 
matchUnannotatedRNA <- findOverlaps(grRNAz,grNoAnnotation,type="within")
txRNAz.unannotated <- grRNAz[matchMatrix(matchUnannotatedRNA)[,1]]

# A bug in GenomicRanges: the last segment is the while genome. I removed it
# from the noncoding ranges.
# matchNoncoding <- findOverlaps(grNoncoding,grTx)

# Expressed transcripts without annotated genes
grTxWGenes <- grTx[unique(matchMatrix(matchGenesTx)[,2])]
grTxWOGenes <- grTx[-unique(matchMatrix(matchGenesTx)[,2])]
grTxWOAnnotation <- grTx[-unique(matchMatrix(matchAnnotationTx)[,2])]

# Find transcripts with genes with conflicted strands
grGenesPlus <- smutansData.txGenes[strand(smutansData.txGenes)=='+']
grGenesMnus <- smutansData.txGenes[strand(smutansData.txGenes)=='-']
#mtchPlus <- findOverlaps(grGenesPlus,grTxWGenes)
#mtchMnus <- findOverlaps(grGenesMnus,grTxWGenes)
#txWithConflictedGenes <- intersect(matchMatrix(mtchPlus)[,2],matchMatrix(mtchMnus)[,2])

# Tx with genes strands
grTx2 <- grTxWGenes[!strand(grTxWGenes)=='*'] 

# Find the number of genes in segments.
matchGenesTx2 <- findOverlaps(smutansData.txGenes,grTx2,type="within")
number.genes.tx <- runLength(Rle(matchMatrix(matchGenesTx2)[,"subject"]))

# Length distribution of UTR regions
# grTx2 <- grTxWGenes[-txWithConflictedGenes] 
# Lengths of 5' and 3' UTRs
mtchPlus <- findOverlaps(grGenesPlus,grTx2,type="within")
mtchMnus <- findOverlaps(grGenesMnus,grTx2,type="within")
length5UTR <- c()
length3UTR <- c()
for (i in unique(matchMatrix(mtchPlus)[,"subject"])) {
  x <- matchMatrix(mtchPlus)[matchMatrix(mtchPlus)[,"subject"]==i,"query"]
  y <- min(start(grGenesPlus[x])) - start(grTx2[i])
  z <- end(grTx2[i]) - max(end(grGenesPlus[x])) 
  length5UTR <- c(length5UTR,y)
  length3UTR <- c(length3UTR,z)
}
for (i in unique(matchMatrix(mtchMnus)[,"subject"])) {
  x <- matchMatrix(mtchMnus)[matchMatrix(mtchMnus)[,"subject"]==i,"query"]
  y <- min(start(grGenesMnus[x])) - start(grTx2[i])
  z <- end(grTx2[i]) - max(end(grGenesMnus[x])) 
  length5UTR <- c(length5UTR,z)
  length3UTR <- c(length3UTR,y)
}

# Correlation of expression levels of adjacent genes.
pileup.x <- lapply(ranges(smutansData.txGenes), function(x) seqselect(smutansData.txPileup, x))
pileup.y <- mapply(function(x,y) sum(x)/y, x=pileup.x, y=width(smutansData.txGenes))
elementMetadata(smutansData.txGenes)["score"] <- pileup.y
mtch3 <- findOverlaps(smutansData.txGenes,grTx2)
x <- rle(matchMatrix(mtch3)[,"subject"])
comp.exp.two.genes <- c()
for (i in x$values[x$lengths > 1]) {
  y <- combn(matchMatrix(mtch3)[,"query"][matchMatrix(mtch3)[,"subject"]==i],2)
  for (j in 1:dim(y)[2]) {
    comp.exp.two.genes <- cbind(comp.exp.two.genes,as.matrix(elementMetadata(smutansData.txGenes)["score"][y[,j],]))
  }
}
# Correlation of expression levels of adjacent genes between transcripts.
comp.exp.two.genes.between.transcript <- c()
for (i in 1:length(x$values)) {
  if (i == length(x$values)) {
    break
  }
  y1 <- matchMatrix(mtch3)[,"query"][matchMatrix(mtch3)[,"subject"]==x$values[i]]
  i2 <- i + 1
  y2 <- matchMatrix(mtch3)[,"query"][matchMatrix(mtch3)[,"subject"]==x$values[i2]]
  l <- as.matrix(do.call(expand.grid, list(y1,y2)))
  for (j in 1:dim(l)[1]) {
    comp.exp.two.genes.between.transcript <- cbind(comp.exp.two.genes.between.transcript,as.matrix(elementMetadata(smutansData.txGenes)["score"][l[j,],]))
  }
}
@

\subsection*{Experimental results from Burne group}
\fixme{Add some if needed.}

\subsection*{Results from Stanhop group when RNA sequencing preparation}
\fixme{Add some if needed.}

\subsection*{Statistics of short reads}

\subsection*{Cluster of RNA samples}
First using the table of read counts of annotated genes we formed a cluster of
the 13 RNA-seq samples (fig.\ \ref{figClust}). The effect of nutrients appeared
to be stronger than that of the deletion of \textit{ccpA} genes from the UA159
genome. 

\subsection*{Predicted transcripts}

We obtained a set of segments of RNA-seq expressions for each of the 13 RNA-seq
samples.  All of the segments for each set of RNA-seq expression segments were
scored by the average proportions of segments shared by the segments in the rest
of sets.  We placed more highly ranked segments first on the reference genome
based on the scores, and purged more lowly ranked segments from the
transcriptome map. In doing so, the
final set of segments was formed by placing non-overlapping highly ranked
segments.  Average site-wise expression levels of the segments 
are shown at figure \ref{figGrCombinedExpression}. 
We designated segments as expressed that had site-wise pileup values
greater than 5 and proportion of sites of zero site-wise pileup values less than
10\%.  A gene was also designated as an expressed if the gene belonged to an
expressed segment.  Some of the predicted segments were not expressed or very
lowly expressed (\Sexpr{length(notExpressedTx)} 
out of 
\Sexpr{length(smutansData.txGrCombined)}).
Of the total \Sexpr{length(smutansData.txGenes)} of genes,
\Sexpr{length(matchMatrix(matchGenesTx)[,1])}
(\Sexpr{as.integer(length(matchMatrix(matchGenesTx)[,1])/length(smutansData.txGenes)*100)}\%) 
were designated as expressed. 
Genes that were not designated as expressed could be due to false gene
annotations or possibility of true negative expression under the experimental
conditions. 
Figure \ref{figLength} shows the length distribution of expressed and
not-expressed genes. 
Of the \Sexpr{length(grTx)} expressed segments, 
\Sexpr{length(grTxWOAnnotation)} contained no annotated genes or RNAs.
Many of the segments might cover gaps between real operons.
\fixme{Examine these unannotated regions in the UCSC genome tracks.}

Figure \ref{figTranscriptLength} shows the length distribution of expressed
segments with annotated coding genes and unannotated regions.  Unannotated
regions could include potential non-coding regions and novel coding genes.
Segments of unannotated regions appeared to be shorter than those with
coding genes. Most of the unannotated regions with length being near 360 bp
could be non-coding regions, and the rest of the unannotated regions with length
being near 1000 bp could be new coding genes. Some of the unannotated genes
could be pseudogene regions.  We used strandnesses of known genes to find
strands of segments although we could not tell strands of the predicted
transcripts solely based on our RNA-seq data because of our strand-nonspecific
RNA-seq technique.  Of
the \Sexpr{length(grTx)} expressed transcripts, 
\Sexpr{sum(strand(grTx)=='*')}
included annotated genes of both strands
(\Sexpr{as.integer(sum(strand(grTx)=='*')/length(grTx)*100)}\%).
We filtered out the transcripts with genes with conflicted both strands leaving
\Sexpr{length(grTx2)} expressed transcripts. 
Figure \ref{figUTRLength} shows length distributions of 5' and 3' UTR of the
expressed transcripts without genes with conflicted strands.  Many transcripts
appeared to start and end near the boundaries of annotated genes.  Gene
expression levels for adjacent genes within expressed transcripts were more
correlated than those between adjacent transcripts 
(fig.\ \ref{figExpressionTwoGene}).
Among the \Sexpr{length(unique(matchMatrix(matchGenesTx2)[,"subject"]))} segments
with genes of all of the same strands, 
\Sexpr{sum(number.genes.tx==1)} segments contained a single gene while 
\Sexpr{sum(number.genes.tx > 1)} contained more than one gene.

\subsection*{Predictions of small RNAs}

We predicted
\Sexpr{length(colnames(smutans.genes.criteria))-1} putative non-coding RNAs, most of
which were already annotated.  
Of the \Sexpr{length(colnames(smutans.genes.criteria))-1} predicted non-coding RNAs
\Sexpr{length(txRNAz.unannotated)} were in  
unannotated non-coding regions (table \ref{tabUnannotatedRNA}).  Using
\program{RNAplex} \citep{Tafer2008} and \program{RNAplfold} \citep{Bernhart2006}
we predicted target genes of the putative non-coding RNAs.  Table \ref{tab:goseq}
shows functional categories enriched with target genes for predicted small RNAs.
We found that a few of the putative small RNAs were structured with stable stems
(figs.\ \ref{figRNALocus1}, \ref{figRNALocus2}, and \ref{figRNALocus3}).  
\fixme{We might be able to 
further experimentally verify potentially important small RNAs.}

\subsection*{Differential expression of RNA transcripts}

<<echo=FALSE,result=hide>>=
glucoseOnly <- newSmutans( smutansGenes, title="Glucose Only" )
glucoseOnly <- smutans.de2( glucoseOnly, condition="glucose", 
                            condA="ua159", condB="tw1" )
save(glucoseOnly, file=file.path("glucoseOnly.RData"))
#load("glucoseOnly.RData")

smutans.de2List( glucoseOnly, "glucoseOnly.csv" )
tableGlucoseOnly <- smutans.de2List( glucoseOnly, qval=1e-3 )

# Gene descriptions
geneDescription <- read.delim("feature-genome.out-geneonly.txt",head=FALSE,sep="\t")
geneDescription <- geneDescription[,c(4,7)]
colnames(geneDescription) <- c("Gene ID","Gene Description")
# Functional classes
functionalDescription <- read.csv("feature-genome.out-geneonly-oralgen-functional.txt",head=FALSE)
functionalDescription <- functionalDescription[,c(2,12)]
colnames(functionalDescription) <- c("Gene ID","Functional Class")
# Find potential differentially expressed genes, and add gene descriptions, and
# functional classes
x <- smutans.de2List( glucoseOnly, qval=1e-3 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","UA159","TW1","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
tableGlucoseOnly <- x
@

Two independent factors of the biological processes included the effect of CcpA
and the treatment of mono-saccharides (glucose versus galactose), resulting in
four pair-wise comparisons. We compared two conditions of one factor with the
other factor being fixed at one condition. For example, we compared UA159
wild-type and TW1 mutant-type strains under the condition of glucose-grown cells
to determine differentially expressed transcripts or genes potentially due to
presence and absence of deactivation of \textit{ccpA} genes.  We performed a
similar comparison with galactose-grown cells. We replicated four
biological experiments for the condition combination of UA159 wild-type grown
in galactose, and three replicates for the other three conditions. 

\subsection*{Comparison between wild-type and mutant-type in glucose-grown cells}

Table \ref{tabGlucoseOnly} lists differentially expressed genes in the
comparison of UA159 wild-type and TW1 mutant-type grown with glucose.
Of \Sexpr{nrow(counts(smutansGenes))} genes, 
\Sexpr{nrow(tableGlucoseOnly)} genes were found to be differentially expressed
at FDR of 0.1\%.  We confirmed that the most up-regulated genes in the TW1
mutant-type encoded the components of the PDH enzyme complex (SMU.1421,
SMU.1422, SMU.1423, and SMU.1424) as shown previously.  
These genes appeared to be transcribed in
a single operon (fig.\ \ref{figUCSC}). 
Figure \ref{figPieGlucoseOnly} shows the
proportions of \program{oralgen} databases functional classes by the number of
differentially expressed genes.  Notably, we found
\Sexpr{length(grep("Energy metabolism",tableGlucoseOnly[,"Functional Class"]))}
genes that were involved in energy metabolism where sugars and fermentation
subclasses were profound. We found 
\Sexpr{length(grep("PTS",tableGlucoseOnly[,"Functional Class"]))}
genes in PTS class of signal transduction. Another subclass, 
``carbohydrates, organic alcohols, and acids,'' in
``transport and binding proteins'' was following next in the number of
differentially expressed genes. We also found  
\Sexpr{length(grep("hypothetical",tableGlucoseOnly[,"Gene Description"]))}
genes annotated as hypothetical, which might constitute the unknown functional
class. These results appeared to be consistent with those of \citet{Abranches2008}.

We further tested functional category enrichments using gene ontology terms and KEGG
pathways. Table \ref{tab:goseqGlucoseOnly} shows that differentially expressed
genes were associated with the following gene ontology terms:
phosphotransferase system, 
glycogen biosynthetic process, and
sugar:hydrogen symporter activity.
Among the KEGG pathways, starch and sucrose metabolism and phosphotransferase
system (PTS) appeared to be highly associated with the differentially expressed
genes (table \ref{tab:kegseqGlucoseOnly}).
Figure \ref{figDotplotGlucoseOnly} shows that differentially expressed genes
tended to increase in expression levels in response to the deletion of
\textit{ccpA} genes.  Profoundly upregulated genes were putative Clp proteinase
(SMU.1425), branched-chain alpha-keto acid dehydrogenase subunit E2 (SMU.1421),
and putative transcriptional regulator (SMU.1599).  Much less proportion of the
differentially expressed genes were found to be decreased in expression levels
in response to the lack of CcpA (fig.\ \ref{figDotplotGlucoseOnly}).
Down-regulated genes include cytoplasmic alpha-amylase (SMU.1590),
beta-D-fructosyltransferase (SMU.2028), and putative glycerol uptake facilitator
protein (SMU.396).  Fructosyltransferase converts sucrose to fructose.

Figure \ref{figFoldPValueGlucoseOnly} shows fold changes and $p$-vallues are
negatively correlated. \fixme{Why did I want to note this observation?}

\subsection*{Comparison between wild-type and mutant-type in galctose-grown cells}

<<echo=FALSE,result=hide>>=
galactoseOnly <- newSmutans( smutansGenes, title="Galactose Only" )
galactoseOnly <- smutans.de2( galactoseOnly, condition="galactose", 
                              condA="ua159", condB="tw1" )
save(galactoseOnly, file=file.path("galactoseOnly.RData"))
#load("galactoseOnly.RData")
x <- smutans.de2List( galactoseOnly, qval=1e-3 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","UA159","TW1","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
tableGalactoseOnly <- x
@
We found that differentially expressed genes in galactose-grown cells were
smaller than those in glucose-grown cells, which was different from the result
of \cite{Abranches2008}. 
Energy metabolism was the most
profoundly differentially expressed (fig.\ \ref{figPieGalactoseOnly}).
Subclasses pyruvate family in amino acid biosynthesis, 
and degradation of proteins, peptides, and glycopeptides in protein fate
were following the next. Almost a quarter of the genes were unknown.
Down-regulated genes included cytoplasmic alpha-amylase (SMU.1590), and
beta-D-fructosyltransferase (SMU.2028)
(table \ref{tabGalactoseOnly}). We could not find gene ontology terms that were
significantly associated with differentially expressed genes. It might be
because many genes were unknown in their functions 
(fig.\ \ref{figDotplotGalactoseOnly}).
However, a few KEGG pathways including valine, leucine and isoleucine
biosynthesis were found to be significant (table \ref{tab:kegseqGalactoseOnly}).

\subsection*{Comparison between glucose- and galactose-grown wild-type cells}

<<echo=FALSE,result=hide>>=
ua159only <- newSmutans( smutansGenes, title="UA159 Only" )
ua159only <- smutans.de2( ua159only, type="ua159", 
                          condA="glucose", condB="galactose" )
save(ua159only, file=file.path("ua159only.RData"))
#load("ua159only.RData")
x <- smutans.de2List( ua159only, qval=1e-10 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","Glucose","Galactose","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
tableUA159Only <- x
@
Among the four pairwise comparisons this showed the largest number of
differentially expressed genes (table \ref{tabUA159Only}). We used a smaller
cutoff value for listing them. 
Figure \ref{figPieUA159Only} shows proportons of functional categories that
included differentially expressed genes between glucose-grown and
galactose-grown UA159 wild-type cells. Qualitatively, energy metabolism and PTS
classes were profound. 
``Energy metabolism; Fermentation'' had highly differentially expressed genes
(fig.\ \ref{figDotplotUA159Only}).
This was similar to the comparison between UA159
wild-type and TW1 mutant-type grown in glucose. 
More than a quarter of the
differentially expressed genes were unknown, unassigned, or hypothetical.
Consecutive genes (SMU.1488c to SMU.1496) were upregulated in 
catabolite-derepressing conditions or in galactose-grown cells
(\fixme{Are these the genes of the lactose (lac) operon?})
(table \ref{tabUA159Only}).  Another set of genes (SMU.886 to SMU.889) were also
upregulated (\fixme{Are these called galKTE?}). Whereas the
four genes appeared to be transcribed together under galactose condition,
two segments seemed to be transcribed by dividing the four genes into two pairs
of genes (fig.\ \ref{figUCSC2}).  Genes from SMU.1535 to SMU.1539, which
appeared to be transcribed together, were upregulated.
Of the \Sexpr{nrow(tableUA159Only)} differentially expressed genes at
$1 \times 10^{-10}$ cutoff of multiple testing adjusted p-value, 
\Sexpr{length(grep("hypothetical",tableUA159Only[,"Gene Description"]))} were
hypothetical, and most of them were down-regulated
(fig.\ \ref{figDotplotUA159Only}).
We also tested functional category enrichments using gene ontology terms 
(table \ref{tab:goseqUA159Only}) and KEGG pathways 
(table \ref{tab:kegseqUA159Only})
where carbohydrate metabolic process and Galactose metabolism were top listed,
respectively.

\subsection*{Comparison between glucose- and galactose-grown mutant-type cells}

We found sugar metabolism and PTS signal transduction to be profound in
differential expression in the comparison of glucose-grown and galactose-grown
TW1 mutant-type cells (fig.\ \ref{figPieTW1Only}). Many of the genes were also
hypothetical. We found similar genes differentially expressed in TW1 mutant-type
cells to those in UA159 wild-type 
(table \ref{tabTW1Only} and fig.\ \ref{figDotplotTW1Only}).
The comparisons between glucose-grown and galactose-grown UA159, 
and that from TW1 mutant-type showed very similar functional categories
although ranks of gene ontology terms were different
(tables \ref{tab:goseqUA159Only} and \ref{tab:goseqTW1Only}).
Lactose metabolic process was top listed in TW1, and
carbohydrate metabolic process was top listed in UA159.
KEGG pathways in TW1 were also similar to those in UA159 
(table \ref{tab:kegseqTW1Only}) .

\subsection*{Comparison between glucose- and galactose-grown cells using all samples}
As previously figure \ref{figClust} shows the nutrient effect was stronger, and
the effects of nutrient changes appeared to be similar whether UA159
wild-type or TW1 mutant-type was used, we used all of the 
\Sexpr{length(colnames(counts(smutansGenes)))} samples to compare glucose-grown
and galactose grown cells.
<<echo=FALSE,results=hide>>=
allSample <- newSmutans( smutansGenes, 
                         title="All Samples for Comparing Conditions" )
allSample <- smutans.de2( allSample, 
                          condA="glucose", condB="galactose" )
save(allSample, file=file.path("allSample.RData"))
#load("allSample.RData")
tabUA159onlyAndAllSample <- smutans.compareResult( ua159only, 
                                                   allSample )
#addmargins( tabUA159onlyAndAllSample )
@
Of the total \Sexpr{sum(tabUA159onlyAndAllSample)} gene,
\Sexpr{tabUA159onlyAndAllSample[2,2]} genes were found to be differentially
expressed in both the analysis of using only UA159-grown cells and the analysis
of using all of the 13 samples. The
analysis using all of the sample found \Sexpr{tabUA159onlyAndAllSample[1,2]}
genes. The analysis using the partial samples found only
\Sexpr{tabUA159onlyAndAllSample[2,1]} genes.  We would expect that it was 
because of higher power of the two-factor analysis with more data.

\subsection*{Differentially expressed small RNAs and transcripts}

<<echo=FALSE,results=hide>>=
ua159only <- newSmutans( smutansGenes2, title="UA159 Only" )
ua159only <- smutans.de2( ua159only, type="ua159", 
                          condA="glucose", condB="galactose" )
allSample <- newSmutans( smutansGenes2, 
                         title="All Samples for Comparing Conditions" )
allSample <- smutans.de2( allSample, 
                          condA="glucose", condB="galactose" )
tabUA159onlyAndAllSample <- smutans.compareResult( ua159only, 
                                                   allSample )
# smutans.de2List( allSample )[grep("locus",rownames(smutans.de2List(allSample))),]
@
We found a few predicted small RNAs that appeared to be differentially
expressed: locus69, locus101, locus7, and locus1. 
Two non-coding regions of locus69 and locus101 seemed to be too weak.
However, locus1 and locus7 seemed to have some noticeable differences between
glucose and galactose conditions.
We also found a few predicted small RNAs that appeared to be differentially
expressed in the comparison of the two types UA159 and TW1: locus68,
locus8, and locus10.  The locus68 was overlapped with T-box region.
The two regions of locus8 and locus10 showed very weak and subtle differences
between UA159 and TW1. 
<<echo=FALSE,results=hide>>=
allSampleType <- newSmutans( smutansGenes2, title="All samples for type" )
allSampleType <- smutans.de2( allSampleType, compareCondition="no", 
                              condA="ua159", condB="tw1" )
# smutans.de2List( allSampleType )[grep("locus",rownames(smutans.de2List(allSampleType))),]
@
Figures \ref{figRNALocus1}, \ref{figRNALocus2}, and \ref{figRNALocus3} shows MFE
structures with encoding positional entropy of differentially expressed putative
structural non-coding RNAs.

\section*{Discussion}

Our current knowledge of the transcriptome is limited to annotated or predicted
genes.  Whole transcriptome sequencing using high throughput sequencing
technologies, or deep RNA sequencing (RNA-seq) allowed to reveal the
transcriptome topography of \textit{S.\ mutans}.  Instead of aiming at
pin-pointing transcript start or end sites we focused on differential expression
of genes in response to nutirent availability and deletion of \textit{ccpA} from
UA159.  Although we did not use strand-specific RNA-seq that could be important
in gene-dense transcriptomes of species like bacteria or archaea, our
application of a hidden Markov model for predicting transcripts to RNA-seq data
in conjunction with known gene annotations allowed to infer segments of
putatively co-transcribed regions. In doing so, we attempted to ensure a
high-quality transcriptome assembly.  Although this map of transcripts did not
precisely tell us where transcripts start and end, it informed us of which open
reading frames were transcribed together.  We also found discrepancies of
annotated genes and their transcriptions: a gene contains two transcripts, and
non-coding regions contain transcripts.  We hope that our study could serve as a
foundation for a comprehensive study of the \textit{S.\ mutans} transcriptome.

\section*{Acknowledgments}
\fixme{Add more peoples or fundings if needed.}
This study was funded by the National Institute of Allergy and Infectious
Disease, US National Institutes of Health, under grant number AI073368-01A2
(to M.J.S.\ and A.S).  Additional support was provided by National Science
Foundation CAREER Award DBI-0644111 and a David and Lucile Packard
Fellowship for Science and Engineering (to A.S.).  
We thank Christopher Browngardt for assistance with labelling functional classes
of \textit{S.\ mutans} genes.

\renewcommand*{\refname}{Literature Cited}
\bibliographystyle{../../latex/bst/mbe}
\bibliography{siepel-mutans}
\clearpage{}

\section*{Tables}

\clearpage{}

\section*{Figures}

\end{document}

