\documentclass{article}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{color} 
\usepackage{setspace} 
\usepackage{rotating}
\usepackage{natbib}
\usepackage{xr}
\usepackage{../../latex/sty/genres}
\externaldocument{supplementary}

% Text layout
\oddsidemargin 0in
\evensidemargin 0in
\topmargin -.5in
\textwidth 6.5in
\textheight 9in
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

\SweaveOpts{keep.source=TRUE,eps=FALSE,include=FALSE,width=6,height=6}
\newcommand{\Rpackage}[1]{R package \textit{#1}}
\newcommand{\program}[1]{{\texttt{#1}}}
\newcommand{\fixme}[1]{{\textbf{Fixme:} \textit{\textcolor{blue}{#1}}}}
\newcommand{\comment}[1]{}

\makeatother
\pagestyle{myheadings}
\markright{Transcriptome of Streptococcus mutans}

\begin{document}

\begin{titlepage}

\title{Defining Transcriptome using RNA Sequencing\\
of {\em Streptococcus mutans} }

\author{Sang Chul Choi$^{1}$, Charles D.\ Danko$^{1}$, 
Melissa J.\ Hubisz$^{1}$, \\
Paulina Dmitrievna Pavinski Bitar$^{2}$, \\
Robert A.\ Burne$^{3}$
Michael J.\ Stanhope$^{2}$, and Adam Siepel$^{1}$}

\date{ }
\maketitle

\begin{footnotesize}
\begin{center}
$^1$Department of Biological Statistics and Computational Biology,\\
Cornell University, Ithaca, NY 14853, USA
\\[1ex]
$^2$Department of Population Medicine and Diagnostic Sciences,\\
College of Veterinary Medicine, Cornell University, Ithaca, NY 14853, USA
\\[1ex]
$^3$Department of Oral Biology,\\
College of Dentistry, University of Florida, Gainesville, Florida 32610 USA
\\
\end{center}
\end{footnotesize}

\vspace{1in}

\begin{tabular}{lp{4.5in}}
{\bf Submission type:}& Research Article
\vspace{1ex}\\
{\bf Keywords:}&Carbon catabolite repression, RNA-seq, non-coding RNAs
\vspace{1ex}\\
{\bf Running Head:}&Transcriptome of {\em Streptococcus mutans}
\vspace{1ex}\\ 
{\bf Corresponding Author:}&
\begin{minipage}[t]{4in}
 Adam Siepel\\
 102E Weill Hall, Cornell University\\
 Ithaca, NY 14853\\
 Phone: +1-607-254-1157\\
 Fax: +1-607-255-4698\\
 Email: acs4@cornell.edu
\end{minipage}
\end{tabular}

\thispagestyle{empty}
\end{titlepage}

\doublespacing

\section{Abstract}
A bacterial transcriptome from the primary etiological agent of human dental
caries, \textit{Stretptococcus mutans}, is described using deep RNA sequencing
data.  We define transcripts boundaries of open reading frames based on sequence
alignments of short reads on the reference genome of UA159 from the deep
sequencing.  Based on homologous sequences and RNA secondary structures we also
report putative small RNAs, several of which are confirmed experimentally using
both RNA sequencing data, northern blot, and qRT-PCR.  CcpA is known to play a
direct role in carbon catabolite repression (CCR).  We use the developed
transcript map, open read frame, and small RNAs to determine differentially
expressed regions between the wild-type (UA159) and CcpA-deletion (TW1) strains
in order to refine the previously established importance of CcpA regulation as
central metabolism and virulence gene expression.  We find that more genes are
differentially expressed in the comparison between wild-type and CcpA-deletion
strains that are grown in galactose than in glucose.  We also find that more
genes differentially expressed in wild-type than in CcpA-deletion strain when
comparing glucose- and galactose-grown cells.  These two findings are different
from those from microarry experiment of \cite{Abranches2008}.  We also predicted
target genes of the small RNAs that were enriched in the following functional
categories: xxx.  We confirmed that CcpA modulates the pathogenic potential of
\textit{S.\ mutans} through global control of gene expression.

\comment{
\section{Abbreviations}
\begin{enumerate}
\item CCR Carbon catabolite repression
\item CcpA ?
\item IPS intracellular polysaccharide
\item CRE catabolite responsive elements
\end{enumerate}
}

\section{Introduction}
Regulation of carbon catabolite repression (CCR) in \textit{Streptococcus
mutans} can involve a phosphoenolpyruvate:sugar phosphotransferase system (PTS),
histidin protein (HPr) and a transcriptional regulator, CcpA \citep{Gorke2008}.
CcpA was shown to regulate metabolism and virulence gene expression in
\textit{S.\ mutans} by \citet{Abranches2008}. They used microarray experiments
to measure expression levels of genes. Microarray experiments tend to be
qualitative. The advent of high throughput sequencing technology motivated us to
revisit the experiment to refine the result.  The deep RNA sequencing data also
allowed us to define the bacterial transcriptome of \textit{S.\ mutans}. So, we
could compare putative operon expression levels as well as gene expression
levels between the wild-type (UA159) and its mutant-type (TW1) whose
\textit{ccpA} was deleted from the wild-type. This allowed us to profile
expression of genes and putative operons affected by the regulation of
\textit{ccpA}. 

\citet{Martin2010} developed a hidden Markov model to predict transcripts by
using site-wise expression levels. I also refined the prediction of transcripts
using multiple samples. I measured expression levels of known genes and compared
them from one experiment condition and from another. I can test whether two
genes are in the same operon or not. I assume that expression levels of two genes 
vary in similar fashion if the two genes belong to the same operon. What I mean
by similar fashion is whether they are expressed differentially with similar
fold changes. I assume that operons maps do not change between the two
experiments, or multiple experiments. This is not true. I hope that additional
samples could help to define operons. Eventually, operon maps will be determined
experimentally. Computational tools could be helpful before those ages come to
us. \fixme{This idea did not work. We should only use the prediction method.}

Small RNAs and their targets can be predicted \citep{Tafer2008a}.
Although our RNA-seq data did not sample specifically non-coding small RNAs, we
retained small RNAs in the sample by attempting to deplete only ribosomal RNAs
from the sample. Because non-coding small RNAs and coding messenger RNAs were
mixed in the sample, we sought a way of predicting small RNAs using
\program{RNAz} developed by \citet{Gruber2010}.  \program{RNAz} used homologous
sequences and RNA secondary structures to predict putative non-coding RNAs. We
confirmed the several predicted small non-coding RNAs experimentally
\citep{Altuvia2007} including the RNA-seq data set. Because small RNAs could
be associated with CCR regulation of genes (CITE), we determined differentially expressed small
RNAs between RNA samples. We found a few small RNAs to be differentially
expressed.

Bacteria can selectively use substrates from different carbon sources. In the
presence of a preferred carbon source, catabolic system of secondary substrates
are often regulated or shutdown: called carbon catabolite repression
\citep{Gorke2008}. CCR also regulates the expression of virulence factors in many
pathogenic bacteria. A classical example is the glucose-lactose diauxie in
\textit{E.\ coli}. A significant number of genes can be subject to CCR.  CCR in
bacteria involves regulatory networks that activate or silence genes in response
to carbon source and availability.  \textit{S.\ mutans} possesses redundant
systems for CCR.  We could measure directly the expression levels of
\textit{fruA} in \textit{S.\ mutans}.  We assess the contribution of CcpA to CCR
of predicted operons.  Clearly, deletion of \textit{ccpA} resulted in
substantial relief of the CCR of \textit{fruA} expression exerted by both
glucose and fructose.

We aims at refining the results of \cite{Abranches2008} in more quantative ways.
Find CCR-sensitive transcripts. We describe predictions made for transcripts and
small RNAs based on RNA-seq data. Combining the predictions and known genes we
determine differentially expressed regions in UA159 genome. 

\fixme{We need to quantify some uncertainty of transcript predictions.}

\section{Materials and Methods}

\subsection{Bacterial strains and growth conditions}
\fixme{Burne group should be able to describe this part. The following is just
from somewhere for the purpose of writing template. Please, rewrite it.}
\textit{S.\ mutans} UA159 is a human cariogenic isolate. The whole genome is
availabe at NCBI \citep{Ajdic2002}. Strain TW1 was derived from UA159 by
replacing a part of \textit{ccpA} gene \citep{Wen2002}.  \textit{S.\ mutans}
strains were inoculated into TV medium supplemented with 0.5\% galactose or
glucose and grown to early exponential phase (OD600, 0.2).  Then fructose was
added to the medium at concentrations (0,0.05,3,5,10 mM) for 3.5h to induce
\textit{fruA} expression.

Expression levels of \textit{fruA} in \textit{S.\ mutans} strains that were
grown to early exponential phase (OD600, 0.2) in TV medium supplemented with
0.5\% glucose or galactose. 

\subsection{RNA isolation}
\fixme{Burne group should be able to describe this part. The following is just
from somewhere for the purpose of writing template. Please, rewrite it.}
RNA was isolated from S. mutans (1) by using 50 ml cultures that were grown
under the desired conditions and harvested by centrifugation at $4^{\circ}$C. Pelleted
cells were resuspended in 400 ul of diethyl pyrocarbonate-treated water, 800 ul
of RNAprotect reagent (Qiagen, Inc., Chatsworth, CA) was added, and the samples
were incubated at room temperature for 5 min with vortexing for 10 s at 1-min
intervals. Cells were then pelleted, resuspended in 250 ul Tris-EDTA (50:10)
buffer, and transferred to 1.5 ml screw-cap tubes containing sterile glass beads
(average diameter, 0.1 mm; Biospec, Bartlesville, OK), 10 ul of 1\% sodium
dodecyl sulfate, and 300 ul of acid phenol-chloroform (5:1). After
centrifugation for 15 min at maximum speed at $4^{\circ}$C, RNA was further purified
using an RNeasy mini kit (Qiagen, Inc., Chatsworth, CA), including on-column
DNase digestion with RNase-free DNase (Qiagen), as recommended by the supplier.

\subsection{RNA sequencing}
\fixme{SC wrote the second paragraph based on email correspondences between PDPB and CDD.
Please rewrite it if necessary. Stanhope group should be able to describe this
part. The following first paragraph is from a paper. Do not use them or rephrase
them in any way. They are for informational purposes.}
Sequencing libraries for the Illumina GA platform were constructed by shearing
the enriched cDNA by nebulisation (35psi, 6 min) followed by end-repair with
Klenow polymerase, T4 DNA polymerase and T4 polynucleotide kinase (to blunt-end
the DNA fragments). A single 3' adenosine moiety was added to the cDNA using
Klenow exo- and dATP. The Illumina adapters (containing primer sites for
sequencing and flowcell surface annealing) were ligated onto the repaired ends
on the cDNA and gel-electrophoresis was used to separate library DNA fragments
from unligated adapters by selecting cDNA fragments between 200-250 bps in size.
Fragmentation followed by gel-electrophoresis were used to separate library DNA
fragments and size fragments were recovered following gel extraction at room
temperature to ensure representation of AT rich sequences. Ligated cDNA
fragments were recovered following gel extraction at room temperature to ensure
representation of AT rich sequences. Libraries were amplified by 18 cycles of
PCR with Phusion polymerase. Sequencing libraries were denatured with sodium
hydroxide and diluted to 3.5 pM in hybridisation buffer for loading onto a
single lane of an Illumina GA flowcell. Cluster formation, primer hybridisation
and single-end, 36 cycle sequencing were performed using proprietary reagents
according to manufacturers' recommended protocol (https://icom.illumina.com/).
The efficacy of each stage of library construction was ascertained in a quality
control step that involved measuring the adapter-cDNA on a Agilent DNA 1000
chip. A final dilution of 2 nM of the library was loaded onto the sequencing
machine.

Deep sequencing of the RNA samples were performed at the Cornell University Life
Sciences Core Laboratories Center (Ithaca, NY).  RNA extraction was performed
using a modified Qiagen RNAease protocol (Qiagen, Inc., Chatsworth, CA): cells
were pelleted, resuspended in RNA protect reagent, frozen, then resuspended in
TE, mechanically disrupted in a bead beater in the presence of glass beads and
acidic phenol. After these steps, we used the Qiagen kit to obtain total RNAs.
The RNA sample was then passed through the Ambion MicrobExpress kit twice, which
removed ribosomal RNAs only using magnetic beads.  This procedure left small
RNAs and transfer RNAs intact.  We followed the TruSeq Illumina kit to prepare
the cDNA libraries.  We did not intend to enrich short reads in either small
RNAs or transcript boundaries.

\subsection{Statistical analysis for differential expressions}
We used an R package, DESeq, \citep{Anders2010} to determine differentially
expressed genes based on negative binomial model originally developed by
\citep{Robinson2007}.  Steps of analyzing RNA-seq data for determining
differentially expressed genes are described in \cite{Oshlack2010}. We used a
short read aligner called \program{BWA} \citep{Li2009a} to map the sequenced
short reads on a reference genome, \textit{Streptococcus mutans} UA159 (NCBI
accession NC\_004350). Because the aligner allowed a few gaps for efficient
alignment of millions of sequences of size being 100 base pairs long, short
reads that consist of RNAs and sequencing adapter sequences would be not mapped.
We used \program{cutadapt} \citep{Martin2011} to remove parts of adapter
sequences from short reads before mapping on the reference genome. Mapped short
read alignments were converted into readable formats using \program{samtools}
\citep{Li2009b}. We counted short reads aligned to annotated genes in the
reference genome, forming a table of read counts of the genes. Statistical
software R \citep{R2011} with \Rpackage{DEseq} \citep{Anders2010} was
subsequently employed to infer differentially expressed in different biological
conditions.  We also generated tracks for our recently released Streptococcus
Genome Browser (http://strep-genome.bscb.cornell.edu) that summarize gene
expression levels based the short read alignments. This allowed us to visualize
parts of the reference genome for comparing different RNA samples. Further
analyses such as functional category enrichment test \cite{Young2010} were also
employed.

\subsection{Transcript predictions}
\citet{Martin2010} developed a hidden Markov model to predict transcripts by
using site-wise expression levels. We used the following options for the
program: ``-c 10 -b 25 -force\_gp'' to bin expression levels into 25 parts to
allow 10 emission states. We also used the genes as the predicted parts of
transcripts.  Because we did not have strand-specific RNA-seq of the 13 samples
due to limited resources, we could not tell in which directions transcripts
based on RNA-seq data were oriented.

\subsection{small RNAs}
We used \program{RNAz} to predict small RNA candidates. Sequence alignments were
a critical step for finding small RNAs in the approach. Phylogenies of
streptococcus species could be considered to choose species related with
Streptococcus mutans.  Alternatively, we found sequences similar to intergenic
regions, and refined the alignments using a multiple sequence alignment tool
such as MUSCLE \citep{Edgar2004a}. First, we exracted intergenic regions
including up- and down-stream. We searched non-redundant database of bacterial
genomes using BLAST \citep{Altschul1990} for similar intergenic sequences,
subsequently, refining the sequence alignments using MUSCLE \citep{Edgar2004a}.
Then, we applied \program{RNAz} to the alignments for scoring regions for small
RNAs \citep{Washietl2005,Gruber2010}.  Targets of transcripts with high scores
for small RNAs were also predicted using \program{RNAplex} \citep{Tafer2008},
and \program{RNAplfold} \citep{Bernhart2006}.  Target genes were scored for each
small RNA candidate, which was used to perform functional category enrichment
tests of targets for each small RNA candidate. We could find functional
categories that a few small RNAs might be associated with.

\subsection{Gene category associations and virulence genes}
We used two approaches for finding functional categories
\citep{Young2010,Suzuki2011}. We briefly descirbe the prodecure of
\citet{Suzuki2011}.  We assigned genes to Gene Ontology (GO) categories by
comparing the \textit{S.\ mutans} genes to bacterial proteins from the Uniref90
database using \program{BLASTP}, and then assigning the same GO classification
as the target gene of the uniProt GOA database if the match had an $E$-value of
$<1.0\times10^{-5}.$  A gene family was assigned a given classification if any
of its genes was assigned that classification.  To test for associations with
replacing gene transfers we used the recombination intensity estimated for each
gene, and for additive gene transfers we used the number of additive transfers
inferred for each family.  To test for significance, we performed a Mann-Whitney
$U$ test of the values (recombination intensities or numbers of transfers)
associated with a given category vs.\ the values for the all other
genes/families.  We used the \cite{Benjamini1995} method to correct for multiple
comparisons.
\fixme{A short description about \citep{Young2010}.}

\comment{
Findings include:
\begin{enumerate}
\item New genes, open reading frames that are not annotated in NCBI's genome database.
\end{enumerate}
}

\section{Results}
We sequenced a whole prokaryotic transcriptome of \textit{S.\ mutans}, and
mapped short reads to the genomic sequence of the reference genome UA159
\citep{Ajdic2002}. 
\citet{Martin2010} applied a hidden Markov model to use short reads mapped to a
genomic sequence in order to parase \textit{B.\ anthracis} genome into a
sequence of transcripts. Most of annotated genes belong to the predicted
transcripts. We found discrepancies of annotated genes and their transcriptions:
1. A gene contains two transcripts, 
2. Non-coding regions contain transcripts.
\fixme{We could find more discrepancies. Try to find them}. 
Expression levels of adjacent genes within a transcript should be more
correlated to each other than those of adjacent genes belonging to separate transcripts
\citep{Martin2010}. 
We identified a set of transcripts, albeit not a full, including small RNAs, and novel
transcripts. 
We hope that our study could serve as a foundation for a comprehensive study of
the \textit{S.\ mutans} transcriptome.
Our current knowledge of the transcriptome is limited to annotated or predicted genes.
Whole transcriptome sequencing using high throughput sequencing (HTS)
technologies, or deep RNA sequencing (RNA-seq) allowed to reveal the
transcriptome topography of \textit{S.\ mutans}. We did not aim at precisely
pin-pointing transcript start or end sites.

We admit that strand-specific RNA-seq would be important in gene-dense
transcriptomes of species like bacteria or archaea.

\subsection{From Reviews}

We could not use short-read assemblers such as Velvet \citep{Zerbino2008} 
and Trinity that had been developed to assembly transcriptome. Because genes in
bacteria are so dense, RNA-seq data without strands did not allow us to apply
any state-of-the-art software of transcriptome assembly.
We ensured a high-quality transcriptome assembly. 


In the data generation phase (FIG. 1a), total RNAs or mRNAs are fragmented and
converted into a library of cDNAs containing sequencing adaptors.  
In the data analysis phase (FIG. 1b), these short reads are pre-processed to
remove sequencing errors and other artefacts. The reads are subsequently
assembled to reconstruct the original RNAs and to assess their abun- dance
(`expression counting').

To increase the number of assem- bled transcripts, especially the less abundant
ones, ribosomal RNA (rRNA) and abundant transcripts are removed during the
first steps of library construction.
In order to retain RNAs without a poly(A) tail in the assem- bled transcriptome,
rRNA contamination can instead be removed by hybridization-based depletion
methods. \fixme{What depletion method did we use?} 
Last, the use of strand-specific RNA-seq protocols aids in the assembly and
quantification of overlapping transcripts that are derived from opposite strands
of the genome. This consideration is especially important for gene-dense
genomes, such as those of bacteria, archaea and lower eukaryotes.

Sequencing. The major factors to consider before sequencing a sample are: the
choice of sequencing plat- form, the sequencing read length and whether to use a
paired-end protocol.

Removing artefacts from RNA- seq data sets before assembly improves the read
quality, which, in turn, improves the accuracy and computa- tional efficiency of
the assembly. This step is straight- forward and can be executed using several
tools.  In general, three types of artefacts should be removed from raw RNA-seq
data: sequencing adaptors, which origi- nate from failed or short DNA
insertions during library preparation; low-complexity reads; and
near-identical reads that are derived from PCR amplification. 
Sequencing errors in NGS reads can be removed or corrected by analysing the
quality score and/or the k-mer frequency.

Reference-based transcriptome assembly is easier to perform for the simple
transcriptomes of bac- terial, archaeal and lower eukaryotic organisms, as these
organisms have few introns and little alternative splicing. Transcription
boundaries can be inferred from regions of contiguous read coverage in the
genome even with- out graph construction and traversal37,51,52. Alternative
transcription start and stop sites can also be inferred based on the 5â€² cap or
poly(A) signals (if cap- or end- specific experimental protocols are used)51,53.
However, complications arise owing to the gene-dense nature of
these genomes. Many genes overlap, resulting in adja- cent genes being assembled
into one transcript, even though they are not from a polycistronic RNA. Strand-
specific RNA-seq has successfully been used to separate adjacent overlapping
genes from opposite strands in the genome51,52. Overlapping genes that are
transcribed from the same strand and that also have comparable expres- sion
levels cannot easily be separated without using cap- or end-specific RNA-seq.

In summary, reference-based assembly is generally preferable for cases in which
a high-quality reference genome already exists.

Meanwhile, experimental RNA-seq and sequencing protocols are continually
improving and should greatly reduce the informatics challenges.  RNA-seq reads
from third-generation sequencers, such as PacBio70, are longer (up to several
kilobases). PacBio sequencers are capable of sequencing a single transcript to
its full length in a single read. If this technology reaches a throughput that
is comparable to the second-generation technologies, then the need for
transcriptome assembly will probably be eliminated. Hopefully, the future of
transcriptome assembly will be ``no assembly required''.

What is the sequencing depth?

\fixme{I might have to try different filters of short reads to see how those
affect the transcript prediction.}

<<manuscript,echo=FALSE,results=hide>>=
# not run when smutans R package is installed
# Use this when smutans R package is being tested.
library(DESeq)
library(GenomicRanges)
library(plotrix)
source("smutans/R/class_and_slots.R")
source("smutans/R/core.R")
source("smutans/R/methods.R")
load("smutans/data/smutansData.tx.RData")
load("smutans/data/smutansData.txGenes.RData")
load("smutans/data/smutansData.txPileup.RData")
load("smutans/data/smutans.feature.genes.RData")
load("smutans/data/smutans.genes.criteria.RData")
load("smutans/data/smutans.go.genes.RData")
load("smutans/data/smutans.cat.desc.RData")
load("smutans/data/smutansGenes.RData")
@
<<manuscript,eval=FALSE,echo=FALSE,results=hide>>=
# not run without smutans R package installed
# Remove eval=FALSE when the R package is installed
library(smutans)
load(smutansData.tx)
load(smutansData.txGenes)
load(smutansData.txPileup)
load(smutans.feature.genes)
data(smutans.genes.criteria)
data(smutans.go.genes)
data(smutans.cat.desc)
data(smutansGenes)
@
<<echo=FALSE,result=hide>>=
grTx <- smutansData.tx[elementMetadata(smutansData.tx)$state > 1]
mtch2 <- findOverlaps(smutansData.txGenes,grTx)
grNoncoding <- gaps(smutansData.txGenes)
# A bug in GenomicRanges: the last segment is the while genome. I removed it
# from the noncoding ranges.
grNoncoding <- grNoncoding[-length(grNoncoding)] 
mtchNoncoding <- findOverlaps(grNoncoding,grTx)
# Expressed transcripts without annotated genes
grTxWGenes <- grTx[unique(matchMatrix(mtch2)[,2])]
grTxWOGenes <- grTx[-unique(matchMatrix(mtch2)[,2])]
# Find transcripts with genes with conflicted strands
grGenesPlus <- smutansData.txGenes[strand(smutansData.txGenes)=='+']
grGenesMnus <- smutansData.txGenes[strand(smutansData.txGenes)=='-']
mtchPlus <- findOverlaps(grGenesPlus,grTxWGenes)
mtchMnus <- findOverlaps(grGenesMnus,grTxWGenes)
txWithConflictedGenes <- intersect(matchMatrix(mtchPlus)[,2],matchMatrix(mtchMnus)[,2])
grTx2 <- grTxWGenes[-txWithConflictedGenes] 
# Lengths of 5' and 3' UTRs
mtchPlus <- findOverlaps(grGenesPlus,grTx2)
mtchMnus <- findOverlaps(grGenesMnus,grTx2)
length5UTR <- c()
length3UTR <- c()
for (i in unique(matchMatrix(mtchPlus)[,"subject"])) {
  x <- matchMatrix(mtchPlus)[matchMatrix(mtchPlus)[,"subject"]==i,"query"]
  y <- min(start(grGenesPlus[x])) - start(grTx2[i])
  z <- end(grTx2[i]) - max(end(grGenesPlus[x])) 
  length5UTR <- c(length5UTR,y)
  length3UTR <- c(length3UTR,z)
}
for (i in unique(matchMatrix(mtchMnus)[,"subject"])) {
  x <- matchMatrix(mtchMnus)[matchMatrix(mtchMnus)[,"subject"]==i,"query"]
  y <- min(start(grGenesMnus[x])) - start(grTx2[i])
  z <- end(grTx2[i]) - max(end(grGenesMnus[x])) 
  length5UTR <- c(length5UTR,z)
  length3UTR <- c(length3UTR,y)
}
# Correlation of expression levels of adjacent genes.
pileup.x <- lapply(ranges(smutansData.txGenes), function(x) seqselect(smutansData.txPileup, x))
pileup.y <- mapply(function(x,y) sum(x)/y, x=pileup.x, y=width(smutansData.txGenes))
elementMetadata(smutansData.txGenes)["score"] <- pileup.y
mtch3 <- findOverlaps(smutansData.txGenes,grTx2)
x <- rle(matchMatrix(mtch3)[,"subject"])
comp.exp.two.genes <- c()
for (i in x$values[x$lengths > 1]) {
  y <- combn(matchMatrix(mtch3)[,"query"][matchMatrix(mtch3)[,"subject"]==i],2)
  for (j in 1:dim(y)[2]) {
    comp.exp.two.genes <- cbind(comp.exp.two.genes,as.matrix(elementMetadata(smutansData.txGenes)["score"][y[,j],]))
  }
}
# Correlation of expression levels of adjacent genes between transcripts.
comp.exp.two.genes.between.transcript <- c()
for (i in 1:length(x$values)) {
  if (i == length(x$values)) {
    break
  }
  y1 <- matchMatrix(mtch3)[,"query"][matchMatrix(mtch3)[,"subject"]==x$values[i]]
  i2 <- i + 1
  y2 <- matchMatrix(mtch3)[,"query"][matchMatrix(mtch3)[,"subject"]==x$values[i2]]
  l <- as.matrix(do.call(expand.grid, list(y1,y2)))
  for (j in 1:dim(l)[1]) {
    comp.exp.two.genes.between.transcript <- cbind(comp.exp.two.genes.between.transcript,as.matrix(elementMetadata(smutansData.txGenes)["score"][l[j,],]))
  }
}
# plot(log10(comp.exp.two.genes.between.transcript[1,]), log10(comp.exp.two.genes.between.transcript[2,]))
# cor(log10(comp.exp.two.genes.between.transcript[1,]), log10(comp.exp.two.genes.between.transcript[2,]))
@

\subsection{My Results}
% Statistics of transcripts, and genes.
The input RNA-seq data is a list of site-wise non-negative integers that
represents a number of short reads mapped to a particular genomic position
\citet{Martin2010}, which is called a RNA-seq ``coverage'' data. We designated a
gene as an expressed if the gene belongs to a transcript with at least C2
level.
The number of predicted expressed transcripts was
\Sexpr{sum(elementMetadata(smutansData.tx)$state > 1)}
(\Sexpr{as.integer(sum(elementMetadata(smutansData.tx)$state > 1)/length(smutansData.tx)*100)}\%)
out of 
\Sexpr{length(smutansData.tx)} 
fragments.
Of the total \Sexpr{length(smutansData.txGenes)} of genes,
\Sexpr{length(matchMatrix(mtch2)[,1])}
(\Sexpr{as.integer(length(matchMatrix(mtch2)[,1])/length(smutansData.txGenes)*100)}\%) 
were designated as expressed. 
Genes that were not designated as expressed could be due to false gene
annotations or possibility of true negative expression under the experimental
conditions. 
Figure \ref{figLength} shows the length distribution of expressed and
not-expressed genes. 
Of the \Sexpr{length(grTx)} expressed transcripts, 
\Sexpr{length(grTxWOGenes)} contained no annotated genes.
Figure \ref{figTranscriptLength} shows the length distribution of expressed
transcripts with annotated coding genes and unannotated regions.
Unannotated regions can include potential non-coding regions and new
coding genes.  Transciprts with unannotated regions appeared to be shorter than those
with coding genes. Most of the unannotated regions with length being near 360 bp
could be non-coding regions, and the rest of the unannotated regions with length
being near 1000 bp could be new coding genes. Some of the unannotated genes
could be pseudogene regions.
Because we used a RNA-seq technique without considering strand-specific, we
could not tell the strandness of the predicted transcripts. Of
the \Sexpr{length(grTx)} expressed transcripts, 
\Sexpr{length(txWithConflictedGenes)}
included annotated genes of both strands
(\Sexpr{as.integer(length(txWithConflictedGenes)/length(grTx)*100)}\%).
We filter out the transcripts with genes with conflicted both strands leaving
\Sexpr{length(grTx2)} expressed transcripts. 
Figure \ref{figUTRLength} shows length distributions of 5' and 3' UTR of the
expressed transcripts without genes with conflicted strands.
Many transcripts appeared to start and end near the boundaries of annotated genes.
Gene expression levels for adjacent genes within expressed transcripts were more
correlated than those between adjacent transcripts 
(Figure \ref{figExpressionTwoGene}).
% Question: What are the promoter scores? 

Name the transcripts.
I found xx transcripts. Among them xx \% were mono-cistronic operons. xx \% were
poly-cistronic operons. xx transcripts could not defined because RNA-seq short
reads were not precise. Transripts were assigned with functional categories, and
differentially expressed transcripts were found. Transcripts and functional
categories: how can I assess that transcripts consist of genes with similar
functional cateogires. 

% Small RNAs predicted
Although we aimed at sequencing messenger RNAs, we expected to find regulatory
small RNAs in the RNA-seq data because we attempted to deplete structrual
ribosomal RNAs. \program{RNAz} is a software package for detection of structural
noncoding RNAs using comparative genomic data \citep{Gruber2010}. The method has
been applied to different genomic data (CITE). We applied the method to predict
potential non-coding small RNAs in \textit{S.\ mutans} as a supplement to
evidence for small RNAs provided by the RNA-seq data. With increased evidence of
putative small RNAs we could proceed to further experimentally verify
potentially important small RNAs.  We first predicted
\Sexpr{length(colnames(smutans.genes.criteria))-1} putative small RNAs. Using
\program{RNAplex} \citep{Tafer2008} and \program{RNAplfold} \citep{Bernhart2006}
we predicted target genes of the putative small RNAs.  Table \ref{tab:goseq}
shows functional categories enriched with target genes for predicted small RNAs.
We found that a few of the putative small RNAs were structured with stable stems
(SHOW FIGURES). 

Name the small RNAs.
RNAz predicted a few small RNAs. Some of them were differentially expressed.
Some functional categories were enriched with target genes of the predicted
genes. 



% UCSC genome browser resources
We created new tracks for our recently released {\em Streptococcus} Genome
Browser (http://strep-genome.bscb.cornell.edu) that summarized the results of
this \textit{S.\ mutans} transcriptome analysis alongside known genes, predicted
small RNAs, predicted target genes and predicted transcripts. These tracks can
be used to inspect loci of interest and to compare the results of different
RNA-seq data sets. They can also be queried and intersected with other tracks
using the UCSC Table Browser.

<<echo=FALSE,result=hide>>=
#glucoseOnly <- newSmutans( smutansGenes, title="Glucose Only" )
#glucoseOnly <- smutans.de2( glucoseOnly, condition="glucose", 
#                            condA="ua159", condB="tw1" )
#save(glucoseOnly, file=file.path("glucoseOnly.RData"))
load("glucoseOnly.RData")
smutans.de2List( glucoseOnly, "glucoseOnly.csv" )
tableGlucoseOnly <- smutans.de2List( glucoseOnly, qval=1e-3 )

# Gene descriptions
geneDescription <- read.delim("feature-genome.out-geneonly.txt",head=FALSE,sep="\t")
geneDescription <- geneDescription[,c(4,7)]
colnames(geneDescription) <- c("Gene ID","Gene Description")
# Functional classes
functionalDescription <- read.csv("feature-genome.out-geneonly-oralgen-functional.txt",head=FALSE)
functionalDescription <- functionalDescription[,c(2,12)]
colnames(functionalDescription) <- c("Gene ID","Functional Class")
# Find potential differentially expressed genes, and add gene descriptions, and
# functional classes
x <- smutans.de2List( glucoseOnly, qval=1e-3 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","UA159","TW1","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
tableGlucoseOnly <- x
@

% Differential expression study
We sequenced 
\Sexpr{length(colnames(counts(smutansGenes)))}
RNA samples for studying differential gene expression to better understand
biological processes. Two independent factors of the biological processes
includes the effect of CcpA and the treatment of monosaccharides (glucose versus
galactose). The two factors results four pair-wise comparisons. We compared two
conditions of one factor with the other factor being fixed at one condition. For
example, we could compare UA159 wild-type and TW1 mutant-type strains under the
condition of glucose-grown cells to find differentially expressed associatively
with deactivation of \textit{ccpA} genes. We performed a similar comparison with
galactose-grown cells. While we replicated four biological experiments for the
condition combination of UA159 wilde-type grown in galactose, we had three for
the other three conditions. Because of limited resources of RNA deep sequencing,
we considered three replicates as a maximum number to but still adequately
detect differentially expressed genes. 

We list differentially expressed genes in the comparison of UA159 wild-type and
TW1 mutant-type in Table \ref{tabGlucoseOnly}.  
Of \Sexpr{nrow(counts(smutansGenes))} genes, 
\Sexpr{nrow(tableGlucoseOnly)} genes were found to be differentially expressed
at FDR of 0.1\%. 
We confirmed that the most upregulated genes in the TW1 mutant-type encoded the
components of the PDH enzyme complex (SMU.1421, SMU.1422, SMU.1423, and
SMU.1424).  
Figure \ref{figPieGlucoseOnly} shows the proportions of functional categories 
of the differentially expressed genes. Notably, we found
\Sexpr{length(grep("Energy metabolism",tableGlucoseOnly[,"Functional Class"]))}
genes that were involved in energy metabolism. Sugars and fermentation
subclasses were profound in energy metabolism. We also found 
\Sexpr{length(grep("PTS",tableGlucoseOnly[,"Functional Class"]))}
genes in PTS class of signal transduction. Another subclass, 
``carbohydrates, organic alcohols, and acids,'' in
``transport and binding proteins'' was following next in the number of
differentially expressed genes. We also found  
\Sexpr{length(grep("hypothetical",tableGlucoseOnly[,"Gene Description"]))}
genes annotated as hypothetical, which might constitute the unknown functional
class. These results were consistent with those of \citet{Abranches2008}.
Profoundly upregulated genes in response to the lack of CcpA were
putative Clp proteinase (SMU.1425), 
branched-chain alpha-keto acid dehydrogenase subunit E2 (SMU.1421), and
putative transcriptional regulator (SMU.1599).
Down-regulated genes include cytoplasmic alpha-amylase (SMU.1590), 
beta-D-fructosyltransferase (SMU.2028), and
putative glycerol uptake facilitator protein (SMU.396).
Fructosyltransferase converts sucrose to fructose.

<<echo=FALSE,result=hide>>=
#galactoseOnly <- newSmutans( smutansGenes, title="Galactose Only" )
#galactoseOnly <- smutans.de2( galactoseOnly, condition="galactose", 
#                              condA="ua159", condB="tw1" )
#save(galactoseOnly, file=file.path("galactoseOnly.RData"))
load("galactoseOnly.RData")
x <- smutans.de2List( galactoseOnly, qval=1e-3 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","UA159","TW1","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
tableGalactoseOnly <- x
@
We found that differentially expressed genes in galactose-grown cells were
smaller than those in glucose-grown cells, which was different from the result
of \cite{Abranches2008}. 
Energy metabolism was the most
profoundly differentially expressed (fig.\ \ref{figPieGalactoseOnly}).
Subclasses pyruvate family in amino acid biosynthesis, 
and degradation of proteins, peptides, and glycopeptides in protein fate
were following the next. Almost quarter of the genes were unknown.
Down-regulated genes included cytoplasmic alpha-amylase (SMU.1590), and
beta-D-fructosyltransferase (SMU.2028)
(table \ref{tabGalactoseOnly}).

<<echo=FALSE,result=hide>>=
#ua159only <- newSmutans( smutansGenes, title="UA159 Only" )
#ua159only <- smutans.de2( ua159only, type="ua159", 
#                          condA="glucose", condB="galactose" )
#save(ua159only, file=file.path("ua159only.RData"))
load("ua159only.RData")
x <- smutans.de2List( ua159only, qval=1e-10 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","Glucose","Galactose","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
tableUA159Only <- x
@
Figure \ref{figPieUA159Only} shows proportons of functional categories that
included differentially expressed genes between glucose-grown and
galactose-grown UA159 wild-type cells. Qualitatively, energy metabolism and PTS
classes were profound. This was silmilar to the comparison between UA159
wild-type and TW1 mutant-type grown in glucose. More than quarter of the
differentially expressed genes were unknown, unassigned, or hypothetical.
Consecutive genes (SMU.1488 to SMU.1496) were upregulated in 
catabolite-derepressing conditions or in galactose-grown cells
(\fixme{Are these the genes of the lactose (lac) operon?})
(table \ref{tabUA159Only}).  Another set of genes (SMU.886 to SMU.889) were also
upregulated (\fixme{Are these called galKTE?}).  
Genes from SMU.1535 to SMU.1538 were upregulated. 
Of the \Sexpr{nrow(tableUA159Only)} differentially expressed genes at
$1 \times 10^{-10}$ cutoff of multiple testing adjusted p-value, 
\Sexpr{length(grep("hypothetical",tableUA159Only[,"Gene Description"]))} were
hypothetical, and most of them were down-regulated. 
A few of the down-regulated genes were annotated.

We also found sugar metabolism and PTS signal transduction to be profound in
differential expression in the comparison of glucose-grown and galactose-grown
TW1 mutant-type cells figure \ref{figPieTW1Only}. Many of the genes were also
hypothetical. We found similar genes differentially expressed in TW1 mutant-type
cells to those in UA159 wild-type (table \ref{tabTW1Only}).  

\fixme{We may need dot plots where x-axis represents fold changes, and y-axis
functional classes. Dots may be colored by p-values. This would show most of
information in one plot. We could label some of dots to emphasixe. Done?}
\fixme{Use KEGG for functional category study. Done?}
\fixme{Display results in UCSC genome browser. Done?}
\fixme{Use two factor analysis. Done?}

%


Deletion of \textit{ccpA} resulted in substantial relief of the CCR of
\textit{fruA} expression exerted by both glucose and fructose. 
When \citet{Wen2002} deleted CRE sequencs in the \textit{fruA} promoter region, CCR was
almost completely alleviated.  They also showed that 
CcpA directly interacted with the \textit{fruA} promoter region. 

The role of CcpA in gene expression during growth on glucose. 

\textit{S.\ mutans} UA159 and TW1 grown under catabolite-repressing conditions
(glucose), and catabolite-derepressing conditions (galactose). Under
glucose-only condition we found xx number of differentially expressed genes. 

\section{Discussion}
We used RNA-seq data to delineate transcript boundaries. Although this map of
transcripts does not precisely tell us where transcripts start and end, it
informs us of which open reading frames can be transcribed together. Potentially
polycistronic operons are defined. We focused on interpretations of operon in
light of CcpA regulations. 

\section{Acknowledgments}
This study was funded by the National Institute of Allergy and Infectious
Disease, US National Institutes of Health, under grant number AI073368-01A2
(to M.J.S.\ and A.S).  Additional support was provided by National Science
Foundation CAREER Award DBI-0644111 and a David and Lucile Packard
Fellowship for Science and Engineering (to A.S.).  
We thank Christopher Browngardt for assistance with labelling functional classes
of \textit{S.\ mutans} genes.

\renewcommand*{\refname}{Literature Cited}
\bibliographystyle{../../latex/bst/mbe}
\bibliography{siepel-mutans}
\clearpage{}

\section{Tables}

%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf Gene Ontology enrichments for target genes of putative small RNAs.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{lcccl}
\hline 
small RNA & $q^a$ & count & GO & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.mannwhitney( qval=0.05,
                     genes.criteria=smutans.genes.criteria,
                     go.genes=smutans.go.genes,
                     cat.desc=smutans.cat.desc )
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:goseq}
\end{table}
%-----------------------------------------------------------------------------

\clearpage{}

\section{Figures}

<<figLength,fig=TRUE,echo=FALSE>>=
l <- list(expressed=width(smutansData.txGenes[matchMatrix(mtch2)[,1]]), 
          notexpressed=width(smutansData.txGenes[-matchMatrix(mtch2)[,1]]))
bin <- c(as.integer(seq.int(0, 3000, length.out=15)), 8500)
multhist(l,breaks=bin)
@
\begin{figure}
\centering
\includegraphics[width=.9\textwidth]{manuscript-figLength}
\caption{{\bf Length distribution for expressed and not-expressed genes.}
Darker gray
bars represent frequency of expressed genes, and lighter gray ones that of
not-expressed genes. The last bin sums from 3000 bp to 8500 bp.
}
\label{figLength}
\end{figure}

<<figTranscriptLength,fig=TRUE,echo=FALSE>>=
l <- list(coding=width(grTxWGenes), 
          noncoding=width(grTxWOGenes))
bin <- c(as.integer(seq.int(0, 10000, length.out=15)), 30000)
multhist(l,breaks=bin)
@
\begin{figure}
\centering
\includegraphics[width=.9\textwidth]{manuscript-figTranscriptLength}
\caption{{\bf Length distribution for expressed transcripts with annotated genes
and unannotated genes.}
Darker gray bars represent frequency of expressed transcripts with annotated
genes, and lighter gray ones that of without annotated genes that can be
potential non-coding regions. The last bin sums from 10000 bp to 30000 bp.
}
\label{figTranscriptLength}
\end{figure}

<<figUTRLength,fig=TRUE,echo=FALSE>>=
l <- list(five=length5UTR, three=length3UTR)
bin <- c(as.integer(seq.int(0, 300, length.out=19)), 6500)
multhist(l,breaks=bin)
@
\begin{figure}
\centering
\includegraphics[width=.9\textwidth]{manuscript-figUTRLength}
\caption{{\bf Length distribution for 5' and 3' UTR.}
Darker gray bars represent frequency of 5' UTR lengths, and lighter gray 3' UTR
lengths. The last bin sums from 300 bp to 6500 bp.
\fixme{Check the too long UTRs because they do not make sense.}
}
\label{figUTRLength}
\end{figure}

<<figExpressionTwoGene,fig=TRUE,echo=FALSE>>=
plot( log10(comp.exp.two.genes[1,]), log10(comp.exp.two.genes[2,]),
      xlab="Expression of first Gene in log10",
      ylab="Expression of second Gene in log10",
      pch=19,cex=0.5)
abline(b=1,lty=2)
@
<<figExpressionTwoGeneBetweenTranscript,fig=TRUE,echo=FALSE>>=
plot( log10(comp.exp.two.genes.between.transcript[1,]),
      log10(comp.exp.two.genes.between.transcript[2,]),
      xlab="Expression of first Gene in log10",
      ylab="Expression of second Gene in log10",
      pch=19,cex=0.5)
abline(b=1,lty=2)
@
\begin{figure}
\centering
\includegraphics[width=.6\textwidth]{manuscript-figExpressionTwoGene}
\includegraphics[width=.6\textwidth]{manuscript-figExpressionTwoGeneBetweenTranscript}
\caption{{\bf Scatter plot of expression levels of pairs of adjacent genes
located in the same transcripts.}
(Top) Two genes in the same tranascripts are plotted using their expression levels in
log10 scale.
(Bottom) Two genes in different tranascripts are plotted using their expression levels in
log10 scale.
}
\label{figExpressionTwoGene}
\end{figure}

\end{document}
