\documentclass{article}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{color} 
\usepackage{setspace} 
\usepackage{rotating}
\usepackage{natbib}
\usepackage{xr}
\usepackage{../../latex/sty/genres}
\usepackage{url}
\usepackage{longtable}

\externaldocument{manuscript}

% Text layout
\oddsidemargin 0in
\evensidemargin 0in
\topmargin -.5in
\textwidth 6.5in
\textheight 9in
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

\SweaveOpts{keep.source=TRUE,eps=FALSE,include=FALSE,width=6,height=6}
\newcommand{\Rpackage}[1]{R package \textit{#1}}
\newcommand{\program}[1]{{\texttt{#1}}}
\newcommand{\fixme}[1]{{\textbf{Fixme:} \textit{\textcolor{blue}{#1}}}}
\newcommand{\comment}[1]{}

\makeatother
\pagestyle{myheadings}
\markright{Transcriptome of Streptococcus mutans}

% 

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}

\makeatother

\renewcommand{\thefootnote}{\textit{\alph{footnote}}}

% Leave date blank
\date{}

\pagestyle{myheadings}
%% ** EDIT HERE **
\markright{Transcriptome of Streptococcus mutans}

\begin{document}

\setcounter{secnumdepth}{-1} 

\noindent \begin{LARGE}{\bf Defining Transcriptome using RNA Sequencing
 \\[0.7ex] 
of {\em Streptococcus mutans}}\end{LARGE}\\[2ex]
{\bf 
Sang Chul Choi, Charles D.\ Danko, Melissa J.\ Hubisz, \\
Paulina Dmitrievna Pavinski Bitar, \\
Robert A.\ Burne, Michael J.\ Stanhope, and Adam Siepel}%\\[0.5ex]

\tableofcontents

\listoftables

\listoffigures

\doublespacing

\clearpage

<<echo=FALSE,results=hide>>=
# not run when smutans R package is installed
# Use this when smutans R package is being tested.
library(DESeq)
library(goseq)
library(GenomicRanges)
library(plotrix)
library(xtable)
library(RColorBrewer)
source("smutans/R/class_and_slots.R")
source("smutans/R/core.R")
source("smutans/R/methods.R")
load("smutans/data/smutansData.tx.RData")
load("smutans/data/smutansData.txGenes.RData")
load("smutans/data/smutansData.txPileup.RData")
load("smutans/data/smutans.feature.genes.RData")
load("smutans/data/smutans.genes.criteria.RData")
load("smutans/data/smutans.go.genes.RData")
load("smutans/data/smutans.cat.desc.RData")
load("smutans/data/smutans.keg.genes.RData")
load("smutans/data/smutans.keg.desc.RData")
load("smutans/data/smutansGenes.RData")
@
<<eval=FALSE,echo=FALSE,results=hide>>=
# not run without smutans R package installed
# Remove eval=FALSE when the R package is installed
library(smutans)
load(smutansData.tx)
load(smutansData.txGenes)
load(smutansData.txPileup)
load(smutans.feature.genes)
data(smutans.genes.criteria)
data(smutans.go.genes)
data(smutans.cat.desc)
data(smutans.keg.genes)
data(smutans.keg.desc)
data(smutansGenes)
@

\section{Supplementary Text}
A
\subsection{Summary statistics for recombinant trees}
\citep{Gorke2008}

\enlargethispage{\baselineskip}

\renewcommand*{\refname}{Literature Cited}
\bibliographystyle{../../latex/bst/mbe}
\bibliography{siepel-mutans}

\clearpage{}\setcounter{figure}{0}
\setcounter{table}{0}

\makeatletter 
\renewcommand{\thefigure}{S\@arabic\c@figure} 
\renewcommand{\thetable}{S\@arabic\c@table} 
\renewcommand{\figurename}{Supplementary Figure}
\renewcommand{\tablename}{Supplementary Table}
\makeatother

%\section{Supplementary Tables}

<<echo=FALSE,results=hide>>=
geneDescription <- read.delim("feature-genome.out-geneonly.txt",head=FALSE,sep="\t")
geneDescription <- geneDescription[,c(4,7)]
colnames(geneDescription) <- c("Gene ID","Gene Description")

functionalDescription <- read.csv("feature-genome.out-geneonly-oralgen-functional.txt",head=FALSE)
functionalDescription <- functionalDescription[,c(2,12)]
colnames(functionalDescription) <- c("Gene ID","Functional Class")
@

<<echo=FALSE,results=tex>>=
#glucoseOnly <- newSmutans( smutansGenes, title="Glucose Only" )
#glucoseOnly <- smutans.de2( glucoseOnly, condition="glucose", 
#                            condA="ua159", condB="tw1" )
#save(glucoseOnly, file=file.path("glucoseOnly.RData"))
load("glucoseOnly.RData")
x <- smutans.de2List( glucoseOnly, qval=1e-3 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","UA159","TW1","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x.big <- xtable( x[order(x[,"p-value"]),],
                 display=c("s","s","d","d","f","e","s"),
                 digits=c(0,0,0,0,1,0,0),
                 align=c('l','l','r','r','r','r','p{3.5in}'),
                 label='tabGlucoseOnly',
                 caption='List of genes for the comparison UA159 and TW1 under glucose-grown cells',
               )
print( x.big,
       tabular.environment='longtable',
       caption.placement="top",
       include.rownames=FALSE,
       floating=FALSE) 
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
@
\clearpage{}

<<figPieGlucoseOnly,fig=TRUE,echo=FALSE,width=6,height=8>>=
par(mai=c(0,0,0,0),fig=c(0,1,0,1),mfrow=c(2,1))
#par(mfrow=c(2,1))
trim <- function (x) gsub("^\\s+|\\s+$", "", x)
y <- x[,"Functional Class"]
z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
z <- table(as.factor(sort(unlist(z))))

slices <- as.vector(z)
lbls <- names(z)
pct <- round(slices/sum(slices)*100)
c1 <- brewer.pal(n=9,name="Set1")
pie(z,labels=pct,main="",col=c1,clockwise=TRUE)
plot(0,0,axes=FALSE,frame.plot=FALSE,xlab="",ylab="",cex=0)
legend(-0.8,1,names(z),bty='n',cex=0.7,fill=c1)
@
\begin{figure}
\centering
\includegraphics[width=0.9\textwidth]{supplementary-figPieGlucoseOnly}
\caption{{\bf Distribution of functional classes of differntially expressed genes
in comparing UA159 wild-type and TW1 mutant-type grown in glucose.}
Labels for functional categories are colored with nine different colors. The
first label with red color, 
``Biosynthesis of cofactors, prosthetic groups, and carriers; Thiamine'', starts
at the top, and next labels read to the clockwise. The numbers circling the pie
chart is in per cent.
}
\label{figPieGlucoseOnly}
\end{figure}

\clearpage{}
<<echo=FALSE,results=tex>>=
#galactoseOnly <- newSmutans( smutansGenes, title="Galactose Only" )
#galactoseOnly <- smutans.de2( galactoseOnly, condition="galactose", 
#                              condA="ua159", condB="tw1" )
#save(galactoseOnly, file=file.path("galactoseOnly.RData"))
load("galactoseOnly.RData")
x <- smutans.de2List( galactoseOnly, qval=1e-3 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","UA159","TW1","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x.big <- xtable( x[order(x[,"p-value"]),],
                 display=c("s","s","d","d","f","e","s"),
                 digits=c(0,0,0,0,1,0,0),
                 align=c('l','l','r','r','r','r','p{3.5in}'),
                 label='tabGalactoseOnly',
                 caption='List of genes for the comparison UA159 and TW1 under galactose-grown cells',
               )
print( x.big,
       tabular.environment='longtable',
       caption.placement="top",
       include.rownames=FALSE,
       floating=FALSE) 
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
@
\clearpage{}

<<figPieGalactoseOnly,fig=TRUE,echo=FALSE,width=6,height=8>>=
par(mai=c(0,0,0,0),fig=c(0,1,0,1),mfrow=c(2,1))
#par(mfrow=c(2,1))
trim <- function (x) gsub("^\\s+|\\s+$", "", x)
y <- x[,"Functional Class"]
z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
z <- table(as.factor(sort(unlist(z))))

slices <- as.vector(z)
lbls <- names(z)
pct <- round(slices/sum(slices)*100)
c1 <- brewer.pal(n=9,name="Set1")
pie(z,labels=pct,main="",col=c1,clockwise=TRUE)
plot(0,0,axes=FALSE,frame.plot=FALSE,xlab="",ylab="",cex=0)
legend(-0.8,1,names(z),bty='n',cex=0.7,fill=c1)
@
\begin{figure}
\centering
\includegraphics[width=0.9\textwidth]{supplementary-figPieGalactoseOnly}
\caption{{\bf Distribution of functional classes of differntially expressed genes
in comparing UA159 wild-type and TW1 mutant-type grown in galactose.}
Labels for functional categories are colored with nine different colors. The
first label with red color, 
``Amino acid biosynthesis; Glutamate family'', starts
at the top, and next labels read to the clockwise. The numbers circling the pie
chart is in per cent.
}
\label{figPieGalactoseOnly}
\end{figure}


\clearpage{}
<<echo=FALSE,results=tex>>=
#ua159only <- newSmutans( smutansGenes, title="UA159 Only" )
#ua159only <- smutans.de2( ua159only, type="ua159", 
#                          condA="glucose", condB="galactose" )
#save(ua159only, file=file.path("ua159only.RData"))
load("ua159only.RData")
x <- smutans.de2List( ua159only, qval=1e-10 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","Glucose","Galactose","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x.big <- xtable( x[order(x[,"p-value"]),],
                 display=c("s","s","d","d","f","e","s"),
                 digits=c(0,0,0,0,1,0,0),
                 align=c('l','l','r','r','r','r','p{3.5in}'),
                 label='tabUA159Only',
                 caption='List of genes for the comparison of UA159 wild-type grown with glucose and galactose. Note that we used 1e-10 as the p-value cutoff.',
               )
print( x.big,
       tabular.environment='longtable',
       caption.placement="top",
       include.rownames=FALSE,
       floating=FALSE) 
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
@
\clearpage{}

<<figPieUA159Only,fig=TRUE,echo=FALSE,width=6,height=8>>=
par(mai=c(0,0,0,0),fig=c(0,1,0,1),mfrow=c(2,1))
trim <- function (x) gsub("^\\s+|\\s+$", "", x)
y <- x[,"Functional Class"]
z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
z <- table(as.factor(sort(unlist(z))))

slices <- as.vector(z)
lbls <- names(z)
pct <- round(slices/sum(slices)*100)
c1 <- brewer.pal(n=9,name="Set1")
pie(z,labels=pct,main="",col=c1,clockwise=TRUE)
plot(0,0,axes=FALSE,frame.plot=FALSE,xlab="",ylab="",cex=0)

legend(-0.8,1,names(z),bty='n',cex=0.7,fill=c1)
#legend(-1.1,1.1,names(z[1:27]),bty='n',cex=0.7,fill=c1)
#legend(0,0.4,names(z[-c(1:27)]),bty='n',cex=0.7,fill=c1)

@
\begin{figure}
\centering
\includegraphics[width=0.9\textwidth]{supplementary-figPieUA159Only}
\caption{{\bf Distribution of functional classes of differntially expressed genes
in comparing UA159 wild-type grown in glucose and galactose.}
Labels for functional categories are colored with nine different colors. The
first label with red color, 
``Amino acid biosynthesis; Glutamate family'', starts
at the top, and next labels read to the clockwise. The numbers circling the pie
chart is in per cent.
}
\label{figPieUA159Only}
\end{figure}

\clearpage{}
<<echo=FALSE,results=tex>>=
#tw1only <- newSmutans( smutansGenes, title="TW1 Only" )
#tw1only <- smutans.de2( tw1only, type="tw1", 
#                          condA="glucose", condB="galactose" )
#save(tw1only, file=file.path("tw1only.RData"))
load("tw1only.RData")
x <- smutans.de2List( tw1only, qval=1e-3 )
x <- x[,c("id","baseMeanA","baseMeanB","log2FoldChange","padj")]
colnames(x) <- c("Gene ID","Glucose","Galactose","fold", "p-value")
x <- merge(x, geneDescription, by.x = "Gene ID", by.y = "Gene ID")
x.big <- xtable( x[order(x[,"p-value"]),],
                 display=c("s","s","d","d","f","e","s"),
                 digits=c(0,0,0,0,1,0,0),
                 align=c('l','l','r','r','r','r','p{3.5in}'),
                 label='tabTW1Only',
                 caption='List of genes for the comparison of TW1 mutant-type grown with glucose and galactose ',
               )
print( x.big,
       tabular.environment='longtable',
       caption.placement="top",
       include.rownames=FALSE,
       floating=FALSE) 
x <- merge(x, functionalDescription, by.x = "Gene ID", by.y = "Gene ID")
@
\clearpage{}

<<figPieTW1Only,fig=TRUE,echo=FALSE,width=6,height=8>>=
par(mai=c(0,0,0,0),fig=c(0,1,0,1),mfrow=c(2,1))
trim <- function (x) gsub("^\\s+|\\s+$", "", x)
y <- x[,"Functional Class"]
z <- lapply(y, function(x) trim(unlist(strsplit(as.character(x),"\\|\\|"))))
z <- table(as.factor(sort(unlist(z))))

slices <- as.vector(z)
lbls <- names(z)
pct <- round(slices/sum(slices)*100)
c1 <- brewer.pal(n=9,name="Set1")
pie(z,labels=pct,main="",col=c1,clockwise=TRUE)
plot(0,0,axes=FALSE,frame.plot=FALSE,xlab="",ylab="",cex=0)
legend(-0.8,1,names(z),bty='n',cex=0.7,fill=c1)
@
\begin{figure}
\centering
\includegraphics[width=0.9\textwidth]{supplementary-figPieTW1Only}
\caption{{\bf Distribution of functional classes of differntially expressed genes
in comparing TW1 mutant-type grown in glucose and galactose.}
Labels for functional categories are colored with nine different colors. The
first label with red color, 
``Amino acid biosynthesis; Histidine family'', starts
at the top, and next labels read to the clockwise. The numbers circling the pie
chart is in per cent.
}
\label{figPieTW1Only}
\end{figure}

\clearpage{}
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf Gene Ontology (GO) enrichments for differentially expressed
genes in the comparison UA159 and TW1 grown in glucose.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq ( glucoseOnly, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc )
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:goseqGlucoseOnly}
\end{table}
%-----------------------------------------------------------------------------

\clearpage{}
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf Gene Ontology (GO) enrichments for differentially expressed
genes in the comparison UA159 and TW1 grown in galactose.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq ( galactoseOnly, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc )
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:goseqGalactoseOnly}
\end{table}
%-----------------------------------------------------------------------------

\clearpage{}
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf Gene Ontology (GO) enrichments for differentially expressed
genes in the comparison of glucose-grown and galactose-grown UA159 wild-type
cells.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq ( ua159only, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc )
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:goseqUA159Only}
\end{table}
%-----------------------------------------------------------------------------

\clearpage{}
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf Gene Ontology (GO) enrichments for differentially expressed
genes in the comparison of glucose-grown and galactose-grown TW1 mutant-type
cells.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq ( tw1only, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.go.genes,
                   cat.desc=smutans.cat.desc )
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:goseqTW1Only}
\end{table}
%-----------------------------------------------------------------------------

\clearpage{}
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf KEGG enrichments for differentially expressed
genes in the comparison UA159 and TW1 grown in glucose.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq ( glucoseOnly, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.keg.genes[c(4,3)],
                   cat.desc=smutans.keg.desc)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:kegseqGlucoseOnly}
\end{table}
%-----------------------------------------------------------------------------

\clearpage{}
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf KEGG enrichments for differentially expressed
genes in the comparison UA159 and TW1 grown in galactose.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq ( galactoseOnly, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.keg.genes[c(4,3)],
                   cat.desc=smutans.keg.desc)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:kegseqGalactoseOnly}
\end{table}
%-----------------------------------------------------------------------------

\clearpage{}
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf KEGG enrichments for differentially expressed
genes in the comparison of glucose-grown and galactose-grown UA159 wild-type
cells.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq ( ua159only, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.keg.genes[c(4,3)],
                   cat.desc=smutans.keg.desc)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:kegseqUA159Only}
\end{table}
%-----------------------------------------------------------------------------

\clearpage{}
%-----------------------------------------------------------------------------
\begin{table}[!h]
\caption{{\bf KEGG enrichments for differentially expressed
genes in the comparison of glucose-grown and galactose-grown TW1 mutant-type
cells.}}
\vspace{1ex}
\noindent \begin{centering}
\begin{tabular}{cl}
\hline 
$q^a$ & Description \\
\hline 
%-----------------------------------------------------------------------------
<<echo=FALSE,results=tex>>=
smutans.de2Goseq ( tw1only, qval=1e-3, 
                   feature.gene=smutans.feature.genes,
                   go.genes=smutans.keg.genes[c(4,3)],
                   cat.desc=smutans.keg.desc)
@
%-----------------------------------------------------------------------------
\hline 
\end{tabular}
\par\end{centering}
\begin{flushleft}
$^a$False discovery rate estimated by the
Benjamini-Hochberg method.  All categories having at least ten genes and
$q\leq 0.05$ are displayed.\\ 
\end{flushleft}
\label{tab:kegseqTW1Only}
\end{table}
%-----------------------------------------------------------------------------

% END OF TABLE

%\section{Supplementary Figures}

\begin{figure}
\centering
\includegraphics[width=.9\textwidth]{supplementary-figClust}
\caption{Heatmap showing the Euclidean distances between the samples
  as calculated from the variance-stabilizing transformation of the
  count data.}
\label{figClust}
\end{figure}
<<figClust,echo=FALSE,result=hide,fig=TRUE>>=
allSample <- newSmutans( smutansGenes )
smutans.de2Clust( allSample )
@



\end{document}
