%\VignetteIndexEntry{Analyzing Streptococcus mutans RNA-seq data}
%\VignettePackage{smutans}

% To compile this document
% library('weaver'); rm(list=ls()); Sweave('smutans.Rnw', driver=weaver()); system('pdflatex smutans')

\documentclass{article}

\usepackage{Sweave}
\usepackage[letterpaper]{geometry}
\usepackage{hyperref,graphicx}
\usepackage{color}
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

\SweaveOpts{keep.source=TRUE,eps=FALSE,include=FALSE,width=4,height=4.5}
\newcommand{\Robject}[1]{object \texttt{#1}}
\newcommand{\Rpackage}[1]{R package \textit{#1}}
\newcommand{\Rclass}[1]{class \textit{#1}}
\newcommand{\Rfunction}[1]{{function \small\texttt{#1}}}
\newcommand{\program}[1]{{\texttt{#1}}}
\newcommand{\fixme}[1]{{\textbf{Fixme:} \textit{\textcolor{blue}{#1}}}}

\renewcommand{\floatpagefraction}{0.7}

\author{Sang Chul Choi\\[1em]Cornell University,\\ Ithaca, NY\\[1em]
\texttt{schoi@cornell.edu}}

\title{\textsf{\textbf{A Toturial of BioConductor R packages}}}

\begin{document}

\maketitle

\begin{abstract}
BioConductor R packages have been useful in analyzing RNA-seq data sets. I wish
to record for the instructional purposes.
\end{abstract}
\tableofcontents

\section{Genomes}

\subsection{Sample sequences from a genome}

\section{Short reads}

Steps of analyzing RNA-seq data for determining differentially expressed genes
are described in \cite{Oshlack2010}. 

\subsection{Create short reads}

\subsection{Read short reads}

\Rpackage{ShortRead}

library(ShortRead)
fastqFile <- "/v4scratch/sc2265/rnaseq/output/ua159/1/bwa/FASTQ001.recovered.fq.gz"
fq <- readFastq(fastqFile, withIds=TRUE)
head(sread(fq), 3)
head(quality(fq), 3)
sampler <- FastqSampler(fastqFile, 100000)
fqSample <- yield(sampler)
report(qa(fqSample),dest="/v4scratch/sc2265/rnaseq/output/ua159/1/bwa/report001")

qa2 <- qa(fq, basename(fastqFile)) 
qa3 <- qa(fqSample, basename(fastqFile)) 
report(qa2,dest="/v4scratch/sc2265/rnaseq/output/ua159/1/bwa/report001")
X11 error in report

indexBam(reads_fn)

Reweight each read.


\subsection{Sample short reads}

\subsection{Correct bias in short reads}

library(seqbias) 
library(Rsamtools) 
ref_fn <- "/v4scratch/sc2265/rnaseq/output/ua159/1/data/NC_004350.fna"
ref_f <- FaFile( ref_fn ) 
open.FaFile( ref_f )
reads_fn <- "/v4scratch/sc2265/rnaseq/output/ua159/1/bwa/FASTQ001.sorted.bam"
indexBam(reads_fn)
ref_seqs <- scanFaIndex( ref_f )

ref_seq <- getSeq(ref_f)
I.all <- GRanges(seqnames=Rle(c("NC_004350.1"),c(2)),
                 ranges=IRanges(c(1,1),width=rep(width(ref_seq),2)),
                 strand=Rle(strand(c("+","-")),c(1,1)))
seqlengths(I.all) <- c(width(ref_seq))

I <- random.intervals( ref_seqs, n = 5, m = 100000 )
seqs <- scanFa( ref_f, I )
neg_idx <- as.logical( I@strand == '-' ) 
seqs[neg_idx] <- reverseComplement( seqs[neg_idx] )
counts <- count.reads( reads_fn, I, binary = T )
freqs <- kmer.freq( seqs, counts )
sb <- seqbias.fit( ref_fn, reads_fn, L = 5, R = 20 )
bias <- seqbias.predict( sb, I )
bias <- seqbias.predict( sb, I.all )
bias[[2]][99996]
seqbias.save(sb,"/v4scratch/sc2265/rnaseq/output/ua159/1/bwa/FASTQ001.yml")
sb <- seqbias.load( ref_fn, "/v4scratch/sc2265/rnaseq/output/ua159/1/bwa/FASTQ001.yml")
sb5 <- seqbias.fit( ref_fn, reads_fn, n=5e5, L = 5, R = 20 )

N <- width(ref_seq) # genome length
cnt <- count.reads(reads_fn, I.all, sb = sb5)
cnt <- count.reads(reads_fn, GRanges("seq1", IRanges(start = 1, width = N)), sb = sb5)

readlen <- 100
cov <- coverage(GRanges("NC_004350.1", IRanges(start = 1:N, width = readlen)), weight = cnt[[1]])

\subsection{Compute the coverage of short reads}

bamFile <- "/v4scratch/sc2265/rnaseq/output/ua159/1/bwa/FASTQ001.sorted.bam"
aln <- readGappedAlignments(bamFile)
aln <- as(aln, "GRanges")
coverage(aln)
cov <- coverage(aln, weight = cnt[[1]])

\subsection{Export coverage to a wiggle file}

rtracklayer
export()


\section{Session Info}
<<sessi>>=
sessionInfo()
@

\bibliographystyle{unsrt}
\bibliography{siepel-mutans}

\end{document}
